{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# W200 Introduction to Data Science Programming, UC Berkeley MIDS\n",
    "\n",
    "\n",
    "## Instructions\n",
    "The final exam is designed to evaluate your grasp of Python theory as well as Python coding.\n",
    "\n",
    "- This is an individual exam.\n",
    "- You have 24 hours to complete the exam, starting from the point at which you first access it.\n",
    "- You will be graded on the quality of your answers.  Use clear, persuasive arguments based on concepts we covered in class.\n",
    "- Please double-click the markdown cells where it says \"Your answer here\" to input answers (if you need more cells please make them markdown cells)\n",
    "- Use only Python standard libraries, NumPy and Pandas for this exam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Samuel Gomez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numpy version: 1.17.2\n",
      "pandas version: 0.25.1\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import datetime,timedelta\n",
    "import matplotlib.ticker as ticker\n",
    "import re\n",
    "\n",
    "print(\"numpy version:\", np.__version__)\n",
    "print(\"pandas version:\", pd.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1: Short Answer Questions (25 pts - each question = 5 pts)\n",
    "\n",
    "a) The following class ```Cart``` and method ```add_to_cart``` are parts of a larger program used by a mobile phone company.  The method ```add_to_cart``` will work when an object of type MobileDevice or of type ServiceContract is passed to it. State whether the method ```add_to_cart``` is a demonstration of the following items (yes/no) and the reasoning (1-2 sentences): \n",
    "\n",
    "    1. Inheritance\n",
    "    2. Polymorphism\n",
    "    3. Duck typing\n",
    "    4. Top-down design\n",
    "    5. Functional programming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method:\n",
    "class Cart():\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.cart = []\n",
    "        self.total = 0\n",
    "    \n",
    "    def add_to_cart(item):\n",
    "        self.cart.append(item)\n",
    "        self.total += item.price"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- a) Your answer here\n",
    "\n",
    "\n",
    "1. Inheritance:  No. The method add_to_cart is not demonstrating inheritance.  First of all, for to utilize inheritance for add_to_cart, there would need to be a class say 'xyz' like so Cart('xyz').  Second, add_to_cart would inherit it's function from the base class, not define its own function.  If Cart() was a child class, add_to_method would override add_to_cart of its parent class.\n",
    "\n",
    "\n",
    "2. Polymorphism:  No.  In OOP, polymorphism refers to the ability of a subclass method to override a base class method.  Also, two child classes of the same parent class can have the same method, but that method can perform a different functions.  The fact that class Cart() is not a subclass negates this possibility.    \n",
    "\n",
    "\n",
    "3. Duck typing:  Yes.  Since the problem statement states add_to_cart works when object types MobileDevice and ServiceContract are passed to it, I can infer that it is demonstrating duck typing.  Duck typing allows different objects of different classes to execute their own function of the same method e.g. add_to_cart.\n",
    "\n",
    "\n",
    "4. Top-down design:  No.  OOP is classified as a bottom-up style of programming.  It typically adds up modules to perform more complex tasks.  Creating classes, subclasses, and methods are characteristics of bottom-up programming.\n",
    "\n",
    "\n",
    "5. Functional programming:  No.  This is not functional programming; this is object-oriented programming.  Functional programming is declarative uses of functions to perform computation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Suppose you have a long list of digits (0-9) that you want to write to a file.  From a storage standpoint, would it be more efficient to use ASCII or UTF-8 as an encoding? What is the most efficient way to create an even smaller file to store the information?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- b) Your answer here\n",
    "\n",
    "ASCII and UTF-8 are equally efficient for encoding integers.  The difference between the two is realized when encoding other characters.  ASCII would be more efficient than UTF-8 in those situations.\n",
    "\n",
    "Encoding the integer in binary form should be used to create smaller file storage."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "c) Why is it important to sanity-check your data before you begin your analysis? What could happen if you don't?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "- c) Your answer here\n",
    "\n",
    "Performing a sanity-check before data analysis is critical.  I haven't done much data analysis prior to this course.  However, just like a lot of the engineering and math applications I have done in the past, it is always best to follow a process.  A data sanity-check should be the step in the data analysis process between importing the data and doing the analysis.  If a proper sanity-check is not performed, the conclusion from the analysis could be wrong, probably more often than not.  An adequate check should verify that all of the data was imported at the very least.  The next step should be checking for correctness  Checking each variable for consistency is also a good practice.  Looking at data distribution for outliers and identifying duplicates in the data is a good check as well.  Only after a thorough data exploration is complete should the analysis begin."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "d) How do you determine which variables in your dataset you should check for issues prior to starting an analysis? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "- d) Your answer here\n",
    "\n",
    "Typically, the nature of the critical data depending on the ensuing analysis dictates which variables I check first or which I check more thoroughly. If I have prior experience with a data set from a particular source, I will check the variables that may have been imported incorrectly in the past.  Next, if the analysis has a constraint--such as the 2016 primary donation HW with donations made to the primary 2016 election--I will check the variables of the constraint for accuracy.  If a specific variable can't have a NaN I will check that as well.   Verifying all variables in a data set is a good practice if time permits."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "e1) Explain why the following code prints what it does."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'function'>\n"
     ]
    }
   ],
   "source": [
    "def f(): pass\n",
    "print(type(f))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- e1) The function type( ) in Python returns the class type of the argument passed as a parameter.  Function f( ) is first passed into type( ); this returns <class 'function'>.  Then, the the print statement prints the output of the type method. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "e2) Explain why the following code prints something different."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "def f(): pass\n",
    "print(type(f()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- e2) This code prints something different because the function f( ) is executed, then the type method is invoked.  Since the function returns nothing (pass) the type( ) is 'NoneType'.  If I were to write 'return 1' in function f( ) instead of 'pass', this code would return type 'int'."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2:  General Coding Questions (15 pts - each question 5 pts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Using a list comprehension: Make a list of the squared numbers that are the square of non-negative integer less than 10, but whose squares are greater than 25. Fill in a list comprehension below so that we get this desired output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[36, 49, 64, 81]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2a) Your code here\n",
    "\n",
    "squares = [num**2 for num in range(10) if num**2 > 25]\n",
    "squares"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\ndef map_compute(my_list):\\n    return list(filter(lambda x: x > 25, map(lambda x: x**2, my_list)))\\n\\nmap_compute(range(0,10))\\n'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "def map_compute(my_list):\n",
    "    return list(filter(lambda x: x > 25, map(lambda x: x**2, my_list)))\n",
    "\n",
    "map_compute(range(0,10))\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Below is a data frame of customers that have different cooling systems. Your data science team lead wants the column ```cooling_system``` to be labeled with the numbers 1-4 instead of the text as shown below:\n",
    "\n",
    "- 1 = Air Conditioning / AC / Air Con\n",
    "- 2 = Heat Pump / HP\n",
    "- 3 = Evaporative Cooler / Evap Cooler\n",
    "- 4 = Fan\n",
    "\n",
    "Make a new column called ```cooling_type``` that maps the text values to the new numeric values. Filter out the values that are not included in the mapping above. Print out this new data frame. Be sure to list any assumptions also!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cust_Number</th>\n",
       "      <th>Cooling_System</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>102</td>\n",
       "      <td>Heat Pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>103</td>\n",
       "      <td>Air Con</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>104</td>\n",
       "      <td>Air Conditioning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>Fan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>106</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>107</td>\n",
       "      <td>Evap Cooler</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>108</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>109</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>110</td>\n",
       "      <td>Evaporative Cooler</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>111</td>\n",
       "      <td>geothermal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>112</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Cust_Number      Cooling_System\n",
       "0           101                  AC\n",
       "1           102           Heat Pump\n",
       "2           103             Air Con\n",
       "3           104    Air Conditioning\n",
       "4           105                 Fan\n",
       "5           106                None\n",
       "6           107         Evap Cooler\n",
       "7           108                None\n",
       "8           109                  AC\n",
       "9           110  Evaporative Cooler\n",
       "10          111          geothermal\n",
       "11          112                   1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas\n",
    "\n",
    "# creating a data frame from scratch - list of lists\n",
    "\n",
    "data = [ [101, 'AC'], \n",
    "         [102, 'Heat Pump'], \n",
    "         [103, 'Air Con'], \n",
    "         [104, 'Air Conditioning'],\n",
    "         [105, 'Fan'],\n",
    "         [106, 'None'],\n",
    "         [107, 'Evap Cooler'],\n",
    "         [108, None],\n",
    "         [109, 'AC'],\n",
    "         [110, 'Evaporative Cooler'],\n",
    "         [111, 'geothermal'],\n",
    "         [112, 1]\n",
    "       ]\n",
    "\n",
    "# create a data frame with column names - list of lists\n",
    "\n",
    "col_names = ['Cust_Number', 'Cooling_System']\n",
    "df = pandas.DataFrame(data, columns=col_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cust_Number</th>\n",
       "      <th>Cooling_System</th>\n",
       "      <th>cooling_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>102</td>\n",
       "      <td>Heat Pump</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>103</td>\n",
       "      <td>Air Con</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>104</td>\n",
       "      <td>Air Conditioning</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>Fan</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>106</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>107</td>\n",
       "      <td>Evap Cooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>108</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>109</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>110</td>\n",
       "      <td>Evaporative Cooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>111</td>\n",
       "      <td>geothermal</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>112</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Cust_Number      Cooling_System cooling_type\n",
       "0           101                  AC            1\n",
       "1           102           Heat Pump            2\n",
       "2           103             Air Con            1\n",
       "3           104    Air Conditioning            1\n",
       "4           105                 Fan            4\n",
       "5           106                None          NaN\n",
       "6           107         Evap Cooler            3\n",
       "7           108                None          NaN\n",
       "8           109                  AC            1\n",
       "9           110  Evaporative Cooler            3\n",
       "10          111          geothermal          NaN\n",
       "11          112                   1          NaN"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2b) Your code here\n",
    "df['cooling_type'] = df.Cooling_System.map({'AC':'1','Air Con':'1','Air Conditioning':'1', 'HP':'2', \n",
    "                                            'Heat Pump':'2', 'Evap Cooler':'3', 'Evaporative Cooler':'3', \n",
    "                                            'Fan':'4'})\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cust_Number</th>\n",
       "      <th>Cooling_System</th>\n",
       "      <th>cooling_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>102</td>\n",
       "      <td>Heat Pump</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>103</td>\n",
       "      <td>Air Con</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>104</td>\n",
       "      <td>Air Conditioning</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>Fan</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>107</td>\n",
       "      <td>Evap Cooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>109</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>110</td>\n",
       "      <td>Evaporative Cooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cust_Number      Cooling_System cooling_type\n",
       "0          101                  AC            1\n",
       "1          102           Heat Pump            2\n",
       "2          103             Air Con            1\n",
       "3          104    Air Conditioning            1\n",
       "4          105                 Fan            4\n",
       "5          107         Evap Cooler            3\n",
       "6          109                  AC            1\n",
       "7          110  Evaporative Cooler            3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[df.cooling_type.notnull()]\n",
    "df.reset_index(inplace=True, drop=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "c) From the dataframe below, use groupby/aggregate in Pandas to show how many total delegates were obtained grouped by favorite color. Print this out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>delegates</th>\n",
       "      <th>color</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>marco</td>\n",
       "      <td>165</td>\n",
       "      <td>blue</td>\n",
       "      <td>FL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>jeb</td>\n",
       "      <td>0</td>\n",
       "      <td>red</td>\n",
       "      <td>FL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chris</td>\n",
       "      <td>0</td>\n",
       "      <td>white</td>\n",
       "      <td>NJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>donald</td>\n",
       "      <td>1543</td>\n",
       "      <td>white</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ted</td>\n",
       "      <td>559</td>\n",
       "      <td>blue</td>\n",
       "      <td>TX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>john</td>\n",
       "      <td>161</td>\n",
       "      <td>red</td>\n",
       "      <td>OH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name  delegates  color state\n",
       "0   marco        165   blue    FL\n",
       "1     jeb          0    red    FL\n",
       "2   chris          0  white    NJ\n",
       "3  donald       1543  white    NY\n",
       "4     ted        559   blue    TX\n",
       "5    john        161    red    OH"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# creating a data frame from scratch - list of lists\n",
    "\n",
    "data = [ ['marco', 165, 'blue', 'FL'], \n",
    "         ['jeb', 0, 'red', 'FL'], \n",
    "         ['chris', 0, 'white', 'NJ'], \n",
    "         ['donald', 1543, 'white', 'NY'],\n",
    "         ['ted', 559, 'blue', 'TX'],\n",
    "         ['john', 161, 'red', 'OH']\n",
    "       ]\n",
    "\n",
    "# create a data frame with column names - list of lists\n",
    "\n",
    "col_names = ['name', 'delegates', 'color', 'state']\n",
    "df = pandas.DataFrame(data, columns=col_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>delegates</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>color</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>blue</td>\n",
       "      <td>724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>red</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>white</td>\n",
       "      <td>1543</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       delegates\n",
       "color           \n",
       "blue         724\n",
       "red          161\n",
       "white       1543"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2c) Your code here\n",
    "df.groupby('color').agg('sum')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3: Bitcoin coding problem (20 points):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Bitcoin\n",
    "\n",
    "Consider a record of a one-time investment in bitcoin with value of that investment tracked monthly, provided as an (ordered) tuple of dictionaries, where each dictionary comprises one key for the month and corresponding one value for the value of the investment, and the first entry (Jan 2018) is the initial investment, like in `data` below.\n",
    "\n",
    "Write Python code to take such a record of any length, and output a table/dataframe comprising a row for each month with columns for date, start balance, and return. **Print out this table/dataframe.**\n",
    "\n",
    "Also, visualize the record as two vertically arranged plots. \n",
    "- The top plot should show a line plot of start balance vs. month\n",
    "- The bottom plot should show a bar plot of return vs. month, with a black horizontal line at return=0, and bars color-coded such that positive returns are green and negative returns are red.  \n",
    "- The two plots' horizontal axes should align. Demonstrate that your code works by applying it to `data`.\n",
    "\n",
    "Notes: <br>\n",
    "- The gain for each period is the end balance minus the start balance. \n",
    "- The growth factor for each period is the end balance divided by the start balance.  \n",
    "- The return for each period is the growth factor minus 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'Jan 2018': 1000},\n",
       " {'Feb 2018': 1100},\n",
       " {'Mar 2018': 1400},\n",
       " {'Apr 2018': 700},\n",
       " {'May 2018': 800},\n",
       " {'Jun 2018': 500})"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = ({\"Jan 2018\":1000},{\"Feb 2018\":1100},{\"Mar 2018\":1400},{\"Apr 2018\":700},{\"May 2018\":800},{\"Jun 2018\":500})\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Line plot of start vs. month"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>invest_balance</th>\n",
       "      <th>invest_return</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Jan 2018</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Feb 2018</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>27.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Mar 2018</td>\n",
       "      <td>1400.0</td>\n",
       "      <td>-50.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Apr 2018</td>\n",
       "      <td>700.0</td>\n",
       "      <td>14.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>May 2018</td>\n",
       "      <td>800.0</td>\n",
       "      <td>-37.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>Jun 2018</td>\n",
       "      <td>500.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       date  invest_balance  invest_return\n",
       "0  Jan 2018          1000.0          10.00\n",
       "1  Feb 2018          1100.0          27.27\n",
       "2  Mar 2018          1400.0         -50.00\n",
       "3  Apr 2018           700.0          14.29\n",
       "4  May 2018           800.0         -37.50\n",
       "5  Jun 2018           500.0            NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnkAAALTCAYAAABqjhQpAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nOzdd3xUZdr/8c+VCgk1BVBaqLFgBaQJREXXLRbWvjbWXWXVXXXZfZ4tP9e2fZ9n1/aAgg1de0VdXddGQCkKKiKI9NBLSKgJpN6/P85JMjOkTMgkk/J9v17zypz7lLnmTMqV69z3fcw5h4iIiIi0LjHRDkBEREREIk9JnoiIiEgrpCRPREREpBVSkiciIiLSCinJExEREWmFlOSJiIiItEJK8kRaGTPLNjPnP7Lrsd9dAftpbiVpM470Z0akuVOSJ22GmWUEJjH+Y7uZJdaw/ZfVbJ/RtFEfFlNgLHdFM5aGCE0ozSwr2jE1ttby2dWloQmTmWVV83PnzKzMzPab2Tdm9k8zG98I4Yu0KkrypK3rDlwZ2mhmE4ATmz4cEalBDNAByASuArLNbHJ0QxJp3pTkicDPq2n7RZNHISLVeQ/4L+BXwJNAYFeCP5pZbFSiEmkBlORJW1bmfx1iZmdXNJrZccC3QrapkZmNNrNnzGy9mR0yswIz+9rMHjCz/tVsPzPgElSOmXUys7/6+xeZ2UYz+4uZJQTsk11NP7k7w7mUbGapZvagmW32j7/azP7LzCyM9xZnZpsCXuPearaZFHJJrVddx63jNUMv12WZ2aVmtsA/t3vN7E0zOyZgnwFmVh6wzxXVHPeBkMv0cSHn6E4zW+Qfv9g/X8+a2fAa4rzczN41sx1mVuJfSswxs3fM7Pdm1sPfLuzPLvRSp5kNNLOXzWy3me0xs9fMbIC/7Ulm9paZ7fMfb5nZsTXEWq/3F/KZOjPrb2Y/MbMlZnbQzHb5+x4dsM9d/vsMvIw6PuQ4k6qLrw7znXP/65z7m3NuEvBWwLpUID0ghjj/3L9lZmvMLN//bPaY2WJ/Xdf6vLiZXWdmz5vZcjPb6Z+7A2a2wsymV3fO7fDuCIlm9hvzLjUX+d9/D5tZpxpes7+Z/d0/33v9fTaZ2b/N7LJqth9hZk+Z2Tr/8ykws6Vmdo+ZpdTn/Uor45zTQ4828QAy8KoAFY9XAp6/HbDdowHtL4fskxFyzHuA8pBtAh8FwIUh+8wMWL8LWF7Dvk8E7JNdy2sExRay7TfAuhq2vzMkrrsC1we0/yagPQ9oF7LfO9Wdxzo+i7tCYskKWJcVsm5uDfHvBNIC9psdsO6tkNeLBbYHrP9rwLphIetCH6XAT0OO99swPo+sBn526/zzHbrtDuB8oLCuc9KA9zcpzM9gBZBYw2da3WNSGN8boZ//XX57DHAssCFg3SEgIWDfDmHEkAP0CHnNwPOeHbJucR3HO0TA928N56Km8/dhNe//BzV8thWPWSHb30Htv4M2AJnR/v2rR3Qelf/JirRBXwEdgbOBc/3/yPPw+vsArMSrGlxU3c7+f9S/C2jKAV4AkoAf4v3BSQKeM7Mhzrm11RwmFegKPAVsBX4MpPnrrjGz3zrntgEPAf8C/idg3/eAdwOW86s5fibeH6GHgIPAjUB7f90UM/uTc66kuvcXYIb/PtsDKcAlwD8BzCwNOCtg20frONaRGAssAv4DnAGM8dvTgR8Bfw147Sz/+Tlmlu6cy/WXJ+D1vwyK08w6Am8GrNsBPId3LicA4/ASxPvNbIlz7mN/u1sCjrUY77MB6I3Xl3NowPoj/ez64X0//g3oD1zst3cDXgdygQeBwcCF/rqgc9KA9xdqLPABMN9/rRP89mP85Rf893MA73usooK9zn//FRbVcPza3Glmd9aw7n+dc8UByw5YDywEtgC78d5fP+AyvJ/HvsDtwE/DfP1cvM9vjX+8EqAHMBHv804E/g8YUssxxgKvAV/j9QHO8NvPMLMRzrlPAMxsGN4l6Yq/zQ7v8/sC73fF2MCDmtnFwN0BTfPwvreSgav9OPsAr5nZCc65Oq9MSCsT7SxTDz2a6sHhlby78C7LVizPwKvMVSz/hMMrGhkBx/ssoH0PwVWlM0L2uz9g3cyQdbcGrLsgZN15Ie/hsApHNe8zO2S7CwLW3Rqy7oSAdXcFrgs55oyAdR8HtE8OaN8BxIf5WdwVEkdWwLqskHWfVBwXiPdfp2LdKwH7tcNLXirW/Sxg3VMB7dkB7T8NaD8E9A5YZ3jJwmEVFP/zrmgfWc37SwM6R+CzGxOwbkvIupF+ewzePwjVnZMjfX+TQl7rVcD8dSl41b+KdX+v5T1kV/c+6/jeCP38a3q8XNP3m3/+v4v3MzwF+CUwJ2DftfWJ2f/eysJLoG/zj/d4SDyB5/aukHX3Bqw7KWRd4PfpSyHrLq0mlgEBzwOrjG9VfEb+umNDjnVhTedcj9b7UCVP2jTn3H/MbDlwPN5/vgX+qjy8/6gP6/8CYGZJwCkBTW8653YFHHe2meVQ9R/7GKpXBkwPWF4Zsr5e/YeqsdU593oEjn8/cL3/fIyZHe+cWw5cHrDNk67uquCReLTiuM65EjNbj1fNgoD4nXOHzOwZqio0VwEP+p/VxMDjBTwPrIwkAhut5q6Kpwc8nwuc5z9/z8w+wav0rMJLnBY658rDfH812eCcmxe4DFT0gVvvnFsI4JwrN7N1wFH+usDP9EjfX6iHnPMyB+dcvpntoqo62NDv0bpUVD0Nr3J2DdAZr8L+jpl9xzlXBGBm7fCqapPwKng16Rnui5vZrXj//FXbfy5AL2BTDeumBTyv7WdwXMDzJc65F0MP5PwrAv739akBq74DlNfx+c6qaaW0Thp4IQL/8L+2w7skAvCwc+5gLft0xfujU2F7NdsEttXU+XmHc+5QwHJRyPqG/ozmhCwf0fH9hO6DgKbJZnYUwX+UHqt3dOHJCVkOfA+h8T8S8Pw0MxuE13+tg9+2G68CVKE+ndJTzKzi9SbjVYbwj32W3/Z3vEtmK80ssx7Hrs6WkOXAy5JbQ9aVBjwPPCdH+v5C5YQs1/YZRFrFwIv/cc7dQnDCfiZwQ8Dyn/CqbXWNuK12bsxQZnY+cB91J3h1HTMn4HltP4OBn9f6Ol4v9HdQXdLr3kRaG1XyROAZvD8OFZWJYrxqQG12410Cqfgl26OabQLbqutzBV7/nkCujtetr0ge/z6q+t9djZdoVPyB+tg5F1qhiJSw34NzbqmZLcYbbABeNS+w2vFMSFId+LnsA35fRywV1axtQJaZ9QFOAwbh9X+8AOgCDMTri3ZmHcerTW1V0dJa1gU6ovcXRiyR/j6tj09Cls/A65sIwZXlZXiDGL7xK8B/w5uKpT4Cj1eA1y9yjnPuoJl9h+CRvjUKrHA751wt1bZ8qqrU/eo4bOjvoA+Bf9ey/dd1RyqtjZI8afOcc0VmNhXvkgzAc8656ipzgfsUmtkSqi7ZnmdmaRWXbM27g0NGwC7ziIxSqn5ukyJ0zHC9hXdJciBeInNnwLrGGHBxpB6lKsm7jmoGXAT4GLjUf94J+Mw5Nzv0gGY2BOhSccnSzE4CljnnNgIbA7abglfNIyCGCtH47I7o/TVQYELYGO/ztJDl+IDnaQHPZzvnvgIws/Z4Fd36CjzeOufcOwHLl4duHAFzqRpgc7KZXeSceyVwAzPLcM7l+L+DvqDqn5geeFcgDoRsH4/XtWBBI8QrzZySPBHPA3ijbQE+DXOf/wGe9Z93BhaZ2fN4f9iuC9iuiKpKQ0Ntpip5nGRmRcBeYJdzbmaEXqNafgXiQbz+eeBd3gavQvRSY752PT2Ll2gl4/WTqrDIOfdlyLZPAv+PqkTw32ZWMQrS8M71GLwRrHfjJU3gVX+7m9mHeJ/JLrzLYVcHHHt3yGtF47M70vfXEJsDng81sweoSoSn1tENojqjzeyXePH2IvgcQ3DMK6ka5Xq9P2/fPrwR4Udy+Xwl3uh7gBPM7AW8CmEWDavS1uSveJejKy43v2RmrwNL8H7HjMbrBnJhwPYv+M+PA5b7n+92vKT+eD/WTniVwdDvSWnllOSJAM65vdSzU7Jz7jkzOwFvHjnw/mD+OmSzg8BVzrk1DQ7S8xJVl5zS8aaCAG+uvZkReo3aPIF3yS+wj9KzzrnCJnjtsDjn9pvZS3id7wMdVm10zu0zs/OAN/AqIYmEX6FJo6pKVp0/hyw3+WfXwPd3pALPfQzws4B1M/F+JurjbKoSrVBfENy14h6gYrBCO6qmutmPNy9mtdMh1eI+vIEeFd/vl1L1mT+BN1VSxDjnFpvZtXh9S9vjJbYXUpXUgTd9TsX2L/pTP93pb9sHbwS9CKCBFyIN4pz7Ld7gg+fwqhXFeH/EVgJTgROdc69G8CV/hzdvWg7h98uKGOfcfrypIwI1p0u1FR4JWS7A+4wO45xbhFfxuB1vZOwevHO7HfjcP9b5VM3HB151bCpe1XcLXrW2GG905avAt5xzD4e8VFQ+uyN8fw15vbfx5nv8ksMHGTRUCd6Ez7PxkpnRzrmKEfE4514Cvo83vVExXuXqDWAkXgWuXvyRrGPx+roV4E1SvBDvfD3VkDdSy2s+gzcP4b3AUry5B0vwPq/3gOdDtr8b7xL248BqvN8/B/EGbszBmyz5FOdcTmPEK81bxbxHIiJhMbPr8ebNA2+ah1Nq215ERKJDl2tFpE7m3Yf1GLw+UYEDLu6vfg8REYk2JXkiEo5z8fogBVqIf3szERFpftQnT0Tqw+H1QZsKfNfpXpgiIs2W+uSJiIiItEK6XBsiLS3NZWRkRDuMNq+goIDk5ORoh9Fs6fzUTeeodjo/ddM5qp3OT/Pw2Wef7XLOVXvbOiV5ITIyMli8eHG0w2jzsrOzycrKinYYzZbOT910jmqn81M3naPa6fw0D2a2oaZ16pMnIiIi0gopyRMRERFphZTkiYiIiLRCSvJEREREWiENvBCRVqOgqJTpc9fy9IIN5BeWkDL3Xa4a1ZfJ4waQnKhfdyLStkSlkmdmGWY2KWB5nJl9bmalZnZxHfvGmtkXZvavgLZ+ZvaJma02sxfMLMFvT/SX1/jrMxrpLYlIlBUUlTJx2jymz1lHfmEJAPmFJUyfs46J0+ZRUFQa5QhFRJpWkyd5ZnYj8B/g92aW7d8TcyMwCXg2jEPcCqwIafsrcK9zbhCwG/iR3/4jYLdzbiBwr7+diLRC0+euJSevkKLS8qD2otJyNuQVMn3u2ihFJiISHU2a5JlZR+Bu4Brgd3iJXYFzLsc5txQor2V3zKwX8F3g0YA2A84EXvabngQu9J9f4C/jrz/L315EWpkn52+guLT6XyFFpeU8vXBjE0ckIhJdTd1JpRxIADoBOOdy6rn/fcB/Ax0D2lKBPc65imsxm4Ge/vOewCb/tUrNbK+//a7Ag5rZDcANAN27dyc7O7ueYUmkHThwQJ9DLXR+gq3IK2PvwZJat9ldUKxzFkDfQ3XTOaqdzk/z16RJnnOuwMyuAf4E9DCzIcAdzrnCuvY1s+8BO51zn5lZVuCq6l4qjHWBcc0AZgAMGzbMaQbv6NNM6rXT+any4uJN/OO9r+rcrmtygs5ZAH0P1U3nqHY6P81fk/fJc869AVwC/A1IB34R5q5jgPPNLAd4HjjTzJ7Gq8p1MbOKhLUXsNV/vhnoDeCv7wzkR+BtiEiUlZc7/vrON/z3y0spKTvsf7cgsTHGVSP7NFFkIiLNQ1P3yetgZn39xf14Ayg61rILZvYNgHPuN865Xs65DOBy4EPn3FXOOQfMBipG5V4LvO4/f8Nfxl//ob+9iLRgB4vLuPnZz3ko+/DBFLExhxfwnXP84DQleSLStjR1JS8emI43COJO4NvA/WY23Mw241X4ppvZcgAzS6P6S66hfgVMMbM1eH3uHvPbHwNS/fYpwK8j+WZEpOnt3HeIy2cs4N/Ltge1J8bF8PdLTuLmMwbQNSk+aF25g+cXbWrKMEVEoq6p++TtBs7156vLcs7N9FdtwbvMGmokMLWa42QD2QHL64DTqtnuEF7iKCKtwIpt+/jRzEVs3XsoqD2tQyKPXjuMk3t3AWDK2ZlMeexdXl1dNRjjsY/X88Mx/ejcPjgBFBFpraJ1W7M9wJK6NnLO/cs590ATxCMizdyH3+zg4ofmH5bgHdOjI7NuHl2Z4FU4u288XQIqevsPlfLYx+ubJFYRkeYgKkmec26Pc67OJE9EBGDmvPX8+MnFFBSXBbVnZabz0k9G0atr0mH7tI8zrh/bP6jtiY/Xs7ew9qlWRERai2hV8kRE6lRaVs4dry/jrje/pjxkyNSk0Rk8es0wOrar+fLrtaMzgvrn7S8q5dGP1zVWuCIizYqSPBFplvYfKuFHTy7mqQUbgtpjDO4+/3juOv944mJr/xXWITGOG8YNCGp7Yl4OewqLIx6viEhzoyRPRJqdTfmFXPTQfOasyg1q75AYx2PXDufa0RlhH+uaUX1JSU6oXD5QVMojH6maJyKtn5I8EWlWPt+4m4nT5rFqx4Gg9p5d2vPyjaM445hu9TpecmIck8cF982bOS+H/AJV80SkdVOSJyLNxptfbuXyGQvZdSA4ATupdxdeu3k0x/TodETHvXpUX9I6VFXzCorLmDFX1TwRad2U5IlI1DnnePCD1fzsuS8oLi0PWvfdE47ihRtG0q1juyM+flJCHJND+uY9tSCHvANFR3xMEZHmTkmeiERVUWkZv3jxS/7+3qrD1v30jIE8eMUptIuPbfDrXDWyL2kdEiuXC1XNE5FWTkmeiERNfkExVz/6Ka9+sSWoPT7W+N9LTuKX38okppp70R6J9gmx3JgVWs3bwC5V80SklVKSJyJRsTb3ABOnzePTnPyg9i5J8Tz9oxFcPLS6Ox02zJUj+tCtY1U172BJGdPnrI3464iINAdK8kSkyc1fs4uJU+exIa8wqL1/WjKv3TSGEf1TG+V128UfXs3758IN7Nx/qIY9RERaLiV5ItKkXli0kWse/5R9h0qD2kf2T+HVm0bTLy25UV//itP60L1TVTXvUEk5D2erb56ItD5K8kSkSZSXO/789gp+9cpXlIbco+zSYb146roRdElKqGHvyGkXH8tNWQOD2p75ZAM796maJyKti5I8EWl0hcWl3PjMZ0yvZjTrr849hr9edCIJcU336+iy4b3p0alqSpai0nKmZatvnoi0LkryRKRR7dh3iMumL+Q/y3cEtbeLj+GhK0/lxqwBmEVmBG242sXHcvOZwdW8Zz/dyPa9quaJSOuhJE9EGs3yrXu54P/m8dWWvUHt6R0TeXHyKL59wlFRisy7RHx056pqXnFpOQ9lr4laPCIikaYkT0Qaxftf7+CShxewPaSv2zE9OvL6zWM4sVeXKEXmSYw7vJr33Keb2LrnYJQiEhGJLCV5IhJRzjke/Wgd1/9zMYXFZUHrzjymGy/fOJqju7SPUnTBLhnam54BsRSXlTNN1TwRaSWU5IlIxJSUlXP7rGX84a0VuOABtFw3ph+PXDOMDolx0QmuGglxMfw0pJr3wqJNbFE1T0RaASV5IhIR+w6VcN3MRTzzycag9tgY4/cXDuGO844jNkK3KIuki4f2olfXqmpeSZlj6mxV80Sk5VOSJyINtim/kIumzeej1buC2jsmxvH4pOFcPbJvlCKrW3xsDLecOSio7aXFm9i8u7CGPUREWgYleSLSIJ9tyOfCqfNYvfNAUHvPLu155abRjB+cHqXIwjfx1J70SUmqXFY1T0RaAyV5InLEXl+yhSse+YS8guKg9lP6dGHWzWMY3L1jlCKrn/jYGH4W0jfvpcWb2ZSvap6ItFxK8kSk3pxz3Pf+Km59fgnFpeVB67534lE8d/1I0jsm1rB38zTxlJ5kpFZV80rLHQ9+uDqKEYmINIySPBGpl0MlZfz8hSXc9/7hCdAtZw7kgctPoV18bBQia5i42Bh+FtI375XPt7AhryBKEYmINIySPBEJW96BIq569BNmLdka1J4QG8M/Lj2JKedkEtMMR9CG64KTj6ZfWnLlclm548EP1TdPRFomJXkiEpY1O/dz4bR5LN6wO6i9a1I8z1w/gu+f2itKkUVOXGwMt5wV3DfvtS+2kLNL1TwRaXmU5IlInT5evYuJ0+azKT94kuD+6cnMunkMwzNSohRZ5J1/Uk/6pwdX8x5Q3zwRaYGU5IlIrZ79ZCPXPvEp+w+VBrWPHpDKazeOoW9qcg17tkyxMcatZwX3zZv1xRbW5h6oYQ8RkeZJSZ6IVKus3PHHt77mt699RVl58D3KLh/emyevO43OSfFRiq5xfe/EoxnYrUPlcrmDBz9QNU9EWhYleSJymIKiUib/8zMe+Wh9ULsZ/PY7x/Dn759AfGzr/fVRXTXvjS+3smanqnki0nJE5be0mWWY2aSA5XFm9rmZlZrZxTXs09vMZpvZCjNbbma3BqxLMbP3zGy1/7Wr325m9oCZrTGzpWZ2aqO/OZEWbtveg1zy8ALeX7EjqL19fCwPXzWUG8YNwKzljqAN13dPOIrB3YOreQ+omiciLUiTJ3lmdiPwH+D3ZpZtZj2AjcAk4Nladi0FfuGcOxYYCdxsZsf5634NfOCcGwR84C8DfBsY5D9uAB6K8NsRaVWWbdnLhVPn8fW2fUHt3Tsl8tJPRvGt43tEKbKmFxNj3HrW4KC2N5duZfWO/VGKSESkfpo0yTOzjsDdwDXA7/ASuwLnXI5zbilQXtO+zrltzrnP/ef7gRVAT3/1BcCT/vMngQsD2p9ynoVAFzM7KrLvSqR1eHf5di55eAE79hUFtR93VCdm3TyGIT07Rymy6Pn2kB4c06Pq1mzOwf2q5olICxHXxK9XDiQAnQCcczlHchAzywBOAT7xm7o757b5x9xmZt389p7ApoBdN/tt20KOdwNepY/u3buTnZ19JGFJBB04cECfQy0ieX6cc7yTU8qLK4txIetO6RbL5ONLWfnFJ6yMyKs1nUidowk9Svlme9XyW0u3MbLjh/Tq2LL7JOpnrG46R7XT+Wn+mjTJc84VmNk1wJ+AHmY2BLjDORf2XcDNrAPwCnCbc25fXZtXF0Y1cc0AZgAMGzbMZWVlhRuONJLs7Gz0OdQsUuenpKycO15fxgsrNx227sen9+M33zmW2BZ6B4tInaNx5Y4PdnzMCv8StgPm7+vMtPOGNvjY0aSfsbrpHNVO56f5a/J/RZ1zbwCXAH8D0oFfhLuvmcXjJXjPOOdeDVi1o+IyrP91p9++GegdsF0vIPh+TCJt1N7CEiY98SnPfRqc4MXGGH+cOITbv3dci03wIikmxrhtQvBI27e/2l6Z9ImINFdN3Sevg5n19Rcr+tV1rGUXzOwb/6sBjwErnHP/CNnsDeBa//m1wOsB7df4o2xHAnsrLuuKtGUb8gr4/kPzmLcmL6i9Y7s4Zv5wOFeO6FvDnm3TOcd15/ijOwW13f+++uaJSPPW1JW8eGA68DJwJ97o1/vNbLiZbcar8E03s+UAZpZG1SXXMcDVwJlmtsR/fMdf9xfgbDNbDZztLwO8DawD1gCPADc19hsUae4W5+Qzcdp81uYG34+1d0p7Xr1xNGMHpUcpsubLzLhtQvBI23eWb2f51r1RikhEpG5N3SdvN3CuP3Aiyzk301+1Be9SaqiRwFR/34+pvo8dzrk84Kxq2h1wc0PjFmktZn2xhf9+eSnFZcED2Yf27cqMq4eS2iExSpE1fxOO7cYJPTvz1ZaqxO7+91cz45phUYxKRKRm0RoetgdYUtdGzrl/OeceaIJ4RFo15xz/eG8Vt72w5LAE74KTj+aZH49QglcHM+PnZwf3zXv36x0s26Jqnog0T1FJ8pxze5xzdSZ5ItJwh0rKuOX5JdXereG2CYO477KTaRcfG4XIWp4zMrtxUu8uQW33vb8qStGIiNSuZU/0JCK12nWgiB88spA3vwweVJ4QG8P9l5/MbRMGt4lblEWK1zcvuJr3/oqdLN28J0oRiYjUTEmeSCu1asd+Lpw6j883BicgKckJPHv9CC44uWcNe0ptsganc/Jh1TyNtBWR5kdJnkgrNHdVLhdNm8/m3QeD2gd268Csm8YwLCMlSpG1fF7fvOCRth9+s5Mlm1TNE5HmRUmeSCvzz4Ub+OHMRewvKg1qHzsojVduHE2f1KQoRdZ6jBuUxql91DdPRJo3JXkirURZueOeN7/md7OWUVYefPe+H4zow+OThtO5fXyUomtdzIwpZ2cGtWWvzOXzjbujFJGIyOGU5Im0AgeKSrnhqcU8Pm99ULsZ3P7dY/njhUOIj9WPeySNGZjK8IyuQW33vqdqnog0H/qtL9LCbd1zkEseXsAH3+wMak9KiGXG1cP48dj+GkHbCMyMn4fcBeOj1bv4bEN+lCISEQmmJE+kBftq814unDqPFdv2BbX36NSOFyeP4uzjukcpsrZh1IBUTusXPIjl3vc00lZEmocmva2ZiBy5gqJSps9dy9MLNpBfWEKH2e9wsKSMkBtYMKRnJx67djjdO7WLTqBtSEU174pHFla2fbxmF4ty8hmuEcwiEmWq5Im0AAVFpUycNo/pc9aRX1gCwIGiwxO8c47rzouTRynBa0KjBqQysn9oNU9980Qk+pTkibQA0+euZUNeIUWl5TVuM3lcfx6+aihJCSrQN7XQvnnz1+axcF1elKIREfEoyRNpxtbvKmDmvPU8lL221gQvOSGW33znWGJiNMAiGkb0T2XMwNSgNlXzRCTa9C+/SDNSWFzKgrV5zFmVy5xVuWzIKwxvv5KyRo5M6vLzCYOZt2ZB5fIn6/OZv3YXowekRTEqEWnLlOSJRJFzjjU7DzBnVS7ZK3P5dH0+xaEd7cLQNSmhEaKT+hiWkcLYQWl8tHpXZdt9761mVP9UTWEjIlGhJE+kie0/VML8tXlkr8xl7qpctuw5WPdOtUiMi+GqkX0iFJ00xG0TBgcleZ/m5DN/bR5jBqqaJyJNT0meSCNzzvHN9v1kr8xlzqqdLM7ZTWnIbcdqkhAbw4j+KYzqn8qLizexbe+hoL55iXEx9E1NYvK4AY0VvtTD0L5dGTc4nbmrcivb7n1vFaMHqJonIk1PSZ5II9h7sIvAF1EAACAASURBVISPV+9izqqdzFmVy459RWHv2ycliazMdLIy0xnZP7VytOy1ozO8efIWbmR3QTFdkxO4amQfJo8bQHKifpSbi59PGBSU5C3esJuP1+xi7KD0KEYlIm2R/jKIREB5uWP51n3MWbWT7JW5fLFpD2VhVusS42IYNSCVrMHpjM/sRr+05Gq3S06MY8rZmUw5O5Ps7GyysrIi+A4kUk7p05UzMtOZvTK4mnf6wDRV80SkSSnJEzlC+QXFfLQ6lzkrc5m7OpddB4rD3rd/ejJZg7sxPjOdEf1SaBcf24iRSlO7bcLgoCTv8417mLMql6zMblGMSkTaGiV5ImEqK3cs3bzH71uXy5eb9+DCK9aRlBDL6AFpjM9MJ2twOr1Tkho3WImqk3p34axjuvHBNzsr2+59fzXjB6ermiciTUZJnkgtcvcXMdefs+6j1bns9m8pFo7M7h0rk7qhGV1JjFO1ri25bcLgoCTvy03ePwhnHKNqnog0DSV5IgFKy8r5YtMe5qzMJXvVTpZt2Rf2vh0T4xgzMI2szHTGDU7n6C7tGzFSae5O6NWZCcd25/0VOyrb7n1/FVmZquaJSNNQkidt3va9h5i7ykvqPlq9i/2HSsPe97ijOpGVmc74wemc2rcr8bG6U6BUuW3CoKAkb+nmvXz4zU7OOrZ7FKMSkbZCSZ60OcWl5Xy2YTfZq3YyZ2Uu32zfH/a+ndvHM3ZQGuMHe4ldt07tGjFSaemG9OzMt47vzn+WB1fzzjymm6p5ItLolORJm7B5d6F3P9iVucxbs4uC4vDu9WoGJ/bs7CV1md04qVdn4lStk3q4bcLgoCRv2ZZ9vPf1Ds45vkcUoxKRtkBJnrRKh0rKWJSTXzkSds3OA2Hvm5KcwLhBaWRldmPsoDRSOyQ2YqTS2h17VCe+PaQH/162vbLtvvdXc/Zx3VXNE5FGpSRPWo0NeQWVSd2CtXkcLAmvWhdjcHLvLmRldmP84HRO6NmZmBj98ZXIuXXCoKAk7+tt+/jP8h2cO0TVPBFpPErypMU6WFzGwnV5zFmVS/bKneTkFYa9b3rHxMp+dWMHpdElKaERI5W27pgenfjuCUfx1lfbKtvue38V5xzXXf9QiEijUZInLYZzjrW5BZVJ3Sfr8ykuLQ9r39gYY2jfrowf7N0T9tgenfTHVZrUrRMG8faybZUTaH+zfT//Wb6db59wVHQDE5FWS0meNBsFRaVMn7uWpxdsIL+whJS573LZ8D4cd3Snyord5t0Hwz7eUZ3bVSZ1owem0aldfCNGL1K7wd078r0Tj+bNL7dWtt33/mq+dXwP/cMhIo0iKkmemWUAWc65mf7yOOA+4ETgcufcyzXsdy5wPxALPOqc+4vf3g94HkgBPgeuds4Vm1ki8BQwFMgDLnPO5TTaG5MjVlBUysRp89iQV0iRX53LLyzhoTlrwz5GfKwxPCPFn7euG4O7d1DHdmlWbj1rIP9aurWymrdyx37eXraN7514dHQDE5FWqcmTPDO7EbgNSDKzScDlwEZgEvDLWvaLBaYCZwObgUVm9oZz7mvgr8C9zrnnzexh4EfAQ/7X3c65gWZ2ub/dZY313uTITZ+7lvW7CigpC/NmsL5eXdtXJnWjB6SSnKjitDRfA7t15PyTjub1JVXVvPvfX823hxxFrKp5IhJhTfoX0cw6AncD5wHHAtlAgXNuu7++tg5WpwFrnHPr/G2fBy4wsxXAmcAP/O2eBO7CS/Iu8J8DvAz8n5mZc+HeVl6aytMLNoSV4CXExTCyf2rlZdj+acmq1kmLcstZg3jzy62U+9/uq3ce4K2vtnH+SarmiUhkNXXZoxxIADoB1PPSaU9gU8DyZmAEkArscc6VBrT3DN3HOVdqZnv97XcFHtjMbgBuAOjevTvZ2dn1CEsiIb+wpM5tpgxNJDMllsTYQijdwKblG4K+IdqSAwcO6Pu0Ds35HI04KpYFW6um+PnzG0vokL+SmCb8h6U5n5/mQueodjo/zV+TJnnOuQIzuwb4E9DDzIYAdzjnwpn7orrffq6W9tr2CY1rBjADYNiwYS4rKyuMcCSSUua+W2uil5KcwC2XTGjCiJq37Oxs9H1au+Z8jvocf4AJ/5hTWc3bVuDY33UwF5zcs/YdI6g5n5/mQueodjo/zV+T35/JOfcGcAnwNyAd+EWYu24Gegcs9wK24lXluphZXEh70D7++s5AfkPil8Zx1ai+JMZV/+2YGBfDVSP7NHFEIo2nf3oHLjwlOKG7//3VlJaFNyWQiEg4mjTJM7MOZtbXX9wPrAA61rHPN/7TRcAgM+tnZgl4Azbe8PvXzQYu9re7Fnjdf/6Gv4y//kP1x2ueJo8bQN/UpMMSvcS4GPqmJjF53IAoRSbSOG45c1DQYIt1uwp4I2B6FRGRhmrqSl48MB1vEMSdwLeB+81suJltxqvwTTez5QBmloZ/ydXvc/dT4D94yeGLzrnl/nF/BUwxszV4fe4e89sfA1L99inArxv/LcqRSE6M47WbxjB5fH9SkhMwvEu0k8f357WbxmjUrLQ6GWnJfD+kmvfAB6rmiUjkNHWfvN3AuaHz5AFb8C6zhhqJN21Kxf5vA29Xc9x1eKNvQ9sP4SWO0gIkJ8Yx5exMppydqb4e0ib87MxBvPbFFkr9znk5eYXMWrKVi4dW9+tQRKR+mrxPnm8PsKSujZxz/3LOPdAE8YiINLk+qUlcdGpwQvfgh6rmiUhkRCXJc87tcc7VmeSJiLR2Pz1zIHEBffM25BXy6hdbohiRiLQW0arkiYgI0DsliUuGHV7NK1E1T0QaqM4kz8wSzexKM/unma0ys33+Y7WZPWtmk8ysfVMEKyLSGt18xkDiY6uqeZvyD/Lq55ujGJGItAY1Jnlmlmxmd+PNOfcU3m3DBgId/McAvPvAPgZsNbPfm1mHxg9ZRKR16dU1iUuH9Q5qe+CDNRSXqponIkeutkreGuB2oCveNCZ5wIfAi8BL/vM8f11n4LfAqsYMVkSktbr5jIEkxFb9St6y5yAvf6ZqnogcudqSvO549329BzjOOdfNOTfBOXe5c+4y/3k34Dh/m83+PiIiUk9Hd2nPZcODq3lTZ6uaJyJHrrYk70fAQOfcXc65b2rayDn3jXPuLrzLtz+OcHwiIm3GTWcMOKya9+LiTVGMSERashqTPOfcE/5dJsLinCt1zj0RmbBERNqeozq354rTDq/mFZWWRSkiEWnJNIWKiEgzctMZA0kIuIfztr2HeHGRqnkiUn/hTKEy0My+Y2bHBbTdZWYbzWynmU01s4TGDVNEpG3o3qkdV47oE9T2f7PXcKhE1TwRqZ9wKnl/AN4ERgOY2Q+BO/DuNZsG/AT4VWMFKCLS1tw4fgCJAdW8HfuKeP7TjVGMSERaonCSvJH+17n+16sBB8wBZuFNoXJF5EMTEWmbunVqx1Uj+wa1Tcteq2qeiNRLbZMh9zGzAUDvqibrAwzzl+/Em0cPoLeZ9fb36RN6LBERqZ+fjB9Au/iqX9E79xfx7Ceq5olI+Gqr5OXgTW5s/mMFsB5I9pdnA1/52yb52+f424iISAOkd0zk6pBq3kNzVM0TkfDVluT1w5v77gDe5dmxwGS8BG8x0B840d+20F/u538VEZEGmjx+AO3jYyuXc/cX8fTCDVGMSERaktrmydvgnMvBu70ZwD+Bv+AlfB845zYAFf9SbvG33+C3i4hIA6V1SOSa0cHVvIfnrKWwOOwpTEWkDQtn4MUDeNW7DCAFKAIe9deN978uiHhkIiLC5HEDSEqoqubtOlCsap6IhKXOJM85NxO4CngVeAbIcs6t9Vf3xxtl+3hjBSgi0palJCdw7eiMoLbpc9apmicidQrrjhfOuWedcxc75652zn0a0P4r59wZzrmPGi9EEZG27Yax/UkOqOblFRTz1AJV80SkdrqtmYhIM9c1OYFJYzKC2mbMXUdBkap5IlKz2ubJu9PMuoR7IDPrYmZ3RiYsEREJdP3Y/nRIjKtczi8o5skFOVGLR0Sav9oqeXcCG8zsMTM7x8ySQzcws2R/3ePABrzbnYmISIR1SUrgumqqefsPlUQnIBFp9mpL8r4GOgKTgH8De8xstZktMLP5ZrYa2OOvu9bf9utGjldEpM360en96diuqpq3p7CEJ+fnRC8gEWnWakvyTsSb/HgN3hQqsXiTI58GjPCfx/rr1gE/AU5qzGBFRNqyzknxXDemX1DbIx+tZ5+qeSJSjdomQy53zj3inMsEzgT+CnyId3uzb/Bua/Y34Czn3CDn3AznXHlTBC0i0lZdd3q/oGre3oMlzJyXE72ARKTZiqt7E3DOZQPZjRqJiIjUqXP7eH58en/ufX9VZdujH63j2tEZdG4fH8XIRKS50RQqIiItzA9Pz6BTQDVv36FSnpi3PooRiUhzpCRPRKSF6dQunuvH9g9qe+yj9ewtVN88EamiJE9EpAWaNCaDLklVl2f3F5Xy2MfrohiRiDQ3SvJERFqgjtVU8x6fl8OewuIoRSQizY2SPBGRFura0Rl0DajmHSgq5dGP1DevIQqKSvnHeys59Z53mfROAafe8y7/eG+lbiEnLVKzTvLMLMPMJgUs9zGz2Wb2hZktNbPvBKz7jZmtMbOVZvatgPZz/bY1ZvbrJn4LIiKNpkNiHDeMGxDU9sS89ewuUDXvSBQUlTJx2jymz1lHvt+/Mb+whOlz1jFx2jwletLi1CvJM7M4MzvNzL7dWAEFvNaNwH+A35tZtpn1AG4HXnTOnQJcDkzztz3OXz4eOBeYZmaxZhYLTAW+DRwHXOFvKyLSKlwzqi8pyQmVywXFZTzykfrm1de+QyX810tfsmbnAYpKg6d8LSotZ0NeIdPnro1SdCJHJuwkz8wuAbYAC4A3/bYPzGydmZ0TyaDMrCNwN3AN8Du8W6sVAA7o5G/WGdjqP78AeN45V+ScW493l47T/Mca59w651wx8Ly/rYhIq5CcGMfkccF982bOzyFf1bxaOedYtmUvU2ev4dKHF3DKPe/x9rLtlLvqty8qLefphRubNkiRBgprMmQzGws8h3cLMwtY9Rbwv8DFwLsRjKscSMBP6JxzOX4cdwHvmtnPgGRggr99T2BhwP6b/TaATSHtI0JfzMxuAG4A6N69O9nZ2ZF5F3LEDhw4oM+hFjo/dWtL5yij1NEpAfb5eV1hcRm3P53NpZkJNe7Tls5PhQPFjuV5ZSzNLWNZXhl7i2rI6Gqwu6C4zZ2z2rTF76GWJqwkD/gNXtXvG+CYgPb3/a+jIhmUc67AzK4B/gT0MLMhwB3AFcBM59zfzWwU8E9/nVV3GKqvVB72U+2cmwHMABg2bJjLysqKzBuRI5adnY0+h5rp/NStrZ2jTYnr+MNbKyqXZ28u554rR5HWIbHa7dvC+Skvdyzdspc5K3OZs2onSzbtqbFSF47YGOOk4aPpmlxz8tyWtIXvoZYu3CRvJF5ydB6wOqC9ouNHz8P2aCDn3BtmttR/zWHAL/D63Z3rr19gZu2ANLwKXe+A3XtRdSm3pnYRkVbjyhF9eXjOOnYdKALgYEkZM+au47ffOTbKkTWtXQeK+Gh1Ltkrc/lo9a56XbbumhTPvkOllNWQCZaWOyZOm8djk4YzIL1DpEIWaTThJnnJ/tfQDgntQ75GhJl1AFL9xf3ACiDFf/2zgJlmdizQDsgF3gCeNbN/AEcDg4BP8Sp8g8ysH15/wsuBH0QyVhGR5qB9Qiw3Zg3g9//6urLtqQU5XD+2P+kdq6/mtQalZeV8uXkP2StzmbMql6+27MWFWa1LTohl9MA0sjLTGTconZTkBCZOm8eGvMLDBl9UyMkr5PvT5vPQVacyekBaBN+JSOSFm+RtAfpy+GXZX/pfN0csIk88MB2vSpeKl9z9AG+wxSNm9nO8yuIk55wDlpvZi8DXQClws3OuDMDMfoo3SjcWeNw5tzzCsYqINAtXjujD9Dlr2bnfq+YdKiln+py13P691jWpwM59h8he5SV1H6/exd6D4d/O7ZgeHRmfmc74wekM65tCQlxwr57XbhrD9LlreXrhRnYXFNMlKZ72CbFs3XOocpu9B0u45rFP+dPEE7h0eO/QlxBpNsJN8v4DTAZmVTSY2Td4FTPnr48Y59xu4FwzywCynHMz/VVbgDE17PNH4I/VtL8NvB3J+EREmqN28V417+43q6p5/1y4gRvG9adbp3ZRjKxhSsrK+WzDbuas8i7Drti2L+x9O7aLY+ygNMYPTmf84G706Fz7eUhOjGPK2ZlMOTuzss9ZcWk5v33tK17+rKqeUVru+O9XlrJ21wF+9a1jiImprmu4SHSFm+T9AW8EbSpVAxcG4V0OzQP+HPnQANgDLGmkY4uItDpXnNaHh+esZcc+r5pXVFrOQ3PWcud5x0c5svrZuuegn9TtZN6aPA7UYyLiIT07MX5wOlmZ3TildxfiYhs2739CXAz/c/GJ9E9P5m/vrAxaN33OOjbsKuTey06mfUJsg15HJNLCSvKcc1vMbAzwAF6fuFigDPgAuM05t6UxgnPOKckTEamHdvGx3HzGQO54vapnyjOfbOQn4wfQvRlX84pKy1ics5vslTuZsyqXVTsOhL1vl6R4xg5KJ2twOmMHp9GtY+Tfp5lxU9ZAMlKT+fkLS4L67L2zfDtbZyzg0WuGteiKqbQ+4VbycM6twruE2g5vEES+c+5QHbuJiEgTu2x4bx7KXsu2vd6v6OLSch7KXstd5zevat6m/MLKpG7+2jwKi8vC2s8MTuzVhazB6YzPTOekXl2IbaLLpd854SiO7tKeHz+5uHIkM8DSzXu5YOo8Hrt2OMcd3amWI4g0nXAnQ+6MN+ih0Dm3C38aEjNLA5KAvc65vY0WpYiIhC0xLpabzhjI72Ytq2x79lOvmldXn7TGdKikjIXr8pizKpc5K3NZt6sg7H1TkxO8fnWZ6Yz1R8JGy8m9u/D6T8fwo5mL+Gb7/sr2bXsPcfHD83nwilM469juUYtPpEK4lbzHgQuBn+Ndsq1wOXA/8Bpenz0REWkGLh3Wi4dmr2FrQDVvWvYa7rlgSJPF4Jxj/a6CygETC9fl1Tg1SagYg1P7dK3sW3f80Z2a1eCGnl3a89JPRnHLc18we2VuZXthcRnXP7WY2797HD8ck4FZ84lZ2p5wk7yKW4G9EtL+Kl7Sd9itwkREJHoS42K5+cyB/L/Xqqp5z3+6iZ+MH8DRXSI6tWmQwuJSFqzNq5y3bmN+Ydj7duuYSFamNwr29IFpdE6Kb7Q4I6Fju3geuWYYf3hrBTPn51S2lzu4519fs27XAe467/gGD/wQOVLhJnnp/tc9Ie17Q9aLiEgzccnQ3kybvZYtew4CUFxWztTZa/jjxBMi9hrOOdbsPFCZ1H26Pp/isvCqdXExxrCMrowf3I2szHSO6dGxxVW+4mJjuOv84+mXlszdby4Pum3a0ws3siGvkKlXnkqnds07YZXWKdwkbz/QFTgH79JshXP8r+EPgxIRkSaREBfDz84cyK9f/aqy7cXFm7gxa0CDjrv/UAnz1nh96+auyq1MIsNxdOd2jM/0krrRA1Lp2EqSn2tHZ9AnNYmfPftF0HQvH63exUXT5vP4pOH0TkmKYoTSFoWb5H0OTAAeN7Pj8W4zdiwwBW/evM8aJzwREWmIi4b2Ymr2Gjble4lYSZlj6uy1fCsl/GM451ixbX/lvHWfbdhNaQ33dw2VEBvDaf1S/Muw6Qzs1qHFVevCdUZmN165cTTXzVwUlPiu3nmAC6fOY8Y1wxjat2sUI5S2Jtwk72G8JK8TcHdAu+EleQ9FOC4REYmA+NgYfnbGIP77laWVbS8t3sSpp9c+ynZvYQkfrfFGwc5ZlVt5q7Rw9ElJqkzqRg1IJSkh7Nm6WrzMHh157ebRXP/UZ3y5qaqHU15BMVc8spD/veQkzj/p6ChGKG1JuJMhv2pm/8Cr3IX6u3NuVjXtIiLSDEw8tSf/N3tN5SCI0nLHf809yJ8Xv8tVo/oyedwA2sfHsmzrXuaszCV7VS5fbNxNmMU6EuNiGDUgtXIkbEZqUqut1oWjW8d2vHDDSH7x0pe8tXRbZXtxaTm3PPcF63MLuOWsgW36HEnTqM9kyL80sxeA84HuwA7gDefcosYKTkREGi4+NoYbxvXj9lnLg9rzC0uYNnstT8zLIT7GyC8sCfuY/dOTK5O6Ef1SaBevW3oFahcfy4OXn0L/tGQe/HBN0Lp7319FTl4Bf7noBBLjdN6k8dSrhu4ndErqRERamO17D1X2rwlUWu7Yf6ju+8ImJcQyekAq4zO7MX5QOn1SNYigLjExxi/OySQjNZlfv7qUkrKqs//aF1vYvLuQ6VcPi+rEztK6hZ3kmVlH4DtAX+CwzhzOuXsiGJeIiETQs59sPCzBq8vg7h3IyuzG+MHpDMvoqqrTEbpoaC96dW3P5Kc/Y09AtXRRzm4unDqPxycNZ2C3DlGMUFqrcG9rNhx4G++etTVRkici0kztDuNSbIfEOE4fmMZ4f9BEY06a3NaM6J/KrJvGcN3MRUG3c9uYX8jEafN4+KqhjBmYFsUIpTUKdxru+4BUvNG01T1ERKQZ61rH3SM6tYvjizvO5uGrh3LFaX2U4DWCjLRkXr1pNKP6pwa17z9UyrWPf8pzn26MUmTSWoWb5J2I15UjG/gZcB3ww4DHdY0RnIiIRMZVo/qSGFf9r/zEuBgmjckgXrffanRdkhJ48rrTuGxY76D20nLHb179ij+9vYKycIc1i9Qh3D55e4Ak4PvOudBbm4mISDM3edwA3lm2nQ15hRSVVt12LDEuhr6pSUwe17C7YEj4EuJi+MtFJ9AvPZm//PuboHUz5q5j/a4C7r/85DY1v6A0jnD/bXvK/zqksQIREZHGk5wYx2s3jWHy+P6kJCdgQEpyApPH9+e1m8aQnKiEoimZGT8ZP4CHrzqVdvHBf4rf+3oHl05fwPa9h6IUnbQW4f5U5wB7gdfN7DFgJRDUi9c591Q1+4mISDORnBjHlLMzmXJ2JtnZ2WRlZUU7pDbv3CFH8WKX9vz4ycVBdxVZtmUfF0z9mMeuHc6Qnp2jGKG0ZOFW8qYDnYEuwC+AGcATAY/HGyU6ERGRVu7EXl2YdfMYjj2qU1D7jn1FXPLwAt77ekeUIpOWrj69bGsaWasRtiIiIg1wdJf2vPSTUZx1TLeg9oMlZdzwz8U8MncdzmlAhtRPuJdrf9ioUYiIiLRxHRLjmHHNMP741goen7e+st05+OPbK1i3q4B7Ljheo6AlbGElec65Jxs7EBERkbYuNsa447zj6J+ezJ1vLA+aTuW5TzeyKb+QqVeeSuf2tc97KAL1u1wrIiIiTeCqkX15YtJwOoaMev54zS6+P20eG/MKoxSZtCRhJ3lmdrWZfW5mBWZWFvKo++7WIiIiErZxg9N55abR9OoafPeRtbkFXDhtHotz8qMUmbQUYSV5ZnYp8CRwEtAeDbwQERFpdIO7d2TWzWM4tU+XoPb8gmJ+8MgnzPpiS5Qik5Yg3Erezf7Xg/5XB+T5z/cAGyIZlIiIiHjSOiTy7PUjOe+ko4Pai8vKue2FJdz73iqNvJVq1ffetRMqGpxz6cCdeJMinxf50ERERASgXXwsD1x+MrecNeiwdfd/sJpbn1/CoZKyKEQmzVm4SV6y//VzvGQPM4sF/g6kAw9EPjQRERGpYGZMOXsw9152Egkh06i88eVWrnz0E3YdKKphb2mLwk3y9vlfDdjvP/82MNR/PiKSQYmIiEj1Jp7Si2euH0FKckJQ+2cbdnPh1Hms3rG/hj2lrQk3ydvqf+0GrPCfvw5k+881xEdERKSJDM9I4bWbRjMgPTmoffPug3x/2nw+Wp0bpcikOQk3yfsCr4o3AniKw0fVNspkyWaWYWaTQtouNbOvzWy5mT0b0H6tma32H9cGtA81s6/MbI2ZPWBmGgksIiItXt/UZF69cQxjBqYGte8vKmXSE4t4eqHGRLZ14SZ5NwFHAW875x4Cfgt8CXwG3A7cFenAzOxG4D/A780s28x6mNkg4DfAGOfc8cBt/rYpeINARgCnAXeaWVf/UA8BNwCD/Me5kY5VREQkGjonxTPzh6dxxWm9g9rLyh23z1rG7//1ddBdM6RtCfe2ZgVAQcDyX4C/NFZQZtYRuBtv1O6xeJeFC4ApwFTn3G4/jp3+Lt8C3nPO5fv7vweca2bZQCfn3AK//SngQuDfjRW7iIhIU4qPjeFPE0+gf1oH/vTvFQTOpvLYx+vZkFfA/ZefQnJiuLerl9bCappbx8z61OdAzrmNEYnIe+1kYAtwCdDTOTfTb58FrALGALHAXc65d8zsl0A759wf/O1+hzenXzbwF+fcBL99LPAr59z3Ql7vBrxqH927dx/6/PPPR+qtyBE6cOAAHTp0iHYYzZbOT910jmqn81O3lniOPt9RysNLiygOmU2lT8cYbhuaSEq7yN3NtCWen9bojDPO+Mw5N6y6dbWl9Tn406WEwdVxrHpxzhWY2TXAn4AeZjYEuMN/jUFAFtAL+MhfV10/O1dLe+jrzQBmAAwbNsxlZWVF4F1IQ2RnZ6PPoWY6P3XTOaqdzk/dWuI5ygLOOX0vP3pyETv2VU2nsnF/OX/5rJzHrh3KCb06R+S1WuL5aWvqSumru31ZTY+Ics69gVfJ+xveXHy/ADYDrzvnSpxz64GVeEnfZiCwQ0IvvBHBm/3noe0iIiKt0pCenXn95tM5/uhOQe079xdx6fQFvLNse5Qik6ZWW/WtUUbMhsPMOgAVw4X2403bkgLMAq4AZppZGjAYWAesBf4UMNjiHOA3zrl8M9tvZiOBT4BrgAeb7p2IiIg0vR6d2/Hi5FHc+vwS3l+xo7L9YEkZNz7zGb8+9xhuGNcfTTjRutWY5DnnftiUgYSIB6YDaXjJ3kbgB3hVuHPM7GugDPgvi3hmrgAAIABJREFU51wegJn9Hljk739PxSAM4EZgJtAeb8CFBl2IiEirl5wYx/Srh/KXf6/gkY/WV7Y7B3/+9zesyy3g9xcOISEucv30pHlplkNt/NGz55pZBpBVMfDCN8V/hO7zOPB4Ne2LgSGNEqiIiEgzFhtj/L/vHke/tA787vVlQdOpvLB4ExvzC3n4qqF0ToqPYpTSWMJO8swsE5gMZOJVxQI559xZkQzMtwdY0gjHFRERaTN+MKIPfVKSuPGZz9h/qLSyfcG6PCY+NI/Hrx1ORlpyLUeQliisGq2ZDQUWA7fiTSY8PuCR5T8izjm3xzmnJE9ERKSBTh+Uxms3jaZ3SnCdZl1uAROnzePT9bpDaWsT7oX43wLJNMGoWhEREWkcA7t1ZNZNYxjWt2tQ++7CEq58dCGvfr45SpFJYwg3yRuNN7/cjf6yA04E3sCbnPjUyIcmIiIikZbaIZGnfzyCC04+Oqi9pMwx5cUv+fu7KynXrdBahXCTvIrpTJ6paHDOLcO7S8Rg4OcRjktEREQaSbv4WO677GR+PmHwYese/HANP3v+Cw6VlFWzp7Qk4SZ5B/2vhyqe+wMxKobjnB/huERERKQRmRm3ThjE/ZeffNg0Km8t3cblMxaSu7+ohr2lJQg3ydvpf03Bu90ZwGxggf+8PIIxiYiISBO54OSePHf9CFKTE4Lal2zaw4VT57Fy+/4oRSYNFW6S95X/9UTgX3gDLroDPfH6570b+dBERESkKQztm8Ksm8cwqFuHoPYtew5y0UPzyV65s4Y9pTkLN8m7G++OEznAH/CSuoqRtR/gTa0iIiIiLVTvlCReuWk0YwelBbUfKCrlupmL+OeCnKjEJUcurCTPOfelc+4F59wa59x+59y5eJduOzvnznHO5TZumCIiItLYOrWL5/FJw/nBiD5B7eUOfvf6cu56Y3nQXTOkeWvIbc0SgIJIBSIiIiLRFx8bwx8vHEL/tGT++PYKXEBON3N+Dtkrd7L3YAm7C0tImfsuV/1/9u48zJKyPvv495Z9EJTFDArIGCVuiFuLGFzGnRgVTdSoESFRIYrGxCVxeWURNS5Eo++LRjQGMETiEgUVxYmxo0YhzAiiiAsiwghK2JRFWX/vH1XtFIfeZvp0nzPF93Nd5zpVz1PLU3VOd9/91PaI3Tj40fdk6y3G8kmpt2uz9uQleUiSdyZ5X5LHtWUvTvK/wCXAVUmOWoqGSpKkpZGEFz/qd/nQ/hMs23yTW9VdcPl1XHndjQBccd2NfPC/zueZ7/9vrr3+pukWpRGaMeQleSTN1bOvBg4BViV5O3AMzaHa0DzD9q+T/MUStFWSJC2hJ9xvOZ/4i0ew07ZbzjjN9Tfdwk8vv44PfvXHS9gyzcdsPXmvpbkPXvcRZq9t6wJc1hnef7EaKEmSRuf+d7sTJ718Hza5w8xPMr3+plv4l9MuXMJWaT5mC3kTNLdHORV4GfAFmkBXwPOq6neAP22nvd9iNlKSJI3O8m23nPNRZ1ded8MStUbzNVvIm7qG+k+q6h9pbqEy5d/b90+179sMu2GSJGl8bLdssznqN5+1XktvtpC3GUBV/ap9/+VURVXd2L5PxfaZ+3AlSdJG7wWP2I0tNp0+Nmyx6R14wd53n7ZOozPn9c5JDp1PmSRJ6q+DH31Pvvjdn/PTy6/j+pvWPc10i03vwG47LOPgR99zhK3TdOZzU5vDOsM1TZkkSeq5rbfYlE+/bB8++NUf8y+nXciV197Adltvzgv2vrv3yRtTc30iHoaVJElAE/Re9cR786on3pvJyUlWrlw56iZpFrOFvCOWrBWSJEkaqhlDXlUZ8iRJkjZSsz7WTJIkSRsnQ54kSVIPGfIkSZJ6yJAnSZLUQ4Y8SZKkHjLkSZIk9ZAhT5IkqYcMeZIkST1kyJMkSeohQ54kSVIPjXXIS7IiyYHTlD8rSSWZ6JS9Psl5SX6Q5Mmd8n3bsvOSvG6Jmi5JkjRSYxvykrwUOBU4Mslkkp3a8m2AvwRO70x7P+C5wP2BfYH3J9kkySbA0cAfAPcDntdOK0mS1GubjroB02mD3BHA04D7ApPAtW31kcA7gdd0ZtkPOLGqrgd+kuQ8YK+27ryqOr9d7onttN9b7G2QJEkapbEMecAtwObAtgBVdQFAkgcDu1bV55J0Q97OwGmd8bVtGcBFA+UPH1xZkoOAgwCWL1/O5OTkUDZCG+6aa67xc5iF+2du7qPZuX/m5j6anftn/I1lyKuqa5O8EHgbsFOSPYBDgfcAB04zS6ZbDNMfjq5p1ncMcAzAxMRErVy5csMarqGZnJzEz2Fm7p+5uY9m5/6Zm/todu6f8TeWIQ+gqk5OcjbNIdsJ4LXAHsBkEoCdgJOTPJ2mh27Xzuy7ABe3wzOVS5Ik9dZYhrwkdwR2aEevBs4Ftq+qHTvTTAKvqarVSX4N/GuSdwN3A3YH/oemh2/3JPcAfkZzccbzl2xDJEmSRmQsQx6wGfBBYEeasHchs4SzqjonycdpLqi4CTikqm4GSPJymqt0NwE+UlXnLHLbJUmSRm4sQ15VXQnsm2QFsLKqjp1mmpUD428F3jrNdKcApyxGOyVJksbV2N4nr3UVcNaoGyFJkrSxGcuevClVZciTJEnaAOPekydJkqQNYMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4a65CXZEWSAzvjr0ryvSRnJ/lykt06dQck+VH7OqBT/tAk30lyXpL3JckSb4YkSdKSG9uQl+SlwKnAkUkmk+wEnAlMVNWewCeBd7bTbg8cBjwc2As4LMl27aI+ABwE7N6+9l3SDZEkSRqBsQx5SbYBjgBeCLwJOBC4tqq+UlXXtZOdBuzSDj8ZWFVVV1TVlcAqYN8kdwW2rapvVlUBxwPPWMJNkSRJGolNR92AGdwCbA5sC1BVF0wzzYuAL7TDOwMXderWtmU7t8OD5beS5CCa3j6WL1/O5OTkghqvhbvmmmv8HGbh/pmb+2h27p+5uY9m5/4Zf2MZ8qrq2iQvBN4G7JRkD+DQqV68JC8AJoDHtLNMd55dzVI+uL5jgGMAJiYmauXKlQveBi3M5OQkfg4zc//MzX00O/fP3NxHs3P/jL+xPFwLUFUnA8+mOe/uLsCrAZI8AXgj8PSqur6dfC2wa2f2XYCL2/JdpimXJEnqtbEMeUnu2Lly9mrgXGCbJA8GPkgT8C7tzHIq8KQk27UXXDwJOLWqLgGuTrJ3e1XtC4GTlm5LJEmSRmMsD9cCm9GEuR2BHYALgecDxwF3BD7R3gnlwqp6elVdkeRI4Ix2/jdX1RXt8EuBY4GtaM7hmzqPT5IkqbfGMuS1V8jum2QFsLKqjm2rnjDLPB8BPjJN+Wpgj+G3UpIkaXyN5eHajquAs0bdCEmSpI3NWPbkTakqQ54kSdIGGPeePEmSJG0AQ54kSVIPGfIkSZJ6yJAnSZLUQ4Y8SZKkHjLkSZIk9ZAhT5IkqYcMeZIkST1kyJMkSeohQ54kSVIPGfIkSZJ6yJAnSZLUQ4Y8SZKkHjLkSZIk9ZAhT5IkqYcMeZIkST1kyJMkSeohQ54kSVIPGfIkSZJ6yJAnSZLUQ4Y8SZKkHjLkSZIk9ZAhT5IkqYcMeZIkST1kyJMkSeohQ54kSVIPGfIkSZJ6yJAnSZLUQ4Y8SZKkHjLkSZIk9VCvQl6SFUkO7IxvkeTfkpyX5PQkK0bWOEmSpCXUm5CX5KXAqcCRSSaT7AS8CLiyqu4FvAd4xyjbKEmStFQ2HXUDhiHJNsARwNOA+wKTwLXAfsDh7WSfBP5fklRVjaCZkiRJS6YXIQ+4Bdgc2Bagqi4ASLIzcFFbdlOSXwI7AJd1Z05yEHBQO3pNkh8sTbM1ix0Z+Jx0K+6fubmPZuf+mZv7aHbun/Gw20wVvQh5VXVtkhcCbwN2SrIHcCiQ6SafZv5jgGMWt5VaH0lWV9XEqNsxrtw/c3Mfzc79Mzf30ezcP+OvFyEPoKpOTnI2zSHbCeDVwFpgV2Btkk2BOwFXjK6VkiRJS6MXF14kuWOSqe7Kq4FzgW2Ak4ED2vJnAf/p+XiSJOn2oC89eZsBH6Q5P2AH4ELg+cDlwEeTnEfTg/fckbVQ68vD57Nz/8zNfTQ798/c3Eezc/+MufSpY6u9D97Kqjp2tC2RJEkarV4cru24Cjhr1I2QJEkatV715EmSJKnRt548SZIkYciTJEnqJUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+6nUgymaTa1+R6zHd4Z75axCZKkobIkKfbvSQruiGmff08yRYzTP/taaZfsbStvk2bum05fJRtWYjBQJlk5ajbtNj68tnNZUP/yejMv3Kan7tKckuSq5N8L8kxSR4wxDYP/m44cFjLlpaCIU+a3nLgTwcLkzwB2HPpmyNpBgHuCNwXeAmwOsmTRtskaTwY8qSZ/fU0Za9e8lZIms4q4LXA4cB/d8o3B/5uFA0apjTuOOp2aONmyJNu6+b2fY8kT5wqTHI/4MkD08woye8nOSHJT5L8Jsm17SGl9yX53WmmP7ZzWOiCJNsmeUc7//VJLkzy9iSbd+aZnOY8ucPmcyg5yQ5J/m+Ste3yf5TktUkyj23bNMlFnXW8Z5ppDuzU35xkl7mWO8c6Bw/XrUzynCTfbPftL5N8Nsl9OvPcsz2cNzXP86ZZ7vsGDtNvOrCPDktyRrv8G9r99a9JHjZDO5+b5EtJfpHkxvZQ4gVJvpjkyCQ7tdPN+7MbPNSZ5F5JPpnkyiRXJfl0knu20z4wyeeT/Kp9fT7JfWdo63pt38BnWkl+N8lfJDkrya+TXNbOe7fOPIe32/mYzqIek4UfBv1GVR1VVUcAjwXWduruN8P27prkXUnObj+X65Ocn+RDSe49MO0FwE8GFvHP3XZ3pp3xkHtmOad1ms91tyT/kuQXNL9jnjXVls50x7af/wlJLm234btJDlivvafbh6ry5et2/QJWANV5faozfEpnug93yj85MM+KgWW+GbhlYJru61rgGQPzHNupvww4Z4Z5/7kzz+Qs67hV2wam/T5w/gzTHzbQrsO79Z3y13fKLwe2HJjvi9Ptxzk+i8MH2rKyU7dyoO6rM7T/UmDHznxf6dR9fmB9mwA/79S/o1M3MVA3+LoJePnA8t4wj89j5QI/u/Pb/T047S+ApwPXzbVPFrB9B87zMzgX2GKGz3S614Hz+G4Mfv6HD9Sv6dRdNM38TwF+NUsbfg38UWf6C+Zqd2fa2dp1+HTzTPO5/nCaz+PAadrybeCXM7TpgFH/PvU1Xq/f/scq6be+A2wDPBHYt+0FuRx4QVv/A+DzwB9PN3OSPwHe1Cm6APg3YBnwZzTnDy0DPpZkj6r68TSL2QHYDjgeuBh4MbBjW/fCJG+oqkuADwCfA97VmXcV8KXO+BXTLP/ewG/a+X8NvBTYqq17VZK3VdWN021fxzHtdm4FbA88G/goQJIdgcd3pv3wHMvaEI8CzgBOpenJ2actvwvwIuAdnXWvbIeflOQuVfW/7fgTaM6/vFU7k2wDfLZT9wvgYzT78gnAo2kC4nuTnFVVX2+n+8vOslbTfDYAu9Kcy/nQTv2Gfnb3oPk+vhP4XdreHuB3gJOA/wX+L/B7wDPaulvtkwVs36BHAV8GvtGua+qih/u04//Wbs81NN+xqR7s89vtn3LGDMufU5KtaQLcAzvF/zowzW7AJ2h+7qDpofs4zc/AfsCDgC2BE5Lcv6rOB95K8w/gGzqL+jeaz3Ux7N6+fwY4E9iZ5nMetCdwJfAemp+9l9B8VgCvA45bpPZpYzTqlOnL16hf3LYn73Caw7JT48fQ9MxNjf8Ft+3RWNFZXrdH4Spu3av02IH53tupO3ag7pWduv0G6p42sA0z9iR0ppkcmG6/Tt0rB+oe0Kk7vFs3sMxjOnVf75Qf3Cn/BbDZPD+LwwfasbJTt3Kg7vSp5QKbteuZqvtUZ74tacLLVN0rOnXHd8onO+Uv75T/Bti1UxfgtE79Zzp1V3XK955m+3YE7jSEz26fTt3PBur2bsvvQPMPwnT7ZEO378CBdf07kLZue5rev6m6v59lGyan2845vhuDn/90r5tovpObD8x7VGeaS4BtO3VbABd26v9hlt8NB87Qthk/Q+bfk1fAX82w/As609wCPLhT956BZWwzjN+Lvvrx8pw8aRpVdSrN4VKA/YGXtcOXM8t/ykmWAQ/uFH22qi7rLPcrNL+wp+zD9G4GPtgZ/8FA/XYztWGeLq6qk4aw/Pd2hvdJcv92+Lmd8uNq7l7BDfHhqeW27z/p1P22/VX1G+CETt0L4Lef1TO7y+sMP6ozvAVwYeecqluAh3fqH9kZ/mpneFWS/0jyj0leleT3gSuq6pfz3sLp/bSquhca/LQz/JOqOg2gqm6h6TGb0v1MN3T7Bn2gqqpd3xU0pxlMt76l8nXgzVV1w0B5d3t3An7Z2d7f0PS0TpltexfTlcD/m8d036yqMzvjw/7doB4x5Ekze3f7viXN4VOAf6yqX88yz3Y0PSFTfj7NNN2y7WdYzi/acDLl+oH6hf7sXjAwvkHLr6pzaA7XTTk4yV1pDvdN+af1bt38XDAw3t2GwfZ/qDO8V5Ldac5fm7p68Uqa8yynzPS5TGf7JFPrOxj4r3b4jjSHrA8G/p7mCtAfDJ7gvwF+NjDeDTQXD9Td1Bnu7pMN3b5BFwyMz/YZDNsqmsOT/0ITTKG5uONrSQa3b3229y4LbNfghUvT3m9zGj+uqpvmnmw4P7u6ffCcPGlmJwBvY915Szcw93/aV9IcMpn6Rb/TNNN0y6Y75wpgsOer5ljv+hrm8v+Bdeff7U8TNKb+0Hy9qgZ7GoZl3ttQVWcnWU1zsQE0vXkP6UxywkCo7n4uvwKOnKMtU71ZlwArk9wd2IvmPKt70xxuvzNwL5pz0R43x/JmM1uv6HxCAmzg9s2jLcP+ns7mG1U1dY7hGprDltAcYn0L63rf4dbb+1Nm/zm+egPa0v2Z32qgbnfm59p5TjfKfa6NjCFPmkFVXZ/kaJrz8QA+VlXT9cx157kuyVmsO2T7tCQ7Th2yTfMEhxWdWf6b4biJdT/Py2abcBF8HjiPJsDcGTisU7cYF1xsqA+zLuT9OdNccNHxdeA57fC2wJr2UPutJNkDuPPUIcskDwS+W1UX0pznNTXdq2h68+i0YcooPrsN2r4F6oaTYW/n+2guQJi6dcpLkryrqqYO4X+dJnRD87l/vqrOHVxIkr1pDt9OGQxUM7X7KtYdJt0rSaqq0jx942nrtynS8BjypNm9j+ZqW4D/mec872Ld1X13As5IciLNH4g/70x3Pc1VkMOwlnXh8cAk19PcZuGyqjp2SOuYVvvH7P+y7vy8Ldv3X9Fc0Tgu/pUmaG0NdO/Zd0ZVfXtg2uOAN7IuCH4hyaeB79H02KygOZ/y94AjaEIENL2/y5P8J81nchnN4b/9O8u+cmBdo/jsNnT7FqJ7H7uHJnkf64Lw0XOcBjGrqrolydtoDt1C87ftjTRXpUPzczZ1BfmWwOlJPgn8mOainXvRnGKwK80V8Ge1811K04M/dW/K17RXjl9Hc3j102356cC+7fBjgNOSXExzpfJv72spLTVDnjSL9iT5z6znPB9r/4N/fVu0gubcoa5fAy+oqvMW3MjGJ2ju/g9NqPg/7fA5NFftLrZ/pjnkt22n7F+r6rolWPe8VNXVST5Bc4Vo1216G6vqV0meBpxMc3h9C259MclsdmRdL9l0Bp/GsOSf3QK3b0N19/0dgFd06o6l+ZlYiBNprmS9Vzt+QJK3VtVPquqCJM+muU3MNu3rz+ZaYFXdmOQkmtsDQXP7miPa4c8DUyHvXcCTWHeawlSv4XU092l87IZulLQQnqApLYKqegNNz8DHaHorbqD5I/YD4Ghgz6r69yGu8k009027gPmflzU0VXU18JGB4nE6VDvlQwPj19J8RrdRVWcA96cJXafRHJK7iebCmW+1y3o66+7HB03v0dE0vb4/o+mtvQG4iOZ2I0+uqn8cWNVIPrsN3L6FrO8Ump61b3PbiwWGsfybuXWAnurNm6r/PM3h3LfTbN/VNIdjf0bzeb2XpueteyU2wEE0t2W5mBmedFNV/0mzr86g2baraALgw7j1FdfSkpq6v5EkLUiSl9D8MQQ4q6oePNv0kqTF5eFaSRsszXNY70Nzjlv3gov3Tj+HJGmpGPIkLcS+NOfjdZ1G+3gzSdLoeE6epGEomnObjgb+sD0/SpI0Qp6TJ0mS1EMerh2w44471ooVK0bdDEmSpDmtWbPmsqqa9nF8Yx/ykmxJcwn6FjTt/WRVHZbkHjT3Rdqe5nL4/avqhiSvoHlW5IXAM9qyRwJ/VFWvmmt9K1asYPXq1Yu1OZIkSUOT5Kcz1W0M5+RdDzyuqh4IPAjYt330zDuA91TV7jR3kH9RO/2LgT2BM4EnJwnNfajmejajJElSb4x9yKvGNe3oZu2raB7w/cm2/DjgGZ3ZNqN5hNSNNI8TOqWqBh8lJEmS1Ftjf7gWIMkmwBqax9UcTfO8wauqauru8GuBndvho2hu4XAOzcPfP8O6ZwrOtPyDaO5qzvLly5mcnBzyFkiSJC2tjerq2iR3pnlUzKHAP1fVvdryXWl66x4wMP1hNA+aLuCFNI8WenVV3TLTOiYmJspz8iRJ0sYgyZqqmpiubuwP13ZV1VXAJLA3cOckUz2Ru9A8V/C3ktwNeFhVnUTzbMY/oTm/7/FL1mBJkqQRGfuQl+QubQ8eSbaieYD0ucBXgGe1kx0AnDQw65E0F1wAbEXTm3cLzbl6kiRJvTb2IQ+4K/CVJGcDZwCrqupzwN8Cr0pyHrAD8E9TMyR5MEBVndkW/RPwHeAhwBeXsO2SJEkjsVGdk7cUPCdPkiRtLHpzTp4kSZLmx5AnSZLUQ4Y8SZKkHjLkSZIk9dBG8cQLaZzliIy6CUuiDvMiLUnamNiTJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9NPYhL8muSb6S5Nwk5yR5ZVu+fZJVSX7Uvm/Xlv9xO93XkuzQlt0zyYmj3A5JkqSlNPYhD7gJeHVV3RfYGzgkyf2A1wFfrqrdgS+34wCvbqc7Hnh+W/YW4E1L2mpJkqQRGvuQV1WXVNW32uGrgXOBnYH9gOPayY4DntEO3wJsASwDbkzyKOCSqvrRkjZckiRphDYddQPWR5IVwIOB04HlVXUJNEEwye+0kx0BnApcDLwA+Djw3DmWexBwEMDy5cuZnJxchNarr476vaNG3YQl4c+FJG1cUlWjbsO8JLkj8F/AW6vq35NcVVV37tRfWVXbDcxzAHBnmlD4GuBK4JVVdd1M65mYmKjVq1cvyjaon3JERt2EJVGHbRy/KyTp9iTJmqqamK5u7A/XAiTZDPgUcEJV/Xtb/Iskd23r7wpcOjDPMuAA4P3A3wF/DqwB/nSp2i1JkjQqYx/ykgT4J+Dcqnp3p+pkmhBH+37SwKx/A7y3qm4EtgKK5ny9ZYvbYkmSpNHbGM7J2wfYH/hOkrPasjcAbwc+nuRFwIXAs6dmSHI3YKKqDm+L/h44DbiKdRdoSJIk9dbYh7yq+jow00lPj59hnouBp3bGPwF8YvitkyRJGk9jf7hWkiRJ68+QJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9tOmoGyBJt3c5IqNuwpKow2rUTZBuV+zJkyRJ6iFDniRJUg8Z8iRJknrIkCdJktRDhjxJkqQeMuRJkiT1kCFPkiSphwx5kiRJPWTIkyRJ6iFDniRJUg8Z8iRJknpo7ENeko8kuTTJdztl2ydZleRH7ft2bfkfJzknydeS7NCW3TPJiaNqvyRJ0iiMfcgDjgX2HSh7HfDlqtod+HI7DvBqYG/geOD5bdlbgDctfjMlSZLGx9iHvKr6KnDFQPF+wHHt8HHAM9rhW4AtgGXAjUkeBVxSVT9airZKkiSNi01H3YANtLyqLgGoqkuS/E5bfgRwKnAx8ALg48Bz51pYkoOAgwCWL1/O5OTkYrR5o7XmkjWjbsKSeOhdH7pB8x31e0cNuSXjyZ+LxeN3SNJiSFWNug1zSrIC+FxV7dGOX1VVd+7UX1lV2w3McwBwZ+B04DXAlcArq+q62dY1MTFRq1evHu4GbORyREbdhCVRh23Yz4L7Rwvld0jShkqypqompqsb+8O1M/hFkrsCtO+XdiuTLAMOAN4P/B3w58Aa4E+XuJ2SJEkjsbGGvJNpQhzt+0kD9X8DvLeqbgS2AormfL1lS9ZCSZKkERr7c/KSfAxYCeyYZC1wGPB24OMcamhgAAAgAElEQVRJXgRcCDy7M/3dgImqOrwt+nvgNOAq1l2gIUmS1GtjH/Kq6nkzVD1+hukvBp7aGf8E8IlFaJokSdLY2lgP10qSJGkWhjxJkqQeMuRJkiT1kCFPkiSphwx5kiRJPWTIkyRJ6iFDniRJUg8Z8iRJknrIkCdJktRDhjxJkqQeMuRJkiT1kCFPkiSphwx5kiRJPWTIkyRJ6iFDniRJUg8Z8iRJknrIkCdJktRDhjxJkqQeMuRJkiT1kCFPkiSphwx5kiRJPWTIkyRJ6qFNh7mwJPcEHg7s1Bb9Ajijqn44zPVIkiRpdgsOeUm2A17Svn53hmkuBD4EfLCqLl/oOiVJkjS7BR2uTfIa4MfA3wH3BDLDazfgSOC8dh5JkiQtooX25L2zfb8O+BzwNeDbwGU04W4H4IHAo4CnAncC3gEctcD1SpIkaRYLDXk/oenF+1hVXTvDNF8Hjk6yDHgB8DcLXOdvJdkXeC+wCfDhqnp7khOABwCfq6o3tNO9CTi7qk4a1rolSZLGWapqw2dONqmqmxd7npmWA/wQeCKwFjgD2B/426r60yRfo+k9XAYcU1VPm89yJyYmavXq1Qtt3qySLOryJUnS6C0kY81XkjVVNTFd3YJ68uYKa0m2Bm6uqt/Md571sBdwXlWd367rROAPga2S3AHYHLgZeDNw6JDWKUmStFEY6i1UpiTZBfgozbl4JPkicGBVXTbE1ewMXNQZX0tz+5YLgW+1678XTW/lmXO09yDgIIDly5czOTk5xGZKkqTbo1HniQUdrp1xocnngT8ArqLpTdsBOKGq9h/iOp4NPLmqXtyO7w/sVVWv6EzzWeBg4M9oLgBZVVUfmm25Hq6VJEnDsFEfrk1yh6q6ZbAMeBxwaFW9pS17CfDWhaxrGmuBXTvjuwAXd9qxH7Aa2BrYo6qek+SrSU6oquuG3Jb1shQfujRWbi//2PizvXj8DknrbaGPNVuTZO8Z6rrf1GGdh9d1BrB7knsk2Rx4LnAyQJLNgFcC76K58GKqLVPn6kmSJPXaQkPelsDXkxyTZHuAtmdvEnhzksuT/ILmaRerFriuW6mqm4CXA6cC5wIfr6pz2upDgOPaHruzgST5DvDfVXXVMNshSZI0jhZ6C5XNgNcCbwB+TXP7ko8kuTtwArAPTS/aKuCFVXXpwpu8uJbinDzpdsdDbVoov0PStGY7J29BPXlVdWNVvQ24P/AN4MNJvg7cuaoeBWwLbFNV+24MAU+SJKkvFnq4FoCq+mlV7Qc8HbgbsDrJu9u6kV7kIEmSdHu04JCX5I5JnpjkD4HTgfvRPNP2ZcD3kzxnoeuQJEnS+llQyEvyQOD7wBdprmz9EfCEqvo/wJ7A94ATk3wpyb0W2lhJkiTNz0J78v4B2AY4Cfgk8BvgwwBV9cOqehLwPJpz9r6zwHVJkiRpnhb6WLOHAM+sqv8ESLITcHGSXavqIoCq+rckpwCHL3BdkiRJmqeFhrwrgecnuRS4CXgWzS1TruxOVFVXA69e4LokSZI0TwsNeccD/4fm2bBTPlNV1yxwuZIkSVqAhYa8I4BrgWfQPC7sP9sySZIkjdCCQl5V3Qy8o31JkiRpTCz0Fiq/uxTzSJIkaf0s9BYq30/y0SSPnmvCJI9O8i/AuQtcpyRJkuaw0HPyNgWeT3OF7SXAfwNnA5fRXGV7F5qbIu8D3BVIWy5JkqRFtNCQ9/vA3wOPoHlm7bPa16C079/EW6lIkiQtugUdrq2q06pqH+CRNLdT+QVNoOu+LgU+CjymqvapqtMW1mRJkiTNZaE9eQBU1TeAbwAk2QXYiSbg/XzqyReSJElaOkMJeV1VtRZYO+zlSpIkaf4WenWtJEmSxpAhT5IkqYcMeZIkST1kyJMkSeohQ54kSVIPGfIkSZJ6aKi3UEmyDfAUYDdgy8H6qnrzMNcnSZKk6Q0t5CV5GHAKsP0skxnyJEmSlsAwe/L+Adhhlvoa4rokSZI0i2GGvD1pgtx/AZ8CrmWBwS7JfYB/Bh4CvLGqjurU7Qu8F9gE+HBVvb0tPwF4APC5qnpDW/Ym4OyqOmkh7ZEkSdpYDDPkXQUsA/6oqq4a0jKvAP4SeEa3MMkmwNHAE2keoXZGkpNpt6eq9kzytSR3atu0V1UdOaQ2SZIkjb1hXl17fPu+x7AWWFWXVtUZwI0DVXsB51XV+VV1A3AisF873VZJ7gBsDtxMcx7gocNqkyRJ0sZgmD15FwC/BE5K8k/ADxgIZ1V1/DTzbYidgYs642uBh1fVuUkuBL4FfBS4F5CqOnO2hSU5CDgIYPny5UxOTg6pmZIAOOqouafpA393LB6/Q9J6S9VwrodIcguzn4NXVbVBoTLJ4cA1U+fkJXk28OSqenE7vj/NIdlXDMz3WeBg4M+ABwKrqupDs61rYmKiVq9evSHNlDSTZNQtWBpD+n2qafgdkqaVZE1VTUxXN+ybIWeO19wLSA5Jclb7utsMk60Fdu2M7wJcPLCc/YDVwNbAHlX1HGD/JMvWY3skSZI2SsM8XPtnw1hIVR1Nc1HFbM4Adk9yD+BnwHOB509VJtkMeCXwVGB31vUwTp2rd90w2ipJkjSuhhLy2gsdvtKOXlpVvxnScnei6Y3bFrglyV8B96uqXyV5OXAqzS1UPlJV53RmPQQ4rqquS3J2s6h8BzhliFf+SpIkja2hnJPX3tLkBpoes92r6icLXuiIeE6etAg8n0oL5XdImtain5NXVTcDv6A57+7nw1imJEmSNtwwL7z4WPv+1CEuU5IkSRtgmBde/JDmCRXHJXk8cCbw6+4EQ7xPniRJkmYxzJD3AZpz8gK8ZJr6Yt1TMSRJkrSIhhnyYN298G4nZ8hKkiSNp7G7T54kSZIWbmghr6qOG9ayJEmStDDDfqyZJEmSxsDQevKSnD/HJFVV9xzW+iRJkjSzYZ6Tt4J1V9dOmbp1dzrDkiRJWmTDDHkXcusgtymwvH2/Abh4iOuSJEnSLIZ54cWKwbIky4A3A3+JV99KkiQtmUW98KKqrgNeC1wPHLmY65IkSdI6S3F17UOBZe27JEmSlsBiX127Jc15eQA/H9a6JEmSNLvFvrq260NDXJckSZJmsZhX10JzLt5PgROAjw5xXZIkSZrFol5dK0mSpNEY2oUXSQ5N8qYZ6h6d5NHDWpckSZJmN8zDtYfTHK6d7lYpk8AtQ16fJEmSZrDot1BJstnU4GKvS5IkSY0F9awleQzwmIGyQwcmu2/7ft1C1iVJkqT5W+jh05VAN9QFOGya6Qo4d4HrkiRJ0jwN4xy5qcOwNTDedRnwN0NYlyRJkuZhoSHvWJqLKgL8J03Qe2ynvoDLgfOq6voFrkuSJEnztKCQV1U/pbnZMUmOb4rqv4bRsHaZfwr8bTt6DfDSqvp2W7cv8F5gE+DDVfX2tvwE4AHA56rqDW3Zm4Czq+qkYbVNkiRpnA3zZsgHAiTZFHgIsENVfWGBi/0J8JiqujLJHwDHAA9PsglwNPBEYC1wRpKTabenqvZM8rUkdwKWAXtV1XS3dpEkSeqlod5CJcmzgZ8B3wQ+25Z9Ocn5SZ60vsurqm9U1ZXt6GnALu3wXjSHgM+vqhuAE4H9gBuBrZLcAdgcuBl4M7e+OESSJKn3hvnEi0cBHwN2pDlHb+oCjM8DK4BnLXAVLwKmegZ3Bi7q1K0Fdq6qc2meofst4OPAvYBU1ZkLXLckSdJGZZhPoHg9TWj8PnCfTvl/tO+P2NAFJ3ksTch75FTRNJMVQFX9VWe+zwIHJ3kj8EBgVVV9aJrlHwQcBLB8+XImJyc3tKmSpnPUUaNuwdLwd8fi8Tskrbdhhry9aYLW04AfdcrPb993ns9CkhwCvKQdfQpNz+CHgT+oqsvb8rXArp3ZdgEuHljOfsBqYGtgj6p6TpKvJjmhqm51Y+aqOobmfD8mJiZq5cqV82mqpPl67GPnnqYPquaeRhvG75C03oZ5Tt7W7fuFA+VbDbzPqqqOrqoHVdWDaELovwP7V9UPO5OdAeye5B5JNgeeC5w8Vdk+Su2VwLtoLryY+qmZOldPkiSp14bZk/czYDdue1j2Ne372g1Y5qHADsD7kwDcVFUTVXVTkpcDp9LcQuUjVXVOZ75DgOOq6rokZwNJ8h3glKq6agPaIUmStFFJDalrOMkHgIOBq4A70/Se/QjYvZ3kA1X18qGsbBFNTEzU6tWrR90MqV8y3Wm0PeShtsXjd0iaVpI1VTUxXd0wD9e+hebpFlMBD5qAF+AK4O+GuC5JkiTNYmghr6p+BuwDfAm4hSbc3dKOP6qtlyRJ0hIY5jl5tBdH7JtkS2B74Iqq+s0w1yFJkqS5DfWJF1Oq6jdVdfFUwEuyZ5LPL8a6JEmSdFsL7slLsozmkWJ3By4FPltVl7V1uwBvp7nFye3krFlJkqTRW1DIa0Pc17n1jYmvSfKHNPej+zRwR5qA5yVDkiRJS2ShPXmH0/TgdW0DfITm5sfbtGW/Am4nz6SRJEkavYWGvMfT9NBdSPNkigDPBO7V1t8MHA0c2XkkmSRJkhbZQkPeXdv3favqBwBJjgG+RxP+9q+qExe4DkmSJK2nhV5duznAVMBrh7/fqf/kApcvSZKkDTCU++QlOXSGqjek8yiaqnrzMNYnSZKk2Q3rZsiHDYzXDOWGPEmSpCUwjJA33/vfeQsVSZKkJbLQkHfEUFohSZKkoVpQyKsqQ54kSdIYWpRn10qSJGm0DHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqobEOeUn2S3J2krOSrE7yyE7dAUl+1L4OaMu2SPLFJN9N8rLOtMckefAotkGSJGkUNh11A+bwZeDkqqokewIfB+6TZHvgMGACKGBNkpOBRwFrgKcA3wLen+SBwB2q6syRbIEkSdIIjHXIq6prOqNb0wQ6gCcDq6rqCoAkq4B9gauArbj1dh0J/MXit1aSJGl8jPXhWoAkz0zyfeDzwJ+3xTsDF3UmW9uWrQJ2Ak4H3pnk6cCaqrp4CZssSZI0cmPdkwdQVZ8GPp3k0TS9ck8AMv2kdRPwfIAkmwGnAk9P8m7g7sDxVXXy4IxJDgIOAli+fDmTk5OLsSnS7ddRR426BUvD3x2Lx++QtN5SVXNPtYSSHAK8pB19SrcXLslPgIcBTwRWVtXBbfkHgcmq+lhn2lfSHL69GHg88Ebgm1W112zrn5iYqNWrVw9xiySR6f4v66Ex+33aK36HpGklWVNVE9PVjd3h2qo6uqoeVFUPApYlzU92kocAmwOX0/TQPSnJdkm2A57UltFOux3wVOB4YBlwC835fFsu6cZIkiSNyLgfrv1j4IVJbgR+DfxJNV2PVyQ5Ejijne7NUxdhtA4F3tJelXsqcAjwHeAfl7DtkiRJIzN2h2tHzcO10iLwUJsWyu+QNK2N6nCtJEmSFs6QJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHNh11AyRJmlPVqFsgbXTsyZMkSeohQ54kSVIPGfIkSZJ6yJAnSZLUQxtFyEvysCQ3J3lWp+yAJD9qXwe0ZVsk+WKS7yZ5WWfaY5I8eBRtlyRJGoWxD3lJNgHeAZzaKdseOAx4OLAXcFiS7YAnA2uAPYGD2mkfCNyhqs5c4qZLkiSNzNiHPOAVwKeASztlTwZWVdUVVXUlsArYF7gR2Ipb3xrmSODQJWqrJEnSWBjr++Ql2Rl4JvA44GGdqp2Bizrja9uyTwD7A6cD70zydGBNVV08x3oOou35W758OZOTk8PaBEkARx016hYsDX93SBojYx3ygH8A/raqbk7SLc8001ZV3QQ8HyDJZjSHeJ+e5N3A3YHjq+rkaWY8BjgGYGJiolauXDnUjZBu9x772FG3YGl4w15JY2TsQl6SQ4CXtKN3Ak5sA96OwFOS3ETTc7eyM9suwOTAol4GHAc8ArgB+BPgm8BtQp4kSVLfjF3Iq6qjgaMHy5McC3yuqj7TXnjxtvZiC4AnAa/vTLsd8NS2/OnALUABWy5u6yVJksbDxnDhxW1U1RU0F1Sc0b7e3JZNORR4S1UVzSHbCeA7wIeWuq2SJEmjkPIckluZmJio1atXj7oZUr9kutNoe8jfp5KWWJI1VTUxXd1G2ZMnSZKk2RnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSesiQJ0mS1EOGPEmSpB4y5EmSJPWQIU+SJKmHDHmSJEk9ZMiTJEnqIUOeJElSDxnyJEmSemisQ16SlUl+meSs9nVop27fJD9Icl6S13XKT0hydpK3dcrelGS/pW6/JEnSqGw66gbMw9eq6qndgiSbAEcDTwTWAmckOZl2e6pqzyRfS3InYBmwV1UducTtliRJGpmx7smbxV7AeVV1flXdAJwI7AfcCGyV5A7A5sDNwJuBQ2dckiRJUg9tDD15j0jybeBi4DVVdQ6wM3BRZ5q1wMOr6twkFwLfAj4K3AtIVZ052wqSHAQcBLB8+XImJyeHvxXS7dlXvjLqFiwNf3dIGiPjHvK+BexWVdckeQrwGWB3INNMWwBV9VdTBUk+Cxyc5I3AA4FVVfWh28xYdQxwDMDExEStXLly2NshSZK0pMbucG2SQ6YutADuWFXXAFTVKcBmSXak6bnbtTPbLjQ9fd3l7AesBrYG9qiq5wD7J1m2FNshSZI0SmMX8qrq6Kp6UFU9CLglSQCS7EXT3suBM4Ddk9wjyebAc4GTp5aRZDPglcC7aC68qLZq6lw9SZKkXhv3w7XPAl6a5Cbg18Bzq6qAm5K8HDgV2AT4SHuu3pRDgOOq6rokZwNJ8h3glKq6aom3QZIkacmlyUyaMjExUatXrx51MyRJkuaUZE1VTUxXN3aHayVJkrRwhjxJkqQeMuRJkiT1kCFPkiSphwx5kiRJPWTIkyRJ6iFDniRJUg8Z8iRJknrImyEPSPK/wE9H3Q6xI3DZqBsxxtw/c3Mfzc79Mzf30ezcP+Nht6q6y3QVhjyNpSSrZ7qDt9w/8+E+mp37Z27uo9m5f8afh2slSZJ6yJAnSZLUQ4Y8jatjRt2AMef+mZv7aHbun7m5j2bn/hlznpMnSZLUQ/bkSZIk9ZAhT5IkqYcMeRqKJNcMYRmvSvK9JGcn+XKS3Tp1ByT5Ufs6oFP+1iQXDa4/yd2TfCXJme3ynrLQ9m2IJDcnOavzWjHLtCuTfG4eyzwhyQ+SfDfJR5Js1pYnyfuSnNdu80M683wxyVWDy0/y+CTfatv29ST32vCtXX9JKslHO+ObJvnf+eyHWZa5a/vZn5vknCSv7NRtn2RV+z1alWS7tvw+Sb6Z5PokrxlY3l+3y/luko8l2XJD27aAbXpmu6/uM4RlvSvJ99vvyKeT3LlT9/r2+/ODJE/ulH8kyaVJvjuwrAclOa39/qxOstdC27ee2zL0709nWU9MsibJd9r3x3XqHtqWn9f+zKUtf3b7XbklyURn+s2SHNfOc26S1y+0fRu4TcP4Pd27/dJrVeXL14JfwDVDWMZjgWXt8EuBf2uHtwfOb9+3a4e3a+v2Bu46uH6aE4Jf2g7fD7hg3PcLsBL43DymewqQ9vWxznY+BfhCW743cHpnnscDTxtcPvBD4L7t8MuAY5d6/wBnAlu1438AnDWf/dBZxqYD43cFHtIOb9Nu4/3a8XcCr2uHXwe8ox3+HeBhwFuB13SWtTPwk077Pg4cOILv0ceBrwGHb8C8mwyMP2lqnwHv6OyD+wHfBrYA7gH8eGpe4NHAQ4DvDizrS8AfdL5/kxvb92eWZT8YuFs7vAfws07d/wCPaH/WvtDZB/cF7g1MAhOd6Z8PnNgOLwMuAFaM4Hs0jN/TvdsvfX7Zk6ehSXLHND1w32r/M9uvLV/R/pf2ofa/uS8l2Wpw/qr6SlVd146eBuzSDj8ZWFVVV1TVlcAqYN92ntOq6pJpmlPAtu3wnYCLh7elC5Nkk7Y35Yy2N+XgTvW2be/K95L8Y5Lb/IxW1SnVovmlOrWf9gOOb6tOA+6c5K7tPF8Grp6mOeOwn74A/GE7/Dya4ApAkr2SfCNNj+w3kty7LT8wySeSfJYmaPxWVV1SVd9qh68GzqUJa9Dso+Pa4eOAZ7TTXVpVZwA3TtO+TYGtkmxK84doSfdRkjsC+wAvAp7bKV+Z5KvTfV+SXJPkzUlOp/mj+1tV9aWquqkd7f6c7UfzB/f6qvoJcB6wVzvPV4Erpmnexvr9+VqSB3Wm++8ke3YXWlVnVtXU9pwDbJlki/Znatuq+mb7M3g8675H51bVD6ZpYwFbt9+hrYAbgF8NYdvXWwaOGCT5f0kObIcvSHJE53f4bXqO+7pf+sqQp2H6DfDMqnoITa/c30911wO7A0dX1f2Bq4A/nmNZL6L55Q3NH+iLOnVrWfdHeyaHAy9IshY4BXjFfDdiyLbKukO1n27LXgT8sqoeRtN79JIk92jr9gJeDTwAuCfwRzMtOM1h2v2BL7ZFG7KfXgyc0u6n/YG3z3vLhudE4LlpDoPuCZzeqfs+8OiqejBwKPC2Tt0jgAOq6nHMIM3h8Qd3lrl86p+C9v13ZmtYVf0MOAq4ELiE5nP70mzzLIJnAF+sqh8CV6RzGJ6Zvy9b0/S6Pbyqvj7Lsv+chf2c/RXwriQX0eynURxu25Dvz4eBAwGS/B6wRVWdPcs6/hg4s6qup9knazt189lPnwSupfkOXQgcVVXTheZxcFn7O/wDwGvmmPb2tF82SoY8DVOAtyU5G/gPmh/w5W3dT6rqrHZ4DbBixoUkLwAmgHd1ljtornv/PI/m0OMuNIeRPjpdr9gS+HVVPah9PbMtexLwwiRn0fxB2oEmBAP8T1WdX1U30/RIPHKWZb8f+GpVfa0d35D99NfAU9r99M/Au+fepOFq/7iuoPnMThmovhPwiTTngr0HuH+nbtVsfxDaHrBPAX9VVRvUO5DmnL39aA5f3o2m1+EFG7KsBXgeTZChfX9ep26m78vNNNs+oyRvBP5/e/cfI0dZx3H8/TGYqiVUaYuIYEvVRtGIqGkqaqyl1koaImpotKZU0IT+gUENUVNCTzH+wZ8CsVVqqlBBEZALaBpFqk1FoyZg+VUp0muqsdfURmrV5s5+/eN5Nvvcdra7t73e0dnPK5nM3DPPPDf77Ozsd555np1RYHMjqSJbp+NnDfD5iDiPdCxt7JB/wvV4/NwDLM8XSlcBm9qVL+ktpNvajRb3XuppAek9OYd0LH1R0rwO20yV+/K803m63+rllOQgzybSSmA28M6IeDuwD2h0Uj9S5Psf6RbYMSQtAdYCl+WrQ0hXhOcV2c6l822hq0n9mIiIR/N+zOr6lZxcAq4tgr/zi9ah1pNi5UlS0jpSXX+hSB5XPUmaDVwYEY2Wjx8CF3f/MibUIKkl6K6W9JuARyLiraQ+heWgh8PtCstf3vcCmyPivmLVvsYt7Dwf7rBfS0gXKPsjYoT0BThpdSRpJrAYuF3SbuB6YEXRQt7uePlvDvzalXslsBxYmW+tQW+fsytpBgX3kG/vToFxHT+5W8jPSQH8FcAPqgqVdC5wP7AqIp7LyXtp3uKG7urpk6TW2JGIGAa2ky5kp8IoY7/7WwcSNc67xztP17FeaslBnk2kGcBwRIxI+gAwp9MGJUkXARtIAV755bsFWCrpVbllZWlOO549pMEGSHoz6US2fzz7cxJtAdaoOSp2vqTped0CSefnVscVwDG32iR9htRP8RMRcbRYNUhqIZSkhaRbi1X9FRsOAjPy7SqAD5L6r02F7wJfi4gdLekzgL/m5dXdFJQDoI3A0xHR2jI5SApMyPMHOhS3B1go6RW53EuY3Dr6OKmf5ZyImJtbzJ6n2WLX8XhpJWkZ8CXS5+zfxapB0m3Pabn7wBtJfT6P52/A+/PyYuDZbl/YBOvl+Lkd+Cbw+6oWYaVRxw8BX4mI7Y30/Jk6JGlhPiZW0d1xtDh/NqeTBkY909Urm3hDwAX5fZ5BPk92q8b1Uk+tIzE8eRrvRLraO0BqKXsU+APpBPo0qbl/LsWoPFI/j4GKcn5Bav17LE+DxbqrSB3BdwGfLtJvJl1BHs3zgZx+Aemq8PFc1tIpqptjRrORLq6+AewAngAeIX0ZLQJ+SWpRewpYD7ykYvtR0sjHRj3dmNMF3JbX7WDsKLZtpCD3P7mePpTTL895HyeNfJv3IqifReTRkaR+d3/O7+VN5FHSpC/sW9uU+V5Si9afijq6NK+bCTxMCkYeBs7M6WfnenmB1Gd0L6kTOcBXSV88TwB3kPpvTVb9bAWWtaR9jtRfqu3xUlWvxfa7SH3vGnWzvli3Nh8/O8kjI3P6XaR+UyO5bq4u6vqP+fj5HakV/0V//BR5n2mt32LdDaTW4seK6ay87l35eHgOuJXm06Muz/VzhHQu25LTTye1dD6Z36vrJ7Oe8j6cBhzIyzfn9/hBUkvs6py+G5hVvMatda+Xuk9+rJmdMEkXAt+JiKm6VWPWdyQtIv3cy/Kp3pdTkaRzSEH0m2Jsi3gt+Tzdn3y71k6IpGtIV/k3TPW+mJl1Q9IqUsvj2j4J8Hye7lNuyTMzMzOrIbfkmZmZmdWQgzwzMzOzGnKQZ2ZmZlZDDvLMrG9JGpAUxXRLRZ7bWvIMTMJ+XZf37bqKdauLfVl9svfFzE5dDvLMzJpWFT9MTV6e7MeYQXom7Lo8N8ZIjsAAAAJ8SURBVDPriYM8M7OmM0iPWmpYmdPMzE45DvLMzJKhPL+mSGss767aQNL7JA1K2i9pRNLfJd0t6W0t+TYVt1gvlnSnpIOSDkj6saSzc75FkoLmIwHnFNtV7cNpktZJGpL0L0m/keQfuzUzoM3Dh83M+tD3gC8D78iBkoCLSI/y2gQMlJklfSpvU14sv5r0DNmPSFoWEVsr/s9DwCuLvz+W/17Swz5/Pf/PhncDP5M0LyL+2UN5ZlYjbskzM0v2A/fm5TV5IqftKzPmvnq3kM6ho6Rnc55Bs+VvGrChzf95Hng9MB8YzmmXSHpNRGyNCNFsVRyKCOVpbkVZ00jPap1JekQXwJnApR1eq5n1AQd5ZmZN38rzFcAVLWml99BsjftpRPwkIg5FxAbSA9sB5kt6Q8W2N0bEXyLiWWBbkT6nIm8nGyPiVxHxD5oBaq9lmVnNOMgzM8siYhvwJPDyPD0VEb+uyDq7WN7Tsm6oWD6rYtudxfLhYvll49jVk1GWmdWMgzwzs7HWF8tVrXiQbu02vK5lXfn3MMcaKZbbPTy824eKd1OWmfUpB3lmZmN9H7g/T3e0ybMdOJiXPyzpMkmnS/osabAGwM6I2NXjPhzI81mSXttjGWbW5xzkmZkVIuKFiPhonipHqEbEYeBa4CjwUuAB4BDw7ZzlCGN/imW8fpvn04G9+SdUNp1AeWbWhxzkmZn1ICI2k0a2PkhqeRsljcL9EbCgzc+ndGsAuJuxt4XNzMZFEe7GYWZmZlY3bskzMzMzqyEHeWZmZmY15CDPzMzMrIYc5JmZmZnVkIM8MzMzsxpykGdmZmZWQw7yzMzMzGrIQZ6ZmZlZDf0fBwQ4O3jsNLkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x864 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 3) \n",
    "\n",
    "df = pd.DataFrame.from_dict(data)\n",
    "\n",
    "rows = []\n",
    "columns = []\n",
    "for i in range(len(df)):\n",
    "    rows.append(df.columns[i])\n",
    "    c = df.columns[i]\n",
    "    columns.append(df[c].max())\n",
    "\n",
    "\n",
    "# create date dataframe\n",
    "df1 = pd.DataFrame(list(rows))\n",
    "df1.columns = ['date']\n",
    "\n",
    "# create start_balance data frame\n",
    "df2 = pd.DataFrame(list(columns))\n",
    "df2.columns = ['invest_balance']\n",
    "\n",
    "# create dataframe equal to df1\n",
    "df=df1\n",
    "\n",
    "\n",
    "\n",
    "df['invest_balance'] = df2\n",
    "column2 = []\n",
    "for num in range(0,len(df)-1):\n",
    "    gain = df.invest_balance[num+1]-df.invest_balance[num]\n",
    "    growth = df.invest_balance[num+1] / df.invest_balance[num]\n",
    "    column2.append(round((growth - 1)*100,2))\n",
    "    \n",
    "c2 = pd.DataFrame(list(column2))\n",
    "\n",
    "df['invest_return'] = c2\n",
    "\n",
    "display(df)\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(2,1, figsize = (10,12), sharex=True)\n",
    "\n",
    "# plot - Monthly Investment Balance\n",
    "ax[0].plot(df.date, df.invest_balance, linewidth=4, marker='o', markersize=8)\n",
    "ax[0].set_title('Monthly Investment Balance', fontsize=20, weight='bold')\n",
    "start, end = ax[0].get_ylim()\n",
    "ax[0].set_ylim(0)\n",
    "ax[0].set_ylabel('Balance ($)', fontsize=16, weight='bold')\n",
    "fmt_dollar = '${x:,.0f}'\n",
    "ax[0].yaxis.set_major_formatter(ticker.StrMethodFormatter(fmt_dollar))\n",
    "\n",
    "# determine positive values for return\n",
    "positive_return = pd.Series(df.invest_return >=0)\n",
    "\n",
    "# plot - Monthly Investment Return\n",
    "ax[1].bar(df.date, df.invest_return, color=positive_return.map({True: 'g', False: 'r'}))\n",
    "ax[1].set_title('Monthly Investment Return', fontsize=20, weight='bold')\n",
    "ax[1].set_xlabel('Month', fontsize=16, weight='bold')\n",
    "ax[1].set_ylabel('Return (%)', fontsize=16, weight='bold')\n",
    "ax[1].hlines(0,-1,len(df.date), linewidth = 4) \n",
    "fmt_percent = '{x:,.0f}%'\n",
    "ax[1].yaxis.set_major_formatter(ticker.StrMethodFormatter(fmt_percent))\n",
    "ax[0].grid()\n",
    "ax[1].yaxis.grid()\n",
    "plt.show()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4: Clinical disease data (40 pts)\n",
    "\n",
    "Your boss comes to you Monday morning and says “I figured out our next step; we are going to pivot from an online craft store and become a data center for genetic disease information! I found **ClinVar** which is a repository that contains expert curated data, and it is free for the taking. This is a gold mine! Look at the file and tell me what gene and mutation combinations are classified as dangerous.”\n",
    "\n",
    "Make sure that you only give your boss the dangerous mutations and include:\n",
    "\n",
    "1) Gene name\n",
    "\n",
    "2) Mutation ID number\n",
    "\n",
    "3) Mutation Position (chromosome & position)\n",
    "\n",
    "4) Mutation value (reference & alternate bases)\n",
    "\n",
    "5) Clinical severity \n",
    "\n",
    "6) Disease that is implicated\n",
    "\n",
    "**Requirements**\n",
    "\n",
    "1) The deliverables are the final result as a dataframe with a short discussion of any specifics. (that is, what data you would present to your boss with the explanation of your results)\n",
    "\n",
    "2) Limit your output to the first 100 harmful mutations and tell your boss how many total harmful mutations were found in the file\n",
    "\n",
    "3) Use the instructor-modified \"clinvar_final.txt\" at this link: https://drive.google.com/file/d/1Zps0YssoJbZHrn6iLte2RDLlgruhAX1s/view?usp=sharing This file was modified to be not exactly the same as 'standard' .vcf file to test your data parsing skills. **This is a large file so do NOT upload it into your github repo!**\n",
    "\n",
    "4) Replace missing values in the dataframe with: 'Not_Given'\n",
    "\n",
    "5) State in your answer how you define harmful mutations\n",
    "\n",
    "**6) Do your best on getting to above requirements and submit whatever you do before the 24 hour deadline. If your work is incomplete be sure to describe the blockers that got in your way and how you might get past them (if given more time).**\n",
    "\n",
    "**Hints** \n",
    "* We do not expect you to have any medical knowledge to solve this problem; look at the data, read the documentation provided, and write down your assumptions!\n",
    "\n",
    "* Correct pseudocode will give you partial credit so start with that. \n",
    "\n",
    "* Map out which fields you want to extract: Are they in the same place every time? What strategy will you use to robustly extract and filter your data of interest? How do you plan to handle missing data?\n",
    "\n",
    "* A good way to start is to print out each line, then practice parsing them to see if you can recover the fields of interest\n",
    "\n",
    "* A starting solution for parsing .vcfs can be found here: https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744 This solution does **NOT** work due to the changes we've made but can be modified to work. As with any solution that needs modifications, it may take less time to make your own solution!\n",
    "\n",
    "* Filter out junk and lines with no mutation data. Just focus on the data your need to deliver to your boss. \n",
    "\n",
    "* Pandas and NumPy parsers correctly recognize the end of each line in in the ClinVar file.\n",
    "\n",
    "* The unit of observation of this dataset is one row per mutation.\n",
    "\n",
    "* This is similar to a task that one of us tackled at work. You can answer the question with the information provided below or using the (partial) data dictionary file at this link: https://drive.google.com/file/d/1lx9yHdlcqmU_OlHiTUXKC_LQDqYBypH_/view?usp=sharing. Our goal is to see that you can put together a sensible plan, implement a solid parsing strategy, and document and justify the decisions that you made."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### VCF file description (Summarized from version 4.1)\n",
    "\n",
    "```\n",
    "* The VCF specification:\n",
    "\n",
    "VCF is a text file format which contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. The format also can contain genotype information on samples for each position.\n",
    "\n",
    "* Fixed fields:\n",
    "\n",
    "There are 8 fixed fields per record. All data lines are **tab-delimited**. In all cases, missing values are specified with a dot (‘.’). \n",
    "\n",
    "1. CHROM - chromosome number\n",
    "2. POS - position DNA nuceleotide count (bases) along the chromosome\n",
    "3. ID - The unique identifier for each mutation\n",
    "4. REF - reference base(s)\n",
    "5. ALT - alternate base(s)\n",
    "6. FILTER - filter status\n",
    "7. QUAL - quality\n",
    "8. INFO - a semicolon-separated series of keys with values in the format: <key>=<data>\n",
    "\n",
    "```\n",
    "### Applicable INFO field specifications\n",
    "\n",
    "```\n",
    "GENEINFO = <Gene name>\n",
    "CLNSIG =  <Clinical significance>\n",
    "CLNDN = <Disease name>\n",
    "```\n",
    "\n",
    "### Sample ClinVar data (vcf file format - not exactly the same as the file to download!)\n",
    "\n",
    "```\n",
    "##fileformat=VCFv4.1\n",
    "##fileDate=2019-03-19\n",
    "##source=ClinVar\n",
    "##reference=GRCh38\t\t\t\t\t\t\t\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n",
    "1\t949523\trs786201005\tC\tT\t.\t.\tGENEINFO=ISG15;CLNSIG=5\n",
    "1\t949696\trs672601345\tC\tCG\t.\t.\tGENEINFO=ISG15;CLNSIG=5;CLNDBN=Cancer\n",
    "1\t949739\trs672601312\tG\tT\t.\t.\tGENEINFO=ISG15;CLNDBN=Cancer\n",
    "1\t955597\trs115173026\tG\tT\t.\t.\tGENEINFO=AGRN;CLNSIG=2; CLNDBN=Cancer\n",
    "1\t955619\trs201073369\tG\tC\t.\t.\tGENEINFO=AGG;CLNDBN=Heart_dis \n",
    "1\t957640\trs6657048\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=3;CLNDBN=Heart_dis \n",
    "1\t976059\trs544749044\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=0;CLNDBN=Heart_dis \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4) Your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in file and create a line for everything that does not start with a #\n",
    "file = open(\"clinvar_final.txt\",\"r\") \n",
    "records = [r for r in file if not r.startswith('#')]\n",
    "file.close() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build the DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check the list comprehension records from above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM\tPOS\tID\tREF\tALT\tFILTER\tQUAL\tINFO\n",
      " \n",
      "\n",
      "1\t1014O42\t475283\tG\tA\t.\t.\tAF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014042G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=143888043\n",
      " \n",
      "\n",
      "1\t1O14122\t542074\tC\tT\t.\t.\tAF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014122C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=150861311\n",
      " \n",
      "\n",
      "1\t1014143\t183381\tC\tT\t.\t.\tALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014143C>T;CLNREVSTAT=no_assertion_criteria_provided;CLNSIG=Pathogenic;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;CLNVI=OMIM_Allelic_Variant:147571.0003;GENEINFO=ISG15:9636;MC=SO:0001587|nonsense;ORIGIN=1;RS=786201005\n",
      " \n",
      "\n",
      "1\t1014179\t542075\tC\tT\t.\t.\tALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014179C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1553169766\n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'1\\t1014O42\\t475283\\tG\\tA\\t.\\t.\\tAF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014042G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=143888043\\n'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for item in range(0,5):\n",
    "    print(records[item], '\\n')\n",
    "    \n",
    "records[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Examine the length of the list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "102322"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "There are 102322 potential rows of data in list lines.  Next I will split all the items in list lines, except \n",
    "for the first item, using split() to split at the '\\t'.  I want to drop the column headers now and rename them \n",
    "in the dataframe later.\n",
    "\"\"\"\n",
    "len(records)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Print the first \"future\" row and column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the first future row in the dataframe:\n",
      " ['1', '1014O42', '475283', 'G', 'A', '.', '.', 'AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014042G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=143888043'] \n",
      "\n",
      "This I will parse to make columns:\n",
      "AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563;CLNDN=Immunodeficiency_38_with_basal_ganglia_calcification;CLNHGVS=NC_000001.11:g.1014042G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=ISG15:9636;MC=SO:0001583|missense_variant;ORIGIN=1;RS=143888043\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Now I will create a new list called rows.  Rows will contain a list of lists split at the '\\t'.  I will print the\n",
    "first list within the list as an example.  I will also print the info column of the first row.\n",
    "\"\"\"\n",
    "# Append everything except the first line to list row\n",
    "rows = []\n",
    "for n in range(1,len(records)):\n",
    "    rows.append(records[n].split())\n",
    "# This is the fist list of lists\n",
    "print(\"This is the first future row in the dataframe:\\n\",rows[0],'\\n')\n",
    "\n",
    "# This is the 'info' column\n",
    "print(\"This I will parse to make columns:\\n\"+rows[0][7])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Ensure all the lists in rows are the same length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All lists in rows are the same length. Checked 102321 rows.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Here I will check to make sure all the the lists within rows are the same length.\n",
    "\"\"\"\n",
    "for n in range(len(rows)):\n",
    "    if len(rows[n]) != len(rows[0]):\n",
    "        print(n)\n",
    "        print('False')\n",
    "        break\n",
    "else: print('All lists in rows are the same length. Checked', n+1, 'rows.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create pandas series for GENEINFO, CLNSIG, and CLNDN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Next, I will use regular expresions to find the gene name (GENEINFO), clinical severity (CLNSIG), and disease \n",
    "implication (CLNDN).  I will create three lists, on for each future variable.  The rows that do not contain \n",
    "information for any of these three future variables I will assign 'Not_Given'.\n",
    "\"\"\"\n",
    "# Break up the info column into three pandas series (gene_info, clinical_severity, and disease_name)\n",
    "gene_name_list = []\n",
    "clin_sev_list = [] \n",
    "dis_name_list = []\n",
    "\n",
    "for n in range(len(rows)):\n",
    "    source = rows[n][7]\n",
    "    GENEINFO = re.compile('(?<=GENEINFO=).*?(?=;)')\n",
    "    CLNSIG = re.compile('(?<=CLNSIG=).*?(?=;)')\n",
    "    CLNDN = re.compile('(?<=CLNDN=).*?(?=;)')\n",
    "\n",
    "    m1 = GENEINFO.search(source)\n",
    "    m2 = CLNSIG.search(source)\n",
    "    m3 = CLNDN.search(source)\n",
    "\n",
    "    # m1\n",
    "    if m1:\n",
    "        item1 = m1.group()\n",
    "    else:\n",
    "        item1 = 'Not_Given'\n",
    "\n",
    "    # m2    \n",
    "    if m2:\n",
    "        item2 = m2.group()\n",
    "    else:\n",
    "        item2 = 'Not_Given'\n",
    "\n",
    "    # m3    \n",
    "    if m3:\n",
    "        item3 = m3.group()\n",
    "    else:\n",
    "        item3 = 'Not_Given'\n",
    "        \n",
    "    gene_name_list.append(item1)\n",
    "    clin_sev_list.append(item2)\n",
    "    dis_name_list.append(item3)\n",
    "    \n",
    "# Convert list to pandas series\n",
    "gene_name_s = pd.Series(gene_name_list)\n",
    "clin_sev_s = pd.Series(clin_sev_list)\n",
    "dis_name_s = pd.Series(dis_name_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Visualize pandas series results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    ISG15:9636\n",
      "1    ISG15:9636\n",
      "2    ISG15:9636\n",
      "3    ISG15:9636\n",
      "4    ISG15:9636\n",
      "dtype: object 0                    Benign\n",
      "1    Uncertain_significance\n",
      "2                Pathogenic\n",
      "3    Uncertain_significance\n",
      "4                    Benign\n",
      "dtype: object 0    Immunodeficiency_38_with_basal_ganglia_calcifi...\n",
      "1    Immunodeficiency_38_with_basal_ganglia_calcifi...\n",
      "2    Immunodeficiency_38_with_basal_ganglia_calcifi...\n",
      "3    Immunodeficiency_38_with_basal_ganglia_calcifi...\n",
      "4    Immunodeficiency_38_with_basal_ganglia_calcifi...\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Visualize the series I created from the lists.\n",
    "print(gene_name_s[0:5], clin_sev_s[0:5], dis_name_s[0:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build the DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinical_severity</th>\n",
       "      <th>disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475283</td>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542074</td>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542075</td>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475278</td>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_info      id chrom      pos ref alt       clinical_severity  \\\n",
       "0  ISG15:9636  475283     1  1014O42   G   A                  Benign   \n",
       "1  ISG15:9636  542074     1  1O14122   C   T  Uncertain_significance   \n",
       "2  ISG15:9636  183381     1  1014143   C   T              Pathogenic   \n",
       "3  ISG15:9636  542075     1  1014179   C   T  Uncertain_significance   \n",
       "4  ISG15:9636  475278     1  1014217   C   T                  Benign   \n",
       "\n",
       "                                  disease_implicated  \n",
       "0  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "4  Immunodeficiency_38_with_basal_ganglia_calcifi...  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a DataFrame from the list of lists, rows.\n",
    "df = pd.DataFrame(rows)\n",
    "df.head(10)\n",
    "\n",
    "# Rename the columns.\n",
    "df.columns=['chrom', 'pos', 'id', 'ref', 'alt', 'filter', 'qual', 'information']\n",
    "\n",
    "# Drop the columns I don't wnat.\n",
    "df.drop(['filter', 'qual','information'],axis=1,inplace=True)\n",
    "\n",
    "# Rearrange the columns.\n",
    "df = df[['id', 'chrom', 'pos', 'ref', 'alt']]\n",
    "\n",
    "# Insert gene info, clinical severity, and disease implicated\n",
    "df.insert(0,'gene_info',gene_name_s)\n",
    "df.insert(6,'clinical_severity', clin_sev_s)\n",
    "df.insert(7,'disease_implicated', dis_name_s)\n",
    "\n",
    "# Print the df DataFrame head.\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data cleaning and sanity check."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### What is the shape of the DataFrame?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(102321, 8)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "The shape of the DataFrame is what I anticipated it to be.  There area 102,321 rows and 8 columns.\n",
    "\"\"\"\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### What type of data types are in the data set?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_info             object\n",
       "id                    object\n",
       "chrom                 object\n",
       "pos                   object\n",
       "ref                   object\n",
       "alt                   object\n",
       "clinical_severity     object\n",
       "disease_implicated    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many clinical severities are there in the DataFrame?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1797 rows that do not have a clinical severity listed.\n"
     ]
    }
   ],
   "source": [
    "print(\"There are\", df.clinical_severity[df.clinical_severity=='Not_Given'].count(), \n",
    "      \"rows that do not have a clinical severity listed.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many different types of clinical severities exist in the data set?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Uncertain_significance                                                                      47980\n",
       "Likely_benign                                                                               17885\n",
       "Pathogenic                                                                                  12313\n",
       "Likely_pathogenic                                                                            6269\n",
       "Benign                                                                                       6138\n",
       "Conflicting_interpretations_of_pathogenicity                                                 5404\n",
       "Benign/Likely_benign                                                                         3338\n",
       "Not_Given                                                                                    1797\n",
       "Pathogenic/Likely_pathogenic                                                                  854\n",
       "risk_factor                                                                                    98\n",
       "association                                                                                    70\n",
       "drug_response                                                                                  41\n",
       "Affects                                                                                        31\n",
       "Conflicting_interpretations_of_pathogenicity,_risk_factor                                      14\n",
       "other                                                                                          12\n",
       "Pathogenic,_risk_factor                                                                        11\n",
       "Pathogenic/Likely_pathogenic,_other                                                             8\n",
       "Conflicting_interpretations_of_pathogenicity,_other                                             8\n",
       "Benign/Likely_benign,_risk_factor                                                               7\n",
       "protective                                                                                      6\n",
       "Pathogenic/Likely_pathogenic,_risk_factor                                                       6\n",
       "Likely_pathogenic,_risk_factor                                                                  5\n",
       "protective,_risk_factor                                                                         4\n",
       "Pathogenic,_other                                                                               2\n",
       "Pathogenic,_Affects                                                                             2\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects                                           1\n",
       "Pathogenic,_association,_protective                                                             1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_other                                    1\n",
       "Conflicting_interpretations_of_pathogenicity,_other,_risk_factor                                1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other        1\n",
       "Conflicting_interpretations_of_pathogenicity,_protective                                        1\n",
       "Likely_pathogenic,_other                                                                        1\n",
       "Benign,_drug_response                                                                           1\n",
       "Benign,_other                                                                                   1\n",
       "Likely_pathogenic,_association                                                                  1\n",
       "Conflicting_interpretations_of_pathogenicity,_association                                       1\n",
       "Pathogenic,_protective                                                                          1\n",
       "Benign/Likely_benign,_other                                                                     1\n",
       "Benign/Likely_benign,_association                                                               1\n",
       "Uncertain_significance,_risk_factor                                                             1\n",
       "Affects,_risk_factor                                                                            1\n",
       "Benign/Likely_benign,_protective                                                                1\n",
       "Likely_benign,_other                                                                            1\n",
       "Name: clinical_severity, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Note there are 1797 clinical severity cases not given.\n",
    "\"\"\"\n",
    "df.clinical_severity.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TTN:7273|TTN-AS1:100506866                        6294\n",
       "TTN:7273                                          4817\n",
       "MSH6:2956                                         3507\n",
       "MSH2:4436                                         2913\n",
       "MLH1:4292                                         2337\n",
       "                                                  ... \n",
       "WARS2:10352|WARS2-AS1:101929147                      1\n",
       "KLF7:8609                                            1\n",
       "EHBP1:23301                                          1\n",
       "OLA1:29789                                           1\n",
       "UBA5:79876|ACAD11:84129|NPHP3-ACAD11:100532724       1\n",
       "Name: gene_info, Length: 1511, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "91228     1\n",
       "378831    1\n",
       "619522    1\n",
       "500094    1\n",
       "237415    1\n",
       "         ..\n",
       "206912    1\n",
       "382294    1\n",
       "485567    1\n",
       "223310    1\n",
       "224317    1\n",
       "Name: id, Length: 102321, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "2    49826\n",
       "1    32381\n",
       "3    20114\n",
       "Name: chrom, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "37025629     31\n",
       "149172318    13\n",
       "73385903     13\n",
       "47414420     11\n",
       "178713381    11\n",
       "             ..\n",
       "240014758     1\n",
       "227164799     1\n",
       "197103405     1\n",
       "15644371      1\n",
       "47801041      1\n",
       "Name: pos, Length: 95548, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "C                         31467\n",
       "G                         31182\n",
       "T                         16032\n",
       "A                         15944\n",
       "CT                          371\n",
       "                          ...  \n",
       "TTAATG                        1\n",
       "GCTTATTGCACTGC                1\n",
       "GTGAAAGGTCAACTGTAGATGC        1\n",
       "ATAGTCCAGCTGTTTTTAACAG        1\n",
       "TCTCCCCATCCAGACCTA            1\n",
       "Name: ref, Length: 2032, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "T                                                                        30115\n",
       "A                                                                        29405\n",
       "C                                                                        19413\n",
       "G                                                                        19070\n",
       "TA                                                                         292\n",
       "                                                                         ...  \n",
       "GGCAAGTATATACTTGCCT                                                          1\n",
       "CGTCAA                                                                       1\n",
       "AAAAAAC                                                                      1\n",
       "ATGTTCTAATAATT                                                               1\n",
       "GTCTTTTCTTTCTTTTAGAAAATAGTGACAGTTTTAATCTCTTTGTAGATATTTGCATTTAAGGTATCA        1\n",
       "Name: alt, Length: 1196, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "Uncertain_significance                                                                      47980\n",
       "Likely_benign                                                                               17885\n",
       "Pathogenic                                                                                  12313\n",
       "Likely_pathogenic                                                                            6269\n",
       "Benign                                                                                       6138\n",
       "Conflicting_interpretations_of_pathogenicity                                                 5404\n",
       "Benign/Likely_benign                                                                         3338\n",
       "Not_Given                                                                                    1797\n",
       "Pathogenic/Likely_pathogenic                                                                  854\n",
       "risk_factor                                                                                    98\n",
       "association                                                                                    70\n",
       "drug_response                                                                                  41\n",
       "Affects                                                                                        31\n",
       "Conflicting_interpretations_of_pathogenicity,_risk_factor                                      14\n",
       "other                                                                                          12\n",
       "Pathogenic,_risk_factor                                                                        11\n",
       "Pathogenic/Likely_pathogenic,_other                                                             8\n",
       "Conflicting_interpretations_of_pathogenicity,_other                                             8\n",
       "Benign/Likely_benign,_risk_factor                                                               7\n",
       "protective                                                                                      6\n",
       "Pathogenic/Likely_pathogenic,_risk_factor                                                       6\n",
       "Likely_pathogenic,_risk_factor                                                                  5\n",
       "protective,_risk_factor                                                                         4\n",
       "Pathogenic,_other                                                                               2\n",
       "Pathogenic,_Affects                                                                             2\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects                                           1\n",
       "Pathogenic,_association,_protective                                                             1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_other                                    1\n",
       "Conflicting_interpretations_of_pathogenicity,_other,_risk_factor                                1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other        1\n",
       "Conflicting_interpretations_of_pathogenicity,_protective                                        1\n",
       "Likely_pathogenic,_other                                                                        1\n",
       "Benign,_drug_response                                                                           1\n",
       "Benign,_other                                                                                   1\n",
       "Likely_pathogenic,_association                                                                  1\n",
       "Conflicting_interpretations_of_pathogenicity,_association                                       1\n",
       "Pathogenic,_protective                                                                          1\n",
       "Benign/Likely_benign,_other                                                                     1\n",
       "Benign/Likely_benign,_association                                                               1\n",
       "Uncertain_significance,_risk_factor                                                             1\n",
       "Affects,_risk_factor                                                                            1\n",
       "Benign/Likely_benign,_protective                                                                1\n",
       "Likely_benign,_other                                                                            1\n",
       "Name: clinical_severity, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "not_specified                                                                                                                                                                                     14400\n",
       "Not_Given                                                                                                                                                                                         12670\n",
       "Hereditary_cancer-predisposing_syndrome                                                                                                                                                            2951\n",
       "Limb-girdle_muscular_dystrophy,_type_2J|Dilated_cardiomyopathy_1G                                                                                                                                  2588\n",
       "Lynch_syndrome                                                                                                                                                                                     1717\n",
       "                                                                                                                                                                                                  ...  \n",
       "Abnormality_of_brain_morphology|Inborn_genetic_diseases|not_specified|not_provided                                                                                                                    1\n",
       "Polymicrogyria|Developmental_delay|Early_infantile_epileptic_encephalopathy_62|not_provided                                                                                                           1\n",
       "Spinocerebellar_ataxia_19|Cardiovascular_phenotype|not_provided                                                                                                                                       1\n",
       "Transitional_cell_carcinoma_of_the_bladder|Squamous_cell_lung_carcinoma|Neoplasm_of_the_large_intestine|Uterine_cervical_neoplasms|Malignant_neoplasm_of_body_of_uterus|Uterine_Carcinosarcoma        1\n",
       "Pitt-Hopkins-like_syndrome_2|not_specified|Pitt-Hopkins-like_syndrome|not_provided                                                                                                                    1\n",
       "Name: disease_implicated, Length: 6140, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Let's take a minute to look at value counts for every column.  One interesting thing is there are 14400 disease\n",
    "implicated that are not specified.\n",
    "\"\"\"\n",
    "display(df.gene_info.value_counts())\n",
    "display(df.id.value_counts())\n",
    "display(df.chrom.value_counts())\n",
    "display(df.pos.value_counts())\n",
    "display(df.ref.value_counts())\n",
    "display(df.alt.value_counts())\n",
    "display(df.clinical_severity.value_counts())\n",
    "display(df.disease_implicated.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check for missing data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_info             0\n",
       "id                    0\n",
       "chrom                 0\n",
       "pos                   0\n",
       "ref                   0\n",
       "alt                   0\n",
       "clinical_severity     0\n",
       "disease_implicated    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for NaNs\n",
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gene info missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation id missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position chromosomes missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position positions missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value references missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value alternate bases missing from DataFrame: 86 \n",
      "\n",
      "Number of clinical severities missing from DataFrame: 0 \n",
      "\n",
      "Number of disease implications missing from DataFrame: 0\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "There are still a few missing values in the DataFrame.  They are all in the alternative bases column.  \n",
    "I will replace the '.' with 'Not_Given'. \n",
    "\"\"\"\n",
    "print(\"Number of gene info missing from DataFrame:\", df[df.gene_info=='.']['gene_info'].count(), '\\n')\n",
    "print(\"Number of mutation id missing from DataFrame:\", df[df.id=='.']['id'].count(), '\\n')\n",
    "print(\"Number of mutation position chromosomes missing from DataFrame:\", df[df.chrom=='.']['chrom'].count(), '\\n')\n",
    "print(\"Number of mutation position positions missing from DataFrame:\", df[df.pos=='.']['pos'].count(), '\\n')\n",
    "print(\"Number of mutation value references missing from DataFrame:\", df[df.ref=='.']['ref'].count(), '\\n')\n",
    "print(\"Number of mutation value alternate bases missing from DataFrame:\", df[df.alt=='.']['alt'].count(), '\\n')\n",
    "print(\"Number of clinical severities missing from DataFrame:\", df[df.clinical_severity=='.']\\\n",
    "      ['clinical_severity'].count(), '\\n')\n",
    "print(\"Number of disease implications missing from DataFrame:\", df[df.disease_implicated=='.']\\\n",
    "      ['disease_implicated'].count())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Replace '.' with 'Not_Given'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinical_severity</th>\n",
       "      <th>disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1469</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>167375</td>\n",
       "      <td>1</td>\n",
       "      <td>5875102</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3068</td>\n",
       "      <td>MTHFR:4524</td>\n",
       "      <td>515933</td>\n",
       "      <td>1</td>\n",
       "      <td>11794400</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5004</td>\n",
       "      <td>DDOST:1650</td>\n",
       "      <td>540210</td>\n",
       "      <td>1</td>\n",
       "      <td>20661380</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Congenital_disorder_of_glycosylation_type_Ir</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5060</td>\n",
       "      <td>ALPL:249</td>\n",
       "      <td>512674</td>\n",
       "      <td>1</td>\n",
       "      <td>21563142</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>not_specified|not_provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6008</td>\n",
       "      <td>RHCE:6006</td>\n",
       "      <td>17708</td>\n",
       "      <td>1</td>\n",
       "      <td>25390874</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>RH_E/e_POLYMORPHISM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98256</td>\n",
       "      <td>CASR:846</td>\n",
       "      <td>237768</td>\n",
       "      <td>3</td>\n",
       "      <td>122284985</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Hypocalcemia,_autosomal_dominant_1|Hypocalciur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100200</td>\n",
       "      <td>BFSP2:8419|BFSP2-AS1:85003</td>\n",
       "      <td>474093</td>\n",
       "      <td>3</td>\n",
       "      <td>133448519</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Cataract_12,_multiple_types</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100545</td>\n",
       "      <td>FOXL2:668</td>\n",
       "      <td>180595</td>\n",
       "      <td>3</td>\n",
       "      <td>138946021</td>\n",
       "      <td>CGCGGCTGCAGCCGCAGCTGCTGCAGCCGCTGCGGCTGCCGC</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Blepharophimosis,_ptosis,_and_epicanthus_inversus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100652</td>\n",
       "      <td>RASA2:5922</td>\n",
       "      <td>496322</td>\n",
       "      <td>3</td>\n",
       "      <td>141608632</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>101126</td>\n",
       "      <td>CP:1356</td>\n",
       "      <td>463162</td>\n",
       "      <td>3</td>\n",
       "      <td>149198448</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Deficiency_of_ferroxidase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gene_info      id chrom        pos  \\\n",
       "1469                  NPHP4:261734  167375     1    5875102   \n",
       "3068                    MTHFR:4524  515933     1   11794400   \n",
       "5004                    DDOST:1650  540210     1   20661380   \n",
       "5060                      ALPL:249  512674     1   21563142   \n",
       "6008                     RHCE:6006   17708     1   25390874   \n",
       "...                            ...     ...   ...        ...   \n",
       "98256                     CASR:846  237768     3  122284985   \n",
       "100200  BFSP2:8419|BFSP2-AS1:85003  474093     3  133448519   \n",
       "100545                   FOXL2:668  180595     3  138946021   \n",
       "100652                  RASA2:5922  496322     3  141608632   \n",
       "101126                     CP:1356  463162     3  149198448   \n",
       "\n",
       "                                               ref alt     clinical_severity  \\\n",
       "1469                                             T   .                Benign   \n",
       "3068                                             G   .                Benign   \n",
       "5004                                             G   .                Benign   \n",
       "5060                                             T   .  Benign/Likely_benign   \n",
       "6008                                             C   .                Benign   \n",
       "...                                            ...  ..                   ...   \n",
       "98256                                            G   .                Benign   \n",
       "100200                                           G   .                Benign   \n",
       "100545  CGCGGCTGCAGCCGCAGCTGCTGCAGCCGCTGCGGCTGCCGC   .                Benign   \n",
       "100652                                           G   .                Benign   \n",
       "101126                                           T   .                Benign   \n",
       "\n",
       "                                       disease_implicated  \n",
       "1469                                        not_specified  \n",
       "3068                                        not_specified  \n",
       "5004         Congenital_disorder_of_glycosylation_type_Ir  \n",
       "5060                           not_specified|not_provided  \n",
       "6008                                  RH_E/e_POLYMORPHISM  \n",
       "...                                                   ...  \n",
       "98256   Hypocalcemia,_autosomal_dominant_1|Hypocalciur...  \n",
       "100200                        Cataract_12,_multiple_types  \n",
       "100545  Blepharophimosis,_ptosis,_and_epicanthus_inversus  \n",
       "100652                                          Not_Given  \n",
       "101126                          Deficiency_of_ferroxidase  \n",
       "\n",
       "[86 rows x 8 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.alt=='.']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Replace '.' with 'Not_Given'\n",
    "df['alt'].replace('.', 'Not_Given', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinical_severity</th>\n",
       "      <th>disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1469</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>167375</td>\n",
       "      <td>1</td>\n",
       "      <td>5875102</td>\n",
       "      <td>T</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3068</td>\n",
       "      <td>MTHFR:4524</td>\n",
       "      <td>515933</td>\n",
       "      <td>1</td>\n",
       "      <td>11794400</td>\n",
       "      <td>G</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5004</td>\n",
       "      <td>DDOST:1650</td>\n",
       "      <td>540210</td>\n",
       "      <td>1</td>\n",
       "      <td>20661380</td>\n",
       "      <td>G</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Congenital_disorder_of_glycosylation_type_Ir</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5060</td>\n",
       "      <td>ALPL:249</td>\n",
       "      <td>512674</td>\n",
       "      <td>1</td>\n",
       "      <td>21563142</td>\n",
       "      <td>T</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>not_specified|not_provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6008</td>\n",
       "      <td>RHCE:6006</td>\n",
       "      <td>17708</td>\n",
       "      <td>1</td>\n",
       "      <td>25390874</td>\n",
       "      <td>C</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>RH_E/e_POLYMORPHISM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98256</td>\n",
       "      <td>CASR:846</td>\n",
       "      <td>237768</td>\n",
       "      <td>3</td>\n",
       "      <td>122284985</td>\n",
       "      <td>G</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Hypocalcemia,_autosomal_dominant_1|Hypocalciur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100200</td>\n",
       "      <td>BFSP2:8419|BFSP2-AS1:85003</td>\n",
       "      <td>474093</td>\n",
       "      <td>3</td>\n",
       "      <td>133448519</td>\n",
       "      <td>G</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Cataract_12,_multiple_types</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100545</td>\n",
       "      <td>FOXL2:668</td>\n",
       "      <td>180595</td>\n",
       "      <td>3</td>\n",
       "      <td>138946021</td>\n",
       "      <td>CGCGGCTGCAGCCGCAGCTGCTGCAGCCGCTGCGGCTGCCGC</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Blepharophimosis,_ptosis,_and_epicanthus_inversus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100652</td>\n",
       "      <td>RASA2:5922</td>\n",
       "      <td>496322</td>\n",
       "      <td>3</td>\n",
       "      <td>141608632</td>\n",
       "      <td>G</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>101126</td>\n",
       "      <td>CP:1356</td>\n",
       "      <td>463162</td>\n",
       "      <td>3</td>\n",
       "      <td>149198448</td>\n",
       "      <td>T</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Deficiency_of_ferroxidase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gene_info      id chrom        pos  \\\n",
       "1469                  NPHP4:261734  167375     1    5875102   \n",
       "3068                    MTHFR:4524  515933     1   11794400   \n",
       "5004                    DDOST:1650  540210     1   20661380   \n",
       "5060                      ALPL:249  512674     1   21563142   \n",
       "6008                     RHCE:6006   17708     1   25390874   \n",
       "...                            ...     ...   ...        ...   \n",
       "98256                     CASR:846  237768     3  122284985   \n",
       "100200  BFSP2:8419|BFSP2-AS1:85003  474093     3  133448519   \n",
       "100545                   FOXL2:668  180595     3  138946021   \n",
       "100652                  RASA2:5922  496322     3  141608632   \n",
       "101126                     CP:1356  463162     3  149198448   \n",
       "\n",
       "                                               ref        alt  \\\n",
       "1469                                             T  Not_Given   \n",
       "3068                                             G  Not_Given   \n",
       "5004                                             G  Not_Given   \n",
       "5060                                             T  Not_Given   \n",
       "6008                                             C  Not_Given   \n",
       "...                                            ...        ...   \n",
       "98256                                            G  Not_Given   \n",
       "100200                                           G  Not_Given   \n",
       "100545  CGCGGCTGCAGCCGCAGCTGCTGCAGCCGCTGCGGCTGCCGC  Not_Given   \n",
       "100652                                           G  Not_Given   \n",
       "101126                                           T  Not_Given   \n",
       "\n",
       "           clinical_severity  \\\n",
       "1469                  Benign   \n",
       "3068                  Benign   \n",
       "5004                  Benign   \n",
       "5060    Benign/Likely_benign   \n",
       "6008                  Benign   \n",
       "...                      ...   \n",
       "98256                 Benign   \n",
       "100200                Benign   \n",
       "100545                Benign   \n",
       "100652                Benign   \n",
       "101126                Benign   \n",
       "\n",
       "                                       disease_implicated  \n",
       "1469                                        not_specified  \n",
       "3068                                        not_specified  \n",
       "5004         Congenital_disorder_of_glycosylation_type_Ir  \n",
       "5060                           not_specified|not_provided  \n",
       "6008                                  RH_E/e_POLYMORPHISM  \n",
       "...                                                   ...  \n",
       "98256   Hypocalcemia,_autosomal_dominant_1|Hypocalciur...  \n",
       "100200                        Cataract_12,_multiple_types  \n",
       "100545  Blepharophimosis,_ptosis,_and_epicanthus_inversus  \n",
       "100652                                          Not_Given  \n",
       "101126                          Deficiency_of_ferroxidase  \n",
       "\n",
       "[86 rows x 8 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.alt=='Not_Given']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gene info missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation id missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position chromosomes missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position positions missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value references missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value alternate bases missing from DataFrame: 0 \n",
      "\n",
      "Number of clinical severities missing from DataFrame: 0 \n",
      "\n",
      "Number of disease implications missing from DataFrame: 0\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "One last check that there are no data with '.'.  \n",
    "\"\"\"\n",
    "print(\"Number of gene info missing from DataFrame:\", df[df.gene_info=='.']['gene_info'].count(), '\\n')\n",
    "print(\"Number of mutation id missing from DataFrame:\", df[df.id=='.']['id'].count(), '\\n')\n",
    "print(\"Number of mutation position chromosomes missing from DataFrame:\", df[df.chrom=='.']['chrom'].count(), '\\n')\n",
    "print(\"Number of mutation position positions missing from DataFrame:\", df[df.pos=='.']['pos'].count(), '\\n')\n",
    "print(\"Number of mutation value references missing from DataFrame:\", df[df.ref=='.']['ref'].count(), '\\n')\n",
    "print(\"Number of mutation value alternate bases missing from DataFrame:\", df[df.alt=='.']['alt'].count(), '\\n')\n",
    "print(\"Number of clinical severities missing from DataFrame:\", df[df.clinical_severity=='.']\\\n",
    "      ['clinical_severity'].count(), '\\n')\n",
    "print(\"Number of disease implications missing from DataFrame:\", df[df.disease_implicated=='.']\\\n",
    "      ['disease_implicated'].count())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How much data is 'Not_Given'?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gene info missing from DataFrame: 18 \n",
      "\n",
      "Number of mutation id missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position chromosomes missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position positions missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value references missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value alternate bases missing from DataFrame: 86 \n",
      "\n",
      "Number of clinical severities missing from DataFrame: 1797 \n",
      "\n",
      "Number of disease implications missing from DataFrame: 12670\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of gene info missing from DataFrame:\", df[df.gene_info=='Not_Given']['gene_info'].count(), '\\n')\n",
    "print(\"Number of mutation id missing from DataFrame:\", df[df.id=='Not_Given']['id'].count(), '\\n')\n",
    "print(\"Number of mutation position chromosomes missing from DataFrame:\", df[df.chrom=='Not_Given']['chrom'].count(), '\\n')\n",
    "print(\"Number of mutation position positions missing from DataFrame:\", df[df.pos=='Not_Given']['pos'].count(), '\\n')\n",
    "print(\"Number of mutation value references missing from DataFrame:\", df[df.ref=='Not_Given']['ref'].count(), '\\n')\n",
    "print(\"Number of mutation value alternate bases missing from DataFrame:\", df[df.alt=='Not_Given']['alt'].count(), '\\n')\n",
    "print(\"Number of clinical severities missing from DataFrame:\", df[df.clinical_severity=='Not_Given']\\\n",
    "      ['clinical_severity'].count(), '\\n')\n",
    "print(\"Number of disease implications missing from DataFrame:\", df[df.disease_implicated=='Not_Given']\\\n",
    "      ['disease_implicated'].count())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12330, 8)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "After examining the value counts of clincal severity I decided to group all harmful cases by 'Pathogenic' or \n",
    "'Pathogenic,'.  I df_all_harmful, gene and mutation combinations are listed.  \n",
    "\"\"\"\n",
    "# df_dangerous = df[df['clinical_severity']=='Pathogenic']\n",
    "df_all_harmful = df[(df['clinical_severity']=='Pathogenic') | (df['clinical_severity'].str.contains('Pathogenic,'))]\n",
    "df_all_harmful.reset_index(inplace=True, drop=True)\n",
    "df_all_harmful.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinical_severity</th>\n",
       "      <th>disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161455</td>\n",
       "      <td>1</td>\n",
       "      <td>1014316</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161454</td>\n",
       "      <td>1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243036</td>\n",
       "      <td>1</td>\n",
       "      <td>1022225</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243037</td>\n",
       "      <td>1</td>\n",
       "      <td>1022313</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>574478</td>\n",
       "      <td>1</td>\n",
       "      <td>1041354</td>\n",
       "      <td>CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>126556</td>\n",
       "      <td>1</td>\n",
       "      <td>1041582</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243038</td>\n",
       "      <td>1</td>\n",
       "      <td>1042136</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243039</td>\n",
       "      <td>1</td>\n",
       "      <td>1050473</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>18241</td>\n",
       "      <td>1</td>\n",
       "      <td>1050575</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>126555</td>\n",
       "      <td>1</td>\n",
       "      <td>1050763</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243040</td>\n",
       "      <td>1</td>\n",
       "      <td>1051775</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>TNFRSF4:7293</td>\n",
       "      <td>96692</td>\n",
       "      <td>1</td>\n",
       "      <td>1213738</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60484</td>\n",
       "      <td>1</td>\n",
       "      <td>1232279</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60493</td>\n",
       "      <td>1</td>\n",
       "      <td>1232294</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60488</td>\n",
       "      <td>1</td>\n",
       "      <td>1232471</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60489</td>\n",
       "      <td>1</td>\n",
       "      <td>1232478</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60497</td>\n",
       "      <td>1</td>\n",
       "      <td>1232594</td>\n",
       "      <td>CGGCGCGCCCTGGAGCGGGAGCA</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60490</td>\n",
       "      <td>1</td>\n",
       "      <td>1232630</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60494</td>\n",
       "      <td>1</td>\n",
       "      <td>1232684</td>\n",
       "      <td>GTGCTGGCCA</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_info      id chrom      pos  \\\n",
       "0       ISG15:9636  183381     1  1014143   \n",
       "1       ISG15:9636  161455     1  1014316   \n",
       "2       ISG15:9636  161454     1  1014359   \n",
       "3      AGRN:375790  243036     1  1022225   \n",
       "4      AGRN:375790  243037     1  1022313   \n",
       "5      AGRN:375790  574478     1  1041354   \n",
       "6      AGRN:375790  126556     1  1041582   \n",
       "7      AGRN:375790  243038     1  1042136   \n",
       "8      AGRN:375790  243039     1  1050473   \n",
       "9      AGRN:375790   18241     1  1050575   \n",
       "10     AGRN:375790  126555     1  1050763   \n",
       "11     AGRN:375790  243040     1  1051775   \n",
       "12    TNFRSF4:7293   96692     1  1213738   \n",
       "13  B3GALT6:126792   60484     1  1232279   \n",
       "14  B3GALT6:126792   60493     1  1232294   \n",
       "15  B3GALT6:126792   60488     1  1232471   \n",
       "16  B3GALT6:126792   60489     1  1232478   \n",
       "17  B3GALT6:126792   60497     1  1232594   \n",
       "18  B3GALT6:126792   60490     1  1232630   \n",
       "19  B3GALT6:126792   60494     1  1232684   \n",
       "\n",
       "                                    ref alt clinical_severity  \\\n",
       "0                                     C   T        Pathogenic   \n",
       "1                                     C  CG        Pathogenic   \n",
       "2                                     G   T        Pathogenic   \n",
       "3                                     G   A        Pathogenic   \n",
       "4                                     A   T        Pathogenic   \n",
       "5   CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG   C        Pathogenic   \n",
       "6                                     C   T        Pathogenic   \n",
       "7                                     T  TC        Pathogenic   \n",
       "8                                     G   A        Pathogenic   \n",
       "9                                     G   C        Pathogenic   \n",
       "10                                    G   T        Pathogenic   \n",
       "11                                    G   A        Pathogenic   \n",
       "12                                    G   A        Pathogenic   \n",
       "13                                    A   G        Pathogenic   \n",
       "14                                    C   T        Pathogenic   \n",
       "15                                    A   G        Pathogenic   \n",
       "16                                    C   T        Pathogenic   \n",
       "17              CGGCGCGCCCTGGAGCGGGAGCA   C        Pathogenic   \n",
       "18                                   GA   G        Pathogenic   \n",
       "19                           GTGCTGGCCA   G        Pathogenic   \n",
       "\n",
       "                                   disease_implicated  \n",
       "0   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2   Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3                      Congenital_myasthenic_syndrome  \n",
       "4                      Congenital_myasthenic_syndrome  \n",
       "5                  Myasthenic_syndrome,_congenital,_8  \n",
       "6   Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "7                      Congenital_myasthenic_syndrome  \n",
       "8                      Congenital_myasthenic_syndrome  \n",
       "9   Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "10  Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "11                     Congenital_myasthenic_syndrome  \n",
       "12                                Immunodeficiency_16  \n",
       "13  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "14          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "15  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "16  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "17  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "18          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "19          Ehlers-Danlos_syndrome,_progeroid_type,_2  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_harmful.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gene info missing from DataFrame: 6 \n",
      "\n",
      "Number of mutation id missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position chromosomes missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation position positions missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value references missing from DataFrame: 0 \n",
      "\n",
      "Number of mutation value alternate bases missing from DataFrame: 3 \n",
      "\n",
      "Number of clinical severities missing from DataFrame: 0 \n",
      "\n",
      "Number of disease implications missing from DataFrame: 1890\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of gene info missing from DataFrame:\", df_all_harmful[df_all_harmful.gene_info=='Not_Given']['gene_info'].count(), '\\n')\n",
    "print(\"Number of mutation id missing from DataFrame:\", df_all_harmful[df_all_harmful.id=='Not_Given']['id'].count(), '\\n')\n",
    "print(\"Number of mutation position chromosomes missing from DataFrame:\", df_all_harmful[df_all_harmful.chrom=='Not_Given']['chrom'].count(), '\\n')\n",
    "print(\"Number of mutation position positions missing from DataFrame:\", df_all_harmful[df_all_harmful.pos=='Not_Given']['pos'].count(), '\\n')\n",
    "print(\"Number of mutation value references missing from DataFrame:\", df_all_harmful[df_all_harmful.ref=='Not_Given']['ref'].count(), '\\n')\n",
    "print(\"Number of mutation value alternate bases missing from DataFrame:\", df_all_harmful[df_all_harmful.alt=='Not_Given']['alt'].count(), '\\n')\n",
    "print(\"Number of clinical severities missing from DataFrame:\", df_all_harmful[df_all_harmful.clinical_severity=='Not_Given']\\\n",
    "      ['clinical_severity'].count(), '\\n')\n",
    "print(\"Number of disease implications missing from DataFrame:\", df_all_harmful[df_all_harmful.disease_implicated=='Not_Given']\\\n",
    "      ['disease_implicated'].count())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinical_severity</th>\n",
       "      <th>disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161455</td>\n",
       "      <td>1</td>\n",
       "      <td>1014316</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161454</td>\n",
       "      <td>1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243036</td>\n",
       "      <td>1</td>\n",
       "      <td>1022225</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243037</td>\n",
       "      <td>1</td>\n",
       "      <td>1022313</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12324</td>\n",
       "      <td>PIK3CA:5290</td>\n",
       "      <td>156449</td>\n",
       "      <td>3</td>\n",
       "      <td>179204509</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Cowden_syndrome_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12326</td>\n",
       "      <td>PIK3CA:5290</td>\n",
       "      <td>39704</td>\n",
       "      <td>3</td>\n",
       "      <td>179204576</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Megalencephaly_cutis_marmorata_telangiectatica...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12327</td>\n",
       "      <td>PIK3CA:5290</td>\n",
       "      <td>156450</td>\n",
       "      <td>3</td>\n",
       "      <td>179204588</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Cowden_syndrome_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12328</td>\n",
       "      <td>PIK3CA:5290</td>\n",
       "      <td>31945</td>\n",
       "      <td>3</td>\n",
       "      <td>179210192</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>CAPILLARY_MALFORMATION_OF_THE_LOWER_LIP,_LYMPH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12329</td>\n",
       "      <td>PIK3CA:5290</td>\n",
       "      <td>39706</td>\n",
       "      <td>3</td>\n",
       "      <td>179210289</td>\n",
       "      <td>TAGA</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Megalencephaly_cutis_marmorata_telangiectatica...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10432 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_info      id chrom        pos   ref alt clinical_severity  \\\n",
       "0       ISG15:9636  183381     1    1014143     C   T        Pathogenic   \n",
       "1       ISG15:9636  161455     1    1014316     C  CG        Pathogenic   \n",
       "2       ISG15:9636  161454     1    1014359     G   T        Pathogenic   \n",
       "3      AGRN:375790  243036     1    1022225     G   A        Pathogenic   \n",
       "4      AGRN:375790  243037     1    1022313     A   T        Pathogenic   \n",
       "...            ...     ...   ...        ...   ...  ..               ...   \n",
       "12324  PIK3CA:5290  156449     3  179204509     G   A        Pathogenic   \n",
       "12326  PIK3CA:5290   39704     3  179204576     G   A        Pathogenic   \n",
       "12327  PIK3CA:5290  156450     3  179204588     G   A        Pathogenic   \n",
       "12328  PIK3CA:5290   31945     3  179210192     T   C        Pathogenic   \n",
       "12329  PIK3CA:5290   39706     3  179210289  TAGA   T        Pathogenic   \n",
       "\n",
       "                                      disease_implicated  \n",
       "0      Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1      Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2      Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3                         Congenital_myasthenic_syndrome  \n",
       "4                         Congenital_myasthenic_syndrome  \n",
       "...                                                  ...  \n",
       "12324                                  Cowden_syndrome_5  \n",
       "12326  Megalencephaly_cutis_marmorata_telangiectatica...  \n",
       "12327                                  Cowden_syndrome_5  \n",
       "12328  CAPILLARY_MALFORMATION_OF_THE_LOWER_LIP,_LYMPH...  \n",
       "12329  Megalencephaly_cutis_marmorata_telangiectatica...  \n",
       "\n",
       "[10432 rows x 8 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "There are currently 1,898 harmful gene and mutation combinations that contain missing data within df_all_harmful.\n",
    "I will filter out all the rows with missing data and put the in df_harmful.\n",
    "\"\"\"\n",
    "df_harmful = df_all_harmful[(df_all_harmful.gene_info != 'Not_Given')&\\\n",
    "                       (df_all_harmful.alt != 'Not_Given')&\\\n",
    "                       (df_all_harmful.disease_implicated != 'Not_Given')]\n",
    "df_harmful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((12330, 8), (10432, 8))"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_harmful.shape, df_harmful.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove all rows that have some form of data not given."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df_harmful.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91228     1\n",
       "425954    1\n",
       "90091     1\n",
       "90327     1\n",
       "14826     1\n",
       "         ..\n",
       "544132    1\n",
       "161334    1\n",
       "94347     1\n",
       "374842    1\n",
       "590783    1\n",
       "Name: id, Length: 10432, dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_harmful.id.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List of 100 harmful gene and mutation variations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinical_severity</th>\n",
       "      <th>disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161455</td>\n",
       "      <td>1</td>\n",
       "      <td>1014316</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161454</td>\n",
       "      <td>1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243036</td>\n",
       "      <td>1</td>\n",
       "      <td>1022225</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243037</td>\n",
       "      <td>1</td>\n",
       "      <td>1022313</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>574478</td>\n",
       "      <td>1</td>\n",
       "      <td>1041354</td>\n",
       "      <td>CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>126556</td>\n",
       "      <td>1</td>\n",
       "      <td>1041582</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243038</td>\n",
       "      <td>1</td>\n",
       "      <td>1042136</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243039</td>\n",
       "      <td>1</td>\n",
       "      <td>1050473</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>18241</td>\n",
       "      <td>1</td>\n",
       "      <td>1050575</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>126555</td>\n",
       "      <td>1</td>\n",
       "      <td>1050763</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome|Myasthenic_synd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243040</td>\n",
       "      <td>1</td>\n",
       "      <td>1051775</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>TNFRSF4:7293</td>\n",
       "      <td>96692</td>\n",
       "      <td>1</td>\n",
       "      <td>1213738</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60484</td>\n",
       "      <td>1</td>\n",
       "      <td>1232279</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60493</td>\n",
       "      <td>1</td>\n",
       "      <td>1232294</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60488</td>\n",
       "      <td>1</td>\n",
       "      <td>1232471</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60489</td>\n",
       "      <td>1</td>\n",
       "      <td>1232478</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60497</td>\n",
       "      <td>1</td>\n",
       "      <td>1232594</td>\n",
       "      <td>CGGCGCGCCCTGGAGCGGGAGCA</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60490</td>\n",
       "      <td>1</td>\n",
       "      <td>1232630</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60494</td>\n",
       "      <td>1</td>\n",
       "      <td>1232684</td>\n",
       "      <td>GTGCTGGCCA</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60486</td>\n",
       "      <td>1</td>\n",
       "      <td>1232744</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>522415</td>\n",
       "      <td>1</td>\n",
       "      <td>1232834</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>22</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60492</td>\n",
       "      <td>1</td>\n",
       "      <td>1232859</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>23</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60495</td>\n",
       "      <td>1</td>\n",
       "      <td>1232897</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60496</td>\n",
       "      <td>1</td>\n",
       "      <td>1232927</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>25</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60485</td>\n",
       "      <td>1</td>\n",
       "      <td>1232972</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60487</td>\n",
       "      <td>1</td>\n",
       "      <td>1233177</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spondyloepimetaphyseal_dysplasia_with_joint_la...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27</td>\n",
       "      <td>B3GALT6:126792</td>\n",
       "      <td>60491</td>\n",
       "      <td>1</td>\n",
       "      <td>1233203</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_progeroid_type,_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>28</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208047</td>\n",
       "      <td>1</td>\n",
       "      <td>1338000</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>29</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208050</td>\n",
       "      <td>1</td>\n",
       "      <td>1338033</td>\n",
       "      <td>TAGGCAGG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208043</td>\n",
       "      <td>1</td>\n",
       "      <td>1338045</td>\n",
       "      <td>AA</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208049</td>\n",
       "      <td>1</td>\n",
       "      <td>1338053</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>32</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208048</td>\n",
       "      <td>1</td>\n",
       "      <td>1338086</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>219223</td>\n",
       "      <td>1</td>\n",
       "      <td>1338092</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>34</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208045</td>\n",
       "      <td>1</td>\n",
       "      <td>1338096</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208044</td>\n",
       "      <td>1</td>\n",
       "      <td>1338098</td>\n",
       "      <td>GGGGGCAGCCGGGT</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208046</td>\n",
       "      <td>1</td>\n",
       "      <td>1338107</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Robinow_syndrome,_autosomal_dominant_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>37</td>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>520999</td>\n",
       "      <td>1</td>\n",
       "      <td>1338111</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>38</td>\n",
       "      <td>ATAD3A:55210</td>\n",
       "      <td>522903</td>\n",
       "      <td>1</td>\n",
       "      <td>1534004</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Harel-Yoon_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>39</td>\n",
       "      <td>TMEM240:339453</td>\n",
       "      <td>161196</td>\n",
       "      <td>1</td>\n",
       "      <td>1535370</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spinocerebellar_ataxia_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40</td>\n",
       "      <td>TMEM240:339453</td>\n",
       "      <td>161193</td>\n",
       "      <td>1</td>\n",
       "      <td>1535392</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spinocerebellar_ataxia_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41</td>\n",
       "      <td>TMEM240:339453</td>\n",
       "      <td>161194</td>\n",
       "      <td>1</td>\n",
       "      <td>1535616</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spinocerebellar_ataxia_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>42</td>\n",
       "      <td>TMEM240:339453</td>\n",
       "      <td>523023</td>\n",
       "      <td>1</td>\n",
       "      <td>1535619</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spinocerebellar_ataxia_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>43</td>\n",
       "      <td>TMEM240:339453</td>\n",
       "      <td>161195</td>\n",
       "      <td>1</td>\n",
       "      <td>1535723</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spinocerebellar_ataxia_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>44</td>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>521025</td>\n",
       "      <td>1</td>\n",
       "      <td>1787340</td>\n",
       "      <td>GATC</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>45</td>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>521213</td>\n",
       "      <td>1</td>\n",
       "      <td>1804496</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46</td>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>208722</td>\n",
       "      <td>1</td>\n",
       "      <td>1806503</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Neurodevelopmental_Disability|hypotonia|Cleft_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47</td>\n",
       "      <td>GNB1:2782</td>\n",
       "      <td>224714</td>\n",
       "      <td>1</td>\n",
       "      <td>1806509</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Neurodevelopmental_Disability|hypotonia|Strabi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>48</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>224869</td>\n",
       "      <td>1</td>\n",
       "      <td>2228828</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37259</td>\n",
       "      <td>1</td>\n",
       "      <td>2228860</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>50</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37262</td>\n",
       "      <td>1</td>\n",
       "      <td>2228866</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases|Shprintzen-Goldberg_sy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37260</td>\n",
       "      <td>1</td>\n",
       "      <td>2228867</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>52</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>39783</td>\n",
       "      <td>1</td>\n",
       "      <td>2228867</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>53</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>39785</td>\n",
       "      <td>1</td>\n",
       "      <td>2228870</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37263</td>\n",
       "      <td>1</td>\n",
       "      <td>2229045</td>\n",
       "      <td>GTCCGACCGC</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>55</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>39784</td>\n",
       "      <td>1</td>\n",
       "      <td>2229045</td>\n",
       "      <td>GTCCGACCGCTCC</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>56</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>37258</td>\n",
       "      <td>1</td>\n",
       "      <td>2229113</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>57</td>\n",
       "      <td>SKI:6497</td>\n",
       "      <td>463409</td>\n",
       "      <td>1</td>\n",
       "      <td>2229401</td>\n",
       "      <td>AGCTCAGCGAGCGCAGCGTCCGCGTGTACCACGAGTGCTTCGGCAA...</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>58</td>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>6771</td>\n",
       "      <td>1</td>\n",
       "      <td>2406526</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>59</td>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>296273</td>\n",
       "      <td>1</td>\n",
       "      <td>2406580</td>\n",
       "      <td>CAG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Zellweger_syndrome|Peroxisome_biogenesis_disor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60</td>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>225040</td>\n",
       "      <td>1</td>\n",
       "      <td>2406719</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>61</td>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>162435</td>\n",
       "      <td>1</td>\n",
       "      <td>2406766</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>62</td>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>6774</td>\n",
       "      <td>1</td>\n",
       "      <td>2406791</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder,_complementatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63</td>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>6770</td>\n",
       "      <td>1</td>\n",
       "      <td>2408451</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>64</td>\n",
       "      <td>PEX10:5192</td>\n",
       "      <td>162431</td>\n",
       "      <td>1</td>\n",
       "      <td>2408714</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>65</td>\n",
       "      <td>PEX10:5192|PLCH2:9651</td>\n",
       "      <td>162434</td>\n",
       "      <td>1</td>\n",
       "      <td>2412501</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>66</td>\n",
       "      <td>PRDM16:63976</td>\n",
       "      <td>60728</td>\n",
       "      <td>1</td>\n",
       "      <td>3402986</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Dilated_cardiomyopathy_1LL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>67</td>\n",
       "      <td>PRDM16:63976</td>\n",
       "      <td>60724</td>\n",
       "      <td>1</td>\n",
       "      <td>3412301</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Left_ventricular_noncompaction_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>68</td>\n",
       "      <td>PRDM16:63976</td>\n",
       "      <td>60726</td>\n",
       "      <td>1</td>\n",
       "      <td>3412644</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Left_ventricular_noncompaction_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>69</td>\n",
       "      <td>PRDM16:63976</td>\n",
       "      <td>60729</td>\n",
       "      <td>1</td>\n",
       "      <td>3414616</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Dilated_cardiomyopathy_1LL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>70</td>\n",
       "      <td>CEP104:9731</td>\n",
       "      <td>221277</td>\n",
       "      <td>1</td>\n",
       "      <td>3816372</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Joubert_syndrome_25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>71</td>\n",
       "      <td>CEP104:9731</td>\n",
       "      <td>221275</td>\n",
       "      <td>1</td>\n",
       "      <td>3835081</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Joubert_syndrome_25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>72</td>\n",
       "      <td>CEP104:9731</td>\n",
       "      <td>542193</td>\n",
       "      <td>1</td>\n",
       "      <td>3837319</td>\n",
       "      <td>AGGGAGTAACG</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Joubert_syndrome_25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73</td>\n",
       "      <td>CEP104:9731</td>\n",
       "      <td>542192</td>\n",
       "      <td>1</td>\n",
       "      <td>3839096</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Joubert_syndrome_25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>74</td>\n",
       "      <td>CEP104:9731</td>\n",
       "      <td>221274</td>\n",
       "      <td>1</td>\n",
       "      <td>3839606</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Joubert_syndrome_25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>75</td>\n",
       "      <td>CEP104:9731</td>\n",
       "      <td>221276</td>\n",
       "      <td>1</td>\n",
       "      <td>3844977</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Joubert_syndrome_25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>76</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>3403</td>\n",
       "      <td>1</td>\n",
       "      <td>5873294</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Nephronophthisis_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>77</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>3402</td>\n",
       "      <td>1</td>\n",
       "      <td>5874946</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Nephronophthisis_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>3400</td>\n",
       "      <td>1</td>\n",
       "      <td>5887394</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Nephronophthisis|Nephronophthisis_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>79</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>3399</td>\n",
       "      <td>1</td>\n",
       "      <td>5887403</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Nephronophthisis_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>80</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>3404</td>\n",
       "      <td>1</td>\n",
       "      <td>5887436</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Nephronophthisis|Senior-Loken_syndrome_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>81</td>\n",
       "      <td>NPHP4:261734</td>\n",
       "      <td>3405</td>\n",
       "      <td>1</td>\n",
       "      <td>5904788</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Nephronophthisis|Senior-Loken_syndrome_4|Nephr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>82</td>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>4420</td>\n",
       "      <td>1</td>\n",
       "      <td>6451926</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deafness,_without_vestibular_involvement,_auto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>83</td>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>4422</td>\n",
       "      <td>1</td>\n",
       "      <td>6452092</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deafness,_without_vestibular_involvement,_auto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>84</td>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>560535</td>\n",
       "      <td>1</td>\n",
       "      <td>6457223</td>\n",
       "      <td>TGGAGGCGGGACCTCCTGC</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Usher_syndrome,_type_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>85</td>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>562076</td>\n",
       "      <td>1</td>\n",
       "      <td>6460076</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deafness,_autosomal_recessive_36,_with_or_with...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>86</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>566096</td>\n",
       "      <td>1</td>\n",
       "      <td>6468294</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Distal_spinal_muscular_atrophy,_autosomal_rece...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>87</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>60779</td>\n",
       "      <td>1</td>\n",
       "      <td>6468378</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Charcot-Marie-Tooth_disease,_recessive_interme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>88</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>468906</td>\n",
       "      <td>1</td>\n",
       "      <td>6468468</td>\n",
       "      <td>TGA</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Distal_spinal_muscular_atrophy,_autosomal_rece...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>89</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>60780</td>\n",
       "      <td>1</td>\n",
       "      <td>6469396</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Charcot-Marie-Tooth_disease,_recessive_interme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>90</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>1019</td>\n",
       "      <td>1</td>\n",
       "      <td>6469444</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Distal_spinal_muscular_atrophy,_autosomal_rece...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>91</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>60778</td>\n",
       "      <td>1</td>\n",
       "      <td>6473051</td>\n",
       "      <td>C</td>\n",
       "      <td>CGTCTTCA</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Charcot-Marie-Tooth_disease,_recessive_interme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>92</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>549691</td>\n",
       "      <td>1</td>\n",
       "      <td>6476021</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Spinal_muscular_atrophy,_facioscapulohumeral_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>93</td>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>60777</td>\n",
       "      <td>1</td>\n",
       "      <td>6477533</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Charcot-Marie-Tooth_disease,_recessive_interme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>94</td>\n",
       "      <td>CAMTA1:23261</td>\n",
       "      <td>521550</td>\n",
       "      <td>1</td>\n",
       "      <td>7249607</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Inborn_genetic_diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>95</td>\n",
       "      <td>PER3:8863</td>\n",
       "      <td>242410</td>\n",
       "      <td>1</td>\n",
       "      <td>7809893</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Advanced_sleep_phase_syndrome,_familial,_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>96</td>\n",
       "      <td>PER3:8863</td>\n",
       "      <td>242411</td>\n",
       "      <td>1</td>\n",
       "      <td>7809900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Advanced_sleep_phase_syndrome,_familial,_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97</td>\n",
       "      <td>PARK7:11315</td>\n",
       "      <td>7065</td>\n",
       "      <td>1</td>\n",
       "      <td>7962863</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Parkinson_disease_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98</td>\n",
       "      <td>PARK7:11315</td>\n",
       "      <td>573287</td>\n",
       "      <td>1</td>\n",
       "      <td>7965336</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Parkinson_disease_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>99</td>\n",
       "      <td>PARK7:11315</td>\n",
       "      <td>7069</td>\n",
       "      <td>1</td>\n",
       "      <td>7965348</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Parkinson_disease,_autosomal_recessive_early-o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_info      id chrom      pos  \\\n",
       "0              ISG15:9636  183381     1  1014143   \n",
       "1              ISG15:9636  161455     1  1014316   \n",
       "2              ISG15:9636  161454     1  1014359   \n",
       "3             AGRN:375790  243036     1  1022225   \n",
       "4             AGRN:375790  243037     1  1022313   \n",
       "5             AGRN:375790  574478     1  1041354   \n",
       "6             AGRN:375790  126556     1  1041582   \n",
       "7             AGRN:375790  243038     1  1042136   \n",
       "8             AGRN:375790  243039     1  1050473   \n",
       "9             AGRN:375790   18241     1  1050575   \n",
       "10            AGRN:375790  126555     1  1050763   \n",
       "11            AGRN:375790  243040     1  1051775   \n",
       "12           TNFRSF4:7293   96692     1  1213738   \n",
       "13         B3GALT6:126792   60484     1  1232279   \n",
       "14         B3GALT6:126792   60493     1  1232294   \n",
       "15         B3GALT6:126792   60488     1  1232471   \n",
       "16         B3GALT6:126792   60489     1  1232478   \n",
       "17         B3GALT6:126792   60497     1  1232594   \n",
       "18         B3GALT6:126792   60490     1  1232630   \n",
       "19         B3GALT6:126792   60494     1  1232684   \n",
       "20         B3GALT6:126792   60486     1  1232744   \n",
       "21         B3GALT6:126792  522415     1  1232834   \n",
       "22         B3GALT6:126792   60492     1  1232859   \n",
       "23         B3GALT6:126792   60495     1  1232897   \n",
       "24         B3GALT6:126792   60496     1  1232927   \n",
       "25         B3GALT6:126792   60485     1  1232972   \n",
       "26         B3GALT6:126792   60487     1  1233177   \n",
       "27         B3GALT6:126792   60491     1  1233203   \n",
       "28              DVL1:1855  208047     1  1338000   \n",
       "29              DVL1:1855  208050     1  1338033   \n",
       "30              DVL1:1855  208043     1  1338045   \n",
       "31              DVL1:1855  208049     1  1338053   \n",
       "32              DVL1:1855  208048     1  1338086   \n",
       "33              DVL1:1855  219223     1  1338092   \n",
       "34              DVL1:1855  208045     1  1338096   \n",
       "35              DVL1:1855  208044     1  1338098   \n",
       "36              DVL1:1855  208046     1  1338107   \n",
       "37              DVL1:1855  520999     1  1338111   \n",
       "38           ATAD3A:55210  522903     1  1534004   \n",
       "39         TMEM240:339453  161196     1  1535370   \n",
       "40         TMEM240:339453  161193     1  1535392   \n",
       "41         TMEM240:339453  161194     1  1535616   \n",
       "42         TMEM240:339453  523023     1  1535619   \n",
       "43         TMEM240:339453  161195     1  1535723   \n",
       "44              GNB1:2782  521025     1  1787340   \n",
       "45              GNB1:2782  521213     1  1804496   \n",
       "46              GNB1:2782  208722     1  1806503   \n",
       "47              GNB1:2782  224714     1  1806509   \n",
       "48               SKI:6497  224869     1  2228828   \n",
       "49               SKI:6497   37259     1  2228860   \n",
       "50               SKI:6497   37262     1  2228866   \n",
       "51               SKI:6497   37260     1  2228867   \n",
       "52               SKI:6497   39783     1  2228867   \n",
       "53               SKI:6497   39785     1  2228870   \n",
       "54               SKI:6497   37263     1  2229045   \n",
       "55               SKI:6497   39784     1  2229045   \n",
       "56               SKI:6497   37258     1  2229113   \n",
       "57               SKI:6497  463409     1  2229401   \n",
       "58             PEX10:5192    6771     1  2406526   \n",
       "59             PEX10:5192  296273     1  2406580   \n",
       "60             PEX10:5192  225040     1  2406719   \n",
       "61             PEX10:5192  162435     1  2406766   \n",
       "62             PEX10:5192    6774     1  2406791   \n",
       "63             PEX10:5192    6770     1  2408451   \n",
       "64             PEX10:5192  162431     1  2408714   \n",
       "65  PEX10:5192|PLCH2:9651  162434     1  2412501   \n",
       "66           PRDM16:63976   60728     1  3402986   \n",
       "67           PRDM16:63976   60724     1  3412301   \n",
       "68           PRDM16:63976   60726     1  3412644   \n",
       "69           PRDM16:63976   60729     1  3414616   \n",
       "70            CEP104:9731  221277     1  3816372   \n",
       "71            CEP104:9731  221275     1  3835081   \n",
       "72            CEP104:9731  542193     1  3837319   \n",
       "73            CEP104:9731  542192     1  3839096   \n",
       "74            CEP104:9731  221274     1  3839606   \n",
       "75            CEP104:9731  221276     1  3844977   \n",
       "76           NPHP4:261734    3403     1  5873294   \n",
       "77           NPHP4:261734    3402     1  5874946   \n",
       "78           NPHP4:261734    3400     1  5887394   \n",
       "79           NPHP4:261734    3399     1  5887403   \n",
       "80           NPHP4:261734    3404     1  5887436   \n",
       "81           NPHP4:261734    3405     1  5904788   \n",
       "82             ESPN:83715    4420     1  6451926   \n",
       "83             ESPN:83715    4422     1  6452092   \n",
       "84             ESPN:83715  560535     1  6457223   \n",
       "85             ESPN:83715  562076     1  6460076   \n",
       "86          PLEKHG5:57449  566096     1  6468294   \n",
       "87          PLEKHG5:57449   60779     1  6468378   \n",
       "88          PLEKHG5:57449  468906     1  6468468   \n",
       "89          PLEKHG5:57449   60780     1  6469396   \n",
       "90          PLEKHG5:57449    1019     1  6469444   \n",
       "91          PLEKHG5:57449   60778     1  6473051   \n",
       "92          PLEKHG5:57449  549691     1  6476021   \n",
       "93          PLEKHG5:57449   60777     1  6477533   \n",
       "94           CAMTA1:23261  521550     1  7249607   \n",
       "95              PER3:8863  242410     1  7809893   \n",
       "96              PER3:8863  242411     1  7809900   \n",
       "97            PARK7:11315    7065     1  7962863   \n",
       "98            PARK7:11315  573287     1  7965336   \n",
       "99            PARK7:11315    7069     1  7965348   \n",
       "\n",
       "                                                  ref       alt  \\\n",
       "0                                                   C         T   \n",
       "1                                                   C        CG   \n",
       "2                                                   G         T   \n",
       "3                                                   G         A   \n",
       "4                                                   A         T   \n",
       "5                 CGCCCGCCAGGAGAATGTCTTCAAGAAGTTCGACG         C   \n",
       "6                                                   C         T   \n",
       "7                                                   T        TC   \n",
       "8                                                   G         A   \n",
       "9                                                   G         C   \n",
       "10                                                  G         T   \n",
       "11                                                  G         A   \n",
       "12                                                  G         A   \n",
       "13                                                  A         G   \n",
       "14                                                  C         T   \n",
       "15                                                  A         G   \n",
       "16                                                  C         T   \n",
       "17                            CGGCGCGCCCTGGAGCGGGAGCA         C   \n",
       "18                                                 GA         G   \n",
       "19                                         GTGCTGGCCA         G   \n",
       "20                                                  G         A   \n",
       "21                                                  T         C   \n",
       "22                                                 CG         C   \n",
       "23                                                  G         C   \n",
       "24                                                  G         A   \n",
       "25                                                  C         T   \n",
       "26                                                  G         C   \n",
       "27                                                  T         A   \n",
       "28                                                 CT         C   \n",
       "29                                           TAGGCAGG         C   \n",
       "30                                                 AA         G   \n",
       "31                                                 TG         T   \n",
       "32                                                 AC         A   \n",
       "33                                                 AG         A   \n",
       "34                                                 CA         C   \n",
       "35                                     GGGGGCAGCCGGGT         G   \n",
       "36                                                 CG         C   \n",
       "37                                                 TG         T   \n",
       "38                                                  C         T   \n",
       "39                                                  G         A   \n",
       "40                                                  G         C   \n",
       "41                                                  G         A   \n",
       "42                                                  C         T   \n",
       "43                                                  G         A   \n",
       "44                                               GATC         G   \n",
       "45                                                  T         C   \n",
       "46                                                  A         G   \n",
       "47                                                  T         C   \n",
       "48                                                  T         G   \n",
       "49                                                  C         G   \n",
       "50                                                  G         T   \n",
       "51                                                  G         A   \n",
       "52                                                  G         T   \n",
       "53                                                  C         A   \n",
       "54                                         GTCCGACCGC         G   \n",
       "55                                      GTCCGACCGCTCC         G   \n",
       "56                                                  G         A   \n",
       "57  AGCTCAGCGAGCGCAGCGTCCGCGTGTACCACGAGTGCTTCGGCAA...         A   \n",
       "58                                                  G         C   \n",
       "59                                                CAG         C   \n",
       "60                                                  C         G   \n",
       "61                                                  G         A   \n",
       "62                                                  C        CT   \n",
       "63                                                  C         T   \n",
       "64                                                 AG         A   \n",
       "65                                                  A         G   \n",
       "66                                                  C         T   \n",
       "67                                                  A         T   \n",
       "68                                                  A         G   \n",
       "69                                                  T         C   \n",
       "70                                                  T         C   \n",
       "71                                                  G        GA   \n",
       "72                                        AGGGAGTAACG         A   \n",
       "73                                                  A         C   \n",
       "74                                                  A         G   \n",
       "75                                                  G         A   \n",
       "76                                                 CA         C   \n",
       "77                                                  A         G   \n",
       "78                                                  G         A   \n",
       "79                                                  C         A   \n",
       "80                                                  G         A   \n",
       "81                                                  G         A   \n",
       "82                                                  A         C   \n",
       "83                                                  G         A   \n",
       "84                                TGGAGGCGGGACCTCCTGC         T   \n",
       "85                                                 AC         A   \n",
       "86                                                  G         A   \n",
       "87                                                  C         G   \n",
       "88                                                TGA         T   \n",
       "89                                                  G         A   \n",
       "90                                                  A         G   \n",
       "91                                                  C  CGTCTTCA   \n",
       "92                                                 CG         C   \n",
       "93                                                 TG         T   \n",
       "94                                                 TC         T   \n",
       "95                                                  C         G   \n",
       "96                                                  A         G   \n",
       "97                                                  G         A   \n",
       "98                                                  G        GT   \n",
       "99                                                  G         T   \n",
       "\n",
       "   clinical_severity                                 disease_implicated  \n",
       "0         Pathogenic  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1         Pathogenic  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2         Pathogenic  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3         Pathogenic                     Congenital_myasthenic_syndrome  \n",
       "4         Pathogenic                     Congenital_myasthenic_syndrome  \n",
       "5         Pathogenic                 Myasthenic_syndrome,_congenital,_8  \n",
       "6         Pathogenic  Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "7         Pathogenic                     Congenital_myasthenic_syndrome  \n",
       "8         Pathogenic                     Congenital_myasthenic_syndrome  \n",
       "9         Pathogenic  Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "10        Pathogenic  Congenital_myasthenic_syndrome|Myasthenic_synd...  \n",
       "11        Pathogenic                     Congenital_myasthenic_syndrome  \n",
       "12        Pathogenic                                Immunodeficiency_16  \n",
       "13        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "14        Pathogenic          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "15        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "16        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "17        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "18        Pathogenic          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "19        Pathogenic          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "20        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "21        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "22        Pathogenic          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "23        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "24        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "25        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "26        Pathogenic  Spondyloepimetaphyseal_dysplasia_with_joint_la...  \n",
       "27        Pathogenic          Ehlers-Danlos_syndrome,_progeroid_type,_2  \n",
       "28        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "29        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "30        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "31        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "32        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "33        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "34        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "35        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "36        Pathogenic             Robinow_syndrome,_autosomal_dominant_2  \n",
       "37        Pathogenic                            Inborn_genetic_diseases  \n",
       "38        Pathogenic                                Harel-Yoon_syndrome  \n",
       "39        Pathogenic                          Spinocerebellar_ataxia_21  \n",
       "40        Pathogenic                          Spinocerebellar_ataxia_21  \n",
       "41        Pathogenic                          Spinocerebellar_ataxia_21  \n",
       "42        Pathogenic                          Spinocerebellar_ataxia_21  \n",
       "43        Pathogenic                          Spinocerebellar_ataxia_21  \n",
       "44        Pathogenic                            Inborn_genetic_diseases  \n",
       "45        Pathogenic                            Inborn_genetic_diseases  \n",
       "46        Pathogenic  Neurodevelopmental_Disability|hypotonia|Cleft_...  \n",
       "47        Pathogenic  Neurodevelopmental_Disability|hypotonia|Strabi...  \n",
       "48        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "49        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "50        Pathogenic  Inborn_genetic_diseases|Shprintzen-Goldberg_sy...  \n",
       "51        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "52        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "53        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "54        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "55        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "56        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "57        Pathogenic                       Shprintzen-Goldberg_syndrome  \n",
       "58        Pathogenic                  Peroxisome_biogenesis_disorder_6B  \n",
       "59        Pathogenic  Zellweger_syndrome|Peroxisome_biogenesis_disor...  \n",
       "60        Pathogenic                            Inborn_genetic_diseases  \n",
       "61        Pathogenic  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...  \n",
       "62        Pathogenic  Peroxisome_biogenesis_disorder,_complementatio...  \n",
       "63        Pathogenic  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...  \n",
       "64        Pathogenic                  Peroxisome_biogenesis_disorder_6B  \n",
       "65        Pathogenic  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...  \n",
       "66        Pathogenic                         Dilated_cardiomyopathy_1LL  \n",
       "67        Pathogenic                   Left_ventricular_noncompaction_8  \n",
       "68        Pathogenic                   Left_ventricular_noncompaction_8  \n",
       "69        Pathogenic                         Dilated_cardiomyopathy_1LL  \n",
       "70        Pathogenic                                Joubert_syndrome_25  \n",
       "71        Pathogenic                                Joubert_syndrome_25  \n",
       "72        Pathogenic                                Joubert_syndrome_25  \n",
       "73        Pathogenic                                Joubert_syndrome_25  \n",
       "74        Pathogenic                                Joubert_syndrome_25  \n",
       "75        Pathogenic                                Joubert_syndrome_25  \n",
       "76        Pathogenic                                 Nephronophthisis_4  \n",
       "77        Pathogenic                                 Nephronophthisis_4  \n",
       "78        Pathogenic                Nephronophthisis|Nephronophthisis_4  \n",
       "79        Pathogenic                                 Nephronophthisis_4  \n",
       "80        Pathogenic           Nephronophthisis|Senior-Loken_syndrome_4  \n",
       "81        Pathogenic  Nephronophthisis|Senior-Loken_syndrome_4|Nephr...  \n",
       "82        Pathogenic  Deafness,_without_vestibular_involvement,_auto...  \n",
       "83        Pathogenic  Deafness,_without_vestibular_involvement,_auto...  \n",
       "84        Pathogenic                             Usher_syndrome,_type_1  \n",
       "85        Pathogenic  Deafness,_autosomal_recessive_36,_with_or_with...  \n",
       "86        Pathogenic  Distal_spinal_muscular_atrophy,_autosomal_rece...  \n",
       "87        Pathogenic  Charcot-Marie-Tooth_disease,_recessive_interme...  \n",
       "88        Pathogenic  Distal_spinal_muscular_atrophy,_autosomal_rece...  \n",
       "89        Pathogenic  Charcot-Marie-Tooth_disease,_recessive_interme...  \n",
       "90        Pathogenic  Distal_spinal_muscular_atrophy,_autosomal_rece...  \n",
       "91        Pathogenic  Charcot-Marie-Tooth_disease,_recessive_interme...  \n",
       "92        Pathogenic  Spinal_muscular_atrophy,_facioscapulohumeral_type  \n",
       "93        Pathogenic  Charcot-Marie-Tooth_disease,_recessive_interme...  \n",
       "94        Pathogenic                            Inborn_genetic_diseases  \n",
       "95        Pathogenic         Advanced_sleep_phase_syndrome,_familial,_3  \n",
       "96        Pathogenic         Advanced_sleep_phase_syndrome,_familial,_3  \n",
       "97        Pathogenic                                Parkinson_disease_7  \n",
       "98        Pathogenic                                Parkinson_disease_7  \n",
       "99        Pathogenic  Parkinson_disease,_autosomal_recessive_early-o...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.options.display.max_rows = 100\n",
    "df_harmful.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2    4857\n",
       "1    3173\n",
       "3    2402\n",
       "Name: chrom, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify which chromosome has the most pathogenic cases.\n",
    "df_harmful.chrom.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47463086     5\n",
       "47803500     4\n",
       "188996179    4\n",
       "47806603     4\n",
       "47806253     4\n",
       "            ..\n",
       "201053539    1\n",
       "111124855    1\n",
       "202464931    1\n",
       "233767092    1\n",
       "219218166    1\n",
       "Name: pos, Length: 9849, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Chromosome positions by pathogenic cases.\n",
    "df_harmful.pos.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MSH2:4436                         610\n",
       "MLH1:4292                         573\n",
       "MSH6:2956                         493\n",
       "COL3A1:1281                       416\n",
       "BMPR2:659                         332\n",
       "                                 ... \n",
       "ACAD9:28976|KIAA1257:57501          1\n",
       "TOR1AIP1:26092|TOR1AIP2:163590      1\n",
       "PTPRF:5792                          1\n",
       "TSGA10:80705                        1\n",
       "IDH1:3417                           1\n",
       "Name: gene_info, Length: 861, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene name by pathogenic cases.\n",
    "df_harmful.gene_info.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    3012\n",
       "C                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    2595\n",
       "T                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    1202\n",
       "A                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    1143\n",
       "TG                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    139\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ... \n",
       "TCTCCCGAGGAGATCAGGGAGCCATCGAAATCCCAAGATCAGACTGATTGAGTTAGAGACCCAGATTCCTTAATGGTCTGGAACCTCCGGAGTGCCTGAAACATGCACACACAACATGCACACAGACACACGTATACACATGCACACACGCTCCCAAACACATGCACATACAGAGTCACTTCCCTCGCTTCACATACTAAGTCCTGAGGTGGCTTGAATTTTCACTTAAGATTCAGGGAGGGAAAGGGGGAATTCCTGTTCTATGTTCTGGTCAGGCAGTGACCAACCCTGGGGCAAGGAACTGAACTTTGGGGGTACACTGGAAGCACTTAAGAAAATGGCAAAAGTTTTAGAGTCTCCTCTCCCTGACCCGGGGGTCTCAAACATCTGCTGGGGGCTATTAGAGCGAGACATCACCCATCCCCAGATCTGGGAAACCAACACTGCCACCTCTCCCAAGTCACAGCTCCCCACCACAAAAACGCAAATGCTCCCCTCTTCCACGAAGTCCCCAGCACCTGCCTTTACCTGAACAGGAACGAGCCTTGCTGCGCTGCTTAGGGGTGAAGCTGGAGGCCGGTCCCCCCAGGAAGCCTCCGGGGTGGAAGAGTCCGCTGCTGTAGTCGTGGGCAGCCGCCGGCACATAGGAGGGGTAGGTGGGGATGGGGTGGTGTGTAGCAGGCTGGGTGCCCATAGTAGCTAGGCCTGGGCGCAGGGGACTGCCACTTTCCATCTTCATGCTCTCCGTCAGTGACACCTGGTACTTGACGCCGTCCTTGTCCTCTCCTCGGGCTGCACTACCCCCCGCGGAAGATGAGGCTGGAGACGCAGCCCCCGTGGTGCTAGGGTCAGGAGACACTTCTTTGGGTGGCGTGGGTGGGAAGCCGAAAAGGTGGGAGCCAGAGTGGGCTGCTGTAGGGGTGAGGGAGGCCACTGAGCTCCCGCTGCCTCCCCCGCTCCCACCCCCAGCCCCTGGGTACACAGAGAGTGGGCCTCCAGGGCCTCCAGCAGCTGAGGGGTGCAGTGGCGTCTTGGAGAAGGGGCTCACGGTCCAGGGGTTGTGGTGGTGGGCCGCAGCGGCAGAGAGGGCTGCTTTGCCCCCGTCCAGCCAGGGCAAACCCGGGCTGTGCAACAAGTGTGGGCGGCACATCTGGCCTCCGGTCAGGCGGGCTGCGGGCAAAGAGAGAGAGGATCAGGGTGGGCAGAAAGATCAGGGTAGGCAGAGCTAGGGACGCCCCTGACAGACATTGAGATCACGACTCCCAGAACCAGCAGTCATCCCCTCCCCAAAGAAAGCCAGAAACATAATACCCCACCGGTAATAATCAGGAATGTCAGTCCAAGCTGAAGGACAAGTGGCATAGAAGGAACCCCACCGGACAGACCCTACAGGGAACCCTCACAGGCCAGCTGGAAGTGGGCAGAAACCCTGTGGGTCCCAGACCCTCCCCAATCGGCCGCTGCTCCCACCTCTCCCGCCCCAATTTTTCAGCAGCTCGATTCCTGCGGATCCTACATCCGGGAAGCAAGCAGACGGGCCCTCCTCCCCTCCCTCGCCTGGCGCGCGGCGCCTGGGTTCTCATCACCACGGGCCCAGTGCTCACCGTGCGCGGGGCTGTAGGAGACGCGCGCCCGCGCGTGAGCGGGGTTGGCATAGTAGGGGTTGCCCTGCGAGTCGAGGTGATTGAAGAAGACGTCCACCTCGTCTGGAGGCAGCAGCTGCGCGGGTTCCATGTAGTTGTGCGCCAGGCCCGGGTGGTGTGAGTCGGGGTGCTGCGCATTCAGCACGGCCGGGTGCGCCATCCAGCGCGGCTGCTCGGGCGCCACCTCCATGGCCGGCGGCGGCGGCTCAGGGTCTGGGTGCAGACGGCAACGGCCCTGCGCGAGGAAGGGGGAGTGAGGCGTGCCGCCAGCGCCTGACACCCCCCAAAGTCCCACCACGAGGTGTCCCGCACGCCACGGAGCCCCAGCCCAGATCCGGCGAGAAAGAGCACCAGTCCCGGGTGGGAGGAAAGCCCAAGGCTCAAAACGAA       1\n",
       "GAAAA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   1\n",
       "GTCT                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    1\n",
       "GCTCGGGGTAATGACGGTGC                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    1\n",
       "GGCCTCTCTGGTGTTCCTGCAGACCCC                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             1\n",
       "Name: ref, Length: 752, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reference base by pathogenic cases\n",
    "df_harmful.ref.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A         3043\n",
       "T         2856\n",
       "C         1801\n",
       "G         1526\n",
       "CT         100\n",
       "          ... \n",
       "CTGCA        1\n",
       "GCTCC        1\n",
       "TAAGTC       1\n",
       "GTTCA        1\n",
       "CAGGG        1\n",
       "Name: alt, Length: 361, dtype: int64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Alternate base by pathogenic cases.\n",
    "df_harmful.alt.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA 0\n",
      "AC 216\n",
      "AG 470\n",
      "AT 221\n",
      "CA 457\n",
      "CC 0\n",
      "CG 351\n",
      "CT 1504\n",
      "GA 1796\n",
      "GC 383\n",
      "GG 0\n",
      "GT 573\n",
      "TA 223\n",
      "TC 475\n",
      "TG 227\n",
      "TT 0\n"
     ]
    }
   ],
   "source": [
    "# Pairs of reference base and alternate base by pathogenic cases.\n",
    "\n",
    "pairs = []\n",
    "for i in ['A', 'C', 'G', 'T']:\n",
    "    for j in ['A', 'C', 'G', 'T']:\n",
    "        pairs.extend((i,j,df_harmful[(df_harmful.ref == i)&(df_harmful.alt == j)]['alt'].count()))\n",
    "        print(i+j, df_harmful[(df_harmful.ref == i)&(df_harmful.alt == j)]['alt'].count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1504"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_harmful[(df_harmful.ref == 'C')&(df_harmful.alt == 'T')]['alt'].count()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MSH2:4436                         610\n",
       "MLH1:4292                         573\n",
       "MSH6:2956                         493\n",
       "COL3A1:1281                       416\n",
       "BMPR2:659                         332\n",
       "                                 ... \n",
       "ACAD9:28976|KIAA1257:57501          1\n",
       "TOR1AIP1:26092|TOR1AIP2:163590      1\n",
       "PTPRF:5792                          1\n",
       "TSGA10:80705                        1\n",
       "IDH1:3417                           1\n",
       "Name: gene_info, Length: 861, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_harmful.gene_info.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# My definition of harmful mutation is any clinical severity listed as 'Pathogenic' or 'Pathogenic,........'.  I found 12,330 harmful gene and mutation variations. I have more description is the cell below.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4) \n",
    "\n",
    "The first step I took in creating a dataframe of dangerous gene and mutation combinations was read in the portion of the text file that did not start with a '#'.  I read all of the records from the .txt file into a list called records.  Next, I printed the first few items in the list to inspect their contents.  There were a total of 102,322 records--including the header.\n",
    "    \n",
    "Once the data was read into the Python Notebook it was time to do some manipulation.  I wanted to eliminate the first row of records and split the records at the '\\t'.  In order to do this, I used a for loop with a split command.  Doing so split up the contents of each list within the list of lists into multiple elements.  For example, I went from a data structure of records[0:n] to rows[0:n][0:n].  To ensure there were no discrepancies in the data, I checked the length of each list.  They were all the same length!\n",
    "\n",
    "Now it was time to extract the GENEINFO, CLNSIG, and CLNDN for each row in rows.  To do this, I used three regular expressions and appended them to three different lists.  If the data structure didn't have one of the GENEINFO, CLNSIG, and CLNDN elements, I set it to 'Not_Given'.  After creating the three lists, I created there pandas series from them.  I then constructed the dataframe and began analyzing the data set.\n",
    "\n",
    "Data cleaning / sanity check and final dataframe:\n",
    "\n",
    "1. The shape of the original dataframe is (102321, 8).\n",
    "\n",
    "2. Check dtypes.  I didn't convert anything.  Althoough I thought about convering some variables to int.\n",
    "\n",
    "3. Check the differnent types of clinical severity.  There are 1,797 gene and muation variations without a clinical sverity listed. I will use Pathogenic and Pathogenic, as the mutations that I consider harmful.\n",
    "\n",
    "4. Inspect all variables in the data set.\n",
    "\n",
    "5. Check for NaNs.\n",
    "\n",
    "6. Search for any variables with '.'. I found 86 in variable alt.\n",
    "\n",
    "7. Replace the '.' with 'Not_Given'\n",
    "\n",
    "8. Look for variabale with 'Not_Given'.  A total of 14,571 variables are missing data.\n",
    "\n",
    "9. Filter data by clinical severity = 'Pathogenic' or 'Pathogenic,'.  Create dataframe df_all_harmful.\n",
    "\n",
    "10. Filter out all 'Not_Given' data from df_all_harmful.  Create dataframe df_harmful.  There were 1,898 rows of data eliminated.\n",
    "\n",
    "11.  All gene and mutation variations have a unique ID.  I use the unique ID to determine the variations that are different\n",
    "\n",
    "Explanation to boss:\n",
    "\n",
    "I have created two data frames, one preliminary that has all the harmful gene and mutation combinations with missing data and one final data frame that has complete records only.  The preliminary data frame is df_all_harmful.  It contains 12,330 gene and mutation combinations.  Within this data frame, the missing data is comprised of the following:\n",
    "Gene Info: 6\n",
    "Alternate Bases: 3\n",
    "Disease Implications: 1890\n",
    "\n",
    "The final data frame I created contains all harmful data that is a complete record.  There are 10,432 records in this data frame.  My definition of harmful mutation is any clinical severity listed as 'Pathogenic' or 'Pathogenic,'.  There were few descriptions such as 'likely pathogenic' or 'conflicting' etc.  Moreover, I left these out from my final data set.  Also, there were 1,797 records missing clinical severity data.\n",
    "\n",
    "Assumptions:\n",
    "\n",
    "1. There are a few REF and ALT columns that have multiple chromosome bases.  The description of the REF and ALT is plural so I wasn't sure if they were correct, however, I left them until told otherwise.\n",
    "\n",
    "2. A few of the GENEINFO entries were seperated by '|'.  I left these as is.\n",
    "\n",
    "Insights\n",
    "\n",
    "1. Chromosome number 2 has the most pathogenic cases.\n",
    "\n",
    "2. Gene MSH2:4436 has the most pathogenic cases.\n",
    "\n",
    "3. Reference base and alternate base GA have the most pathogenic cases.\n",
    "\n",
    "All in all, the final data set contains all harmful gene and mutation combinations with complete records."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
